{"cells":[{"cell_type":"code","execution_count":1,"metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[33mWARNING: The directory '/home/dnanexus/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.\u001b[0m\u001b[33m\n","\u001b[0m\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager, possibly rendering your system unusable.It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv. Use the --root-user-action option if you know what you are doing and want to suppress this warning.\u001b[0m\u001b[33m\n","\u001b[0m\u001b[33mWARNING: The directory '/home/dnanexus/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.\u001b[0m\u001b[33m\n","\u001b[0m\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager, possibly rendering your system unusable.It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv. Use the --root-user-action option if you know what you are doing and want to suppress this warning.\u001b[0m\u001b[33m\n","\u001b[0m"]}],"source":["# -*- coding: utf-8 -*-\n","\n","\"\"\"\n","ensemble GA for binary class with n-Fold CV\n","\n","\"\"\"\n","!pip install -q scikit-learn knnimpute scikit-survival\n","import os\n","import random\n","import statistics\n","\n","import numpy as np\n","import scipy.stats as st\n","import time\n","import math\n","import sys\n","\n","from copy import deepcopy\n","import argparse\n","\n","import copy\n","\n","from sklearn.model_selection import StratifiedKFold\n","\n","from sklearn.ensemble import RandomForestClassifier\n","from collections import Counter # for counting classes from labels list\n","from sklearn.metrics import fbeta_score, make_scorer # to make f2 for cross_validate\n","from scipy.spatial import distance\n","\n","from sklearn.decomposition import PCA\n","from sklearn.neighbors import LocalOutlierFactor\n","import pandas as pd\n","from sklearn import svm\n","from joblib import Parallel, delayed\n","from joblib import dump, load\n","\n","from sklearn.metrics import precision_score\n","from sklearn.metrics import recall_score\n","from sklearn.metrics import roc_auc_score, roc_curve, auc\n","from sklearn.metrics import accuracy_score\n","\n","from sklearn.preprocessing import label_binarize\n","from sklearn.multiclass import OneVsRestClassifier\n","\n","import itertools # for permutation of multi-labels\n","from sklearn.metrics import confusion_matrix\n","from sklearn.preprocessing import RobustScaler\n","from sksurv.metrics import concordance_index_censored\n","\n","import joblib\n","\n","        \n","from sklearn.feature_selection import SelectKBest\n","from sklearn.feature_selection import mutual_info_classif\n","\n","## install from: https://pypi.org/project/knnimpute/\n","from knnimpute import (\n","    knn_impute_few_observed,\n","    knn_impute_with_argpartition,\n","    knn_impute_optimistic,\n","    knn_impute_reference,\n",")\n","\n","\"\"\"\n","ensemble GA for binary class with n-Fold CV\n","\n","\"\"\"\n","!pip install -q scikit-learn knnimpute scikit-survival\n","import os\n","import random\n","import statistics\n","\n","import numpy as np\n","import scipy.stats as st\n","import time\n","import math\n","import sys\n","\n","from copy import deepcopy\n","import argparse\n","\n","import copy\n","\n","from sklearn.model_selection import StratifiedKFold\n","\n","from sklearn.ensemble import RandomForestClassifier\n","from collections import Counter # for counting classes from labels list\n","from sklearn.metrics import fbeta_score, make_scorer # to make f2 for cross_validate\n","from scipy.spatial import distance\n","\n","from sklearn.decomposition import PCA\n","from sklearn.neighbors import LocalOutlierFactor\n","import pandas as pd\n","from sklearn import svm\n","from joblib import Parallel, delayed\n","from joblib import dump, load\n","\n","from sklearn.metrics import precision_score\n","from sklearn.metrics import recall_score\n","from sklearn.metrics import roc_auc_score, roc_curve, auc\n","from sklearn.metrics import accuracy_score\n","\n","from sklearn.preprocessing import label_binarize\n","from sklearn.multiclass import OneVsRestClassifier\n","\n","import itertools # for permutation of multi-labels\n","from sklearn.metrics import confusion_matrix\n","from sklearn.preprocessing import RobustScaler\n","from sksurv.metrics import concordance_index_censored\n","\n","import joblib\n","\n","        \n","from sklearn.feature_selection import SelectKBest\n","from sklearn.feature_selection import mutual_info_classif\n","\n","## install from: https://pypi.org/project/knnimpute/\n","from knnimpute import (\n","    knn_impute_few_observed,\n","    knn_impute_with_argpartition,\n","    knn_impute_optimistic,\n","    knn_impute_reference,\n",")"]},{"cell_type":"code","execution_count":2,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["import multiprocessing.managers\n","import copy\n","import concurrent.futures"]},{"cell_type":"code","execution_count":3,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["import argparse"]},{"cell_type":"code","execution_count":4,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["path = os.getcwd() #get cwd and append the file path to it  \n","\n","# create output folder for pre-processing with timestamp to the path\n","tstamp = time.strftime('%Y_%m_%d')\n","output_folder='Output_PreProcess'+ os.sep+ str(tstamp) # get the path separator based on os\n","if not os.path.exists(output_folder):\n","    os.makedirs(output_folder)\n","pathProcess = os.path.join(path + os.sep, output_folder)\n","\n","# create output folder for feature Selection with timestamp to the path\n","output_folder='Output_FeatSelect'+ os.sep+ str(tstamp) # get the path separator based on os\n","if not os.path.exists(output_folder):\n","    os.makedirs(output_folder)\n","pathFeature = os.path.join(path + os.sep, output_folder)\n","\n","# create output folder for Model Saving with timestamp to the path\n","output_folder='Output_Model'+ os.sep+ str(tstamp) # get the path separator based on os\n","if not os.path.exists(output_folder):\n","    os.makedirs(output_folder)\n","pathModel = os.path.join(path + os.sep, output_folder)\n","'SEED CHANGE'\n","SEED = 1234 # seed for random state\n","'SEED CHANGE'\n","random.seed(1234) #Initialize random generator seed with the current local time\n","#NUM_CORES = cpu_count()\n","import psutil\n","NUM_CORES = psutil.cpu_count()"]},{"cell_type":"code","execution_count":5,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def load_model(file):\n","    remote_path = f\"SADIQ/21 models/{file}\"  \n","    !dx download \"{remote_path}\" --overwrite\n","    model = joblib.load(file)\n","    return model"]},{"cell_type":"code","execution_count":6,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["cleaned_model_to_csvs = {\n","    \"Sociodemographics\": [\n","        \"Sociodemographics\"\n","    ],\n","\n","    \"Sociodemographics_oneDrop_qrisk\": [\n","        \"Sociodemographics\",\n","        \"oneDrop\",\n","        \"qrisk\"\n","    ],\n","    \"Sociodemographics_Health and medical history\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\"\n","    ],\n","\n","    \"Sociodemographics_Health and medical history_oneDrop_qrisk\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"oneDrop\",\n","        \"qrisk\"\n","    ],\n","    \"Sociodemographics_Health and medical history_Sex-specific factors\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\"\n","    ],\n","\n","    \"Sociodemographics_Health and medical history_Sex-specific factors_oneDrop_qrisk\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\",\n","        \"oneDrop\",\n","        \"qrisk\"\n","    ],\n","    \"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\",\n","        \"Early life factors\"\n","    ],\n","\n","    \"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_oneDrop_qrisk\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\",\n","        \"Early life factors\",\n","        \"oneDrop\",\n","        \"qrisk\"\n","    ],\n","    \"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\",\n","        \"Early life factors\",\n","        \"Family history\"\n","    ],\n","\n","    \"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_oneDrop_qrisk\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\",\n","        \"Early life factors\",\n","        \"Family history\",\n","        \"oneDrop\",\n","        \"qrisk\"\n","    ],\n","    \"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\",\n","        \"Early life factors\",\n","        \"Family history\",\n","        \"Lifestyle and environment\"\n","    ],\n","    \"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment_oneDrop_qrisk\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\",\n","        \"Early life factors\",\n","        \"Family history\",\n","        \"Lifestyle and environment\",\n","        \"oneDrop\",\n","        \"qrisk\"\n","    ],\n","    \"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment_Psychosocial factors\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\",\n","        \"Early life factors\",\n","        \"Family history\",\n","        \"Lifestyle and environment\",\n","        \"Psychosocial factors\"\n","    ],\n","    \"Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\",\n","        \"Early life factors\",\n","        \"Family history\",\n","        \"Lifestyle and environment\",\n","        \"Psychosocial factors\",\n","        \"oneDrop\"\n","    ],\n","    \"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment_Psychosocial factors_oneDrop\":[\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\",\n","        \"Early life factors\",\n","        \"Family history\",\n","        \"Lifestyle and environment\",\n","        \"Psychosocial factors\",\n","        \"oneDrop\"\n","    ],\n","    \"Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop_qrisk\": [\n","        \"Sociodemographics\",\n","        \"Health and medical history\",\n","        \"Sex-specific factors\",\n","        \"Early life factors\",\n","        \"Family history\",\n","        \"Lifestyle and environment\",\n","        \"Psychosocial factors\",\n","        \"oneDrop\",\n","        \"qrisk\"\n","    ],\n","    \n","\n","\n","}"]},{"cell_type":"code","execution_count":7,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def to_binary(x):                                 ### FIX – was missing\n","    \"\"\"\n","    Normalise typical yes/no, true/false, 1/0 strings to 0 / 1.\n","    Leaves other values unchanged so pandas keeps dtypes.\n","    \"\"\"\n","    if isinstance(x, str):\n","        s = x.strip().lower()\n","        if s in {\"yes\", \"y\", \"true\", \"1\"}:\n","            return 1\n","        if s in {\"no\", \"n\", \"false\", \"0\"}:\n","            return 0\n","    return x\n"]},{"cell_type":"code","execution_count":8,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def predictIndividual(model, columns_to_use, dfTestIndividual):\n","    # 1. Filter test DataFrame for required columns\n","    filtered_df = dfTestIndividual[\n","        [col for col in dfTestIndividual.columns if any(key in col for key in columns_to_use)]\n","        + ['clinicalrisk_Age.at.recruitment']\n","    ].copy()  \n","\n","    # 2. Ensure correct feature order\n","    model_features = model.feature_names_in_\n","\n","    for feature in model_features:\n","        if feature not in filtered_df.columns:\n","            filtered_df[feature] = 0  # Safe to modify now\n","\n","    filtered_df = filtered_df[model_features]\n","\n","    # 3. Return the risk score\n","    return model.predict(filtered_df)\n"]},{"cell_type":"code","execution_count":9,"metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[2K[===========================================================>] Completed 967,397,153 of 967,397,153 bytes (100%) /opt/notebooks/dfTrain.csvv\n","\u001b[2K[===========================================================>] Completed 968,855,153 of 968,855,153 bytes (100%) /opt/notebooks/dfTest.csvv\n"]},{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Unnamed: 0</th>\n","      <th>eid</th>\n","      <th>oneDrop_metabolomics_Apolipoprotein.B</th>\n","      <th>oneDrop_metabolomics_Apolipoprotein.A1</th>\n","      <th>oneDrop_metabolomics_Alanine</th>\n","      <th>oneDrop_metabolomics_Glutamine</th>\n","      <th>oneDrop_metabolomics_Glycine</th>\n","      <th>oneDrop_metabolomics_Histidine</th>\n","      <th>oneDrop_metabolomics_Total.Concentration.of.Branched.Chain.Amino.Acids..Leucine...Isoleucine...Valine.</th>\n","      <th>oneDrop_metabolomics_Isoleucine</th>\n","      <th>...</th>\n","      <th>MEL_at_base</th>\n","      <th>CRC_at_base</th>\n","      <th>PC_at_base</th>\n","      <th>BC_at_base</th>\n","      <th>PD_at_base</th>\n","      <th>OP_at_base</th>\n","      <th>CAT_at_base</th>\n","      <th>POAG_at_base</th>\n","      <th>HT_at_base</th>\n","      <th>testtrain</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>0</td>\n","      <td>-1.000592</td>\n","      <td>-0.313596</td>\n","      <td>-1.083320</td>\n","      <td>-0.317599</td>\n","      <td>-0.774265</td>\n","      <td>0.168655</td>\n","      <td>-1.016457</td>\n","      <td>-0.442465</td>\n","      <td>0.200798</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>1</td>\n","      <td>-1.000577</td>\n","      <td>-0.260364</td>\n","      <td>1.600931</td>\n","      <td>-0.218882</td>\n","      <td>0.122720</td>\n","      <td>0.652646</td>\n","      <td>0.621709</td>\n","      <td>0.150410</td>\n","      <td>-0.006246</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>5</td>\n","      <td>-1.000543</td>\n","      <td>0.724643</td>\n","      <td>0.212878</td>\n","      <td>-0.148123</td>\n","      <td>0.167354</td>\n","      <td>1.267813</td>\n","      <td>-0.874707</td>\n","      <td>0.025624</td>\n","      <td>-0.043533</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>6</td>\n","      <td>-1.000539</td>\n","      <td>0.743573</td>\n","      <td>-0.346315</td>\n","      <td>0.702509</td>\n","      <td>1.137018</td>\n","      <td>-0.169060</td>\n","      <td>0.863078</td>\n","      <td>1.844256</td>\n","      <td>2.094369</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>7</td>\n","      <td>-1.000536</td>\n","      <td>0.527013</td>\n","      <td>0.528782</td>\n","      <td>-0.346036</td>\n","      <td>-0.342773</td>\n","      <td>0.034109</td>\n","      <td>-0.589745</td>\n","      <td>-0.844637</td>\n","      <td>-0.657025</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>120294</th>\n","      <td>257225</td>\n","      <td>1.000970</td>\n","      <td>0.011964</td>\n","      <td>0.025446</td>\n","      <td>0.064152</td>\n","      <td>0.242324</td>\n","      <td>0.345130</td>\n","      <td>-0.261630</td>\n","      <td>0.698133</td>\n","      <td>0.225687</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>120295</th>\n","      <td>257227</td>\n","      <td>1.000981</td>\n","      <td>-0.547458</td>\n","      <td>0.683941</td>\n","      <td>-0.363845</td>\n","      <td>-0.061498</td>\n","      <td>-0.375194</td>\n","      <td>1.805369</td>\n","      <td>1.318727</td>\n","      <td>1.213963</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>120296</th>\n","      <td>257229</td>\n","      <td>1.000988</td>\n","      <td>0.501306</td>\n","      <td>-0.301319</td>\n","      <td>-0.632038</td>\n","      <td>-0.200532</td>\n","      <td>-0.735962</td>\n","      <td>-0.212039</td>\n","      <td>-0.325872</td>\n","      <td>-0.267634</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>120297</th>\n","      <td>257233</td>\n","      <td>1.001018</td>\n","      <td>0.438723</td>\n","      <td>0.680217</td>\n","      <td>-0.181540</td>\n","      <td>-0.272936</td>\n","      <td>-1.094466</td>\n","      <td>1.705383</td>\n","      <td>1.684035</td>\n","      <td>1.512493</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>120298</th>\n","      <td>257235</td>\n","      <td>1.001029</td>\n","      <td>0.428388</td>\n","      <td>0.711559</td>\n","      <td>0.935465</td>\n","      <td>0.461278</td>\n","      <td>0.062285</td>\n","      <td>-0.052809</td>\n","      <td>0.570204</td>\n","      <td>0.314674</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>train</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>120299 rows × 1625 columns</p>\n","</div>"],"text/plain":["        Unnamed: 0       eid  oneDrop_metabolomics_Apolipoprotein.B  \\\n","0                0 -1.000592                              -0.313596   \n","1                1 -1.000577                              -0.260364   \n","2                5 -1.000543                               0.724643   \n","3                6 -1.000539                               0.743573   \n","4                7 -1.000536                               0.527013   \n","...            ...       ...                                    ...   \n","120294      257225  1.000970                               0.011964   \n","120295      257227  1.000981                              -0.547458   \n","120296      257229  1.000988                               0.501306   \n","120297      257233  1.001018                               0.438723   \n","120298      257235  1.001029                               0.428388   \n","\n","        oneDrop_metabolomics_Apolipoprotein.A1  oneDrop_metabolomics_Alanine  \\\n","0                                    -1.083320                     -0.317599   \n","1                                     1.600931                     -0.218882   \n","2                                     0.212878                     -0.148123   \n","3                                    -0.346315                      0.702509   \n","4                                     0.528782                     -0.346036   \n","...                                        ...                           ...   \n","120294                                0.025446                      0.064152   \n","120295                                0.683941                     -0.363845   \n","120296                               -0.301319                     -0.632038   \n","120297                                0.680217                     -0.181540   \n","120298                                0.711559                      0.935465   \n","\n","        oneDrop_metabolomics_Glutamine  oneDrop_metabolomics_Glycine  \\\n","0                            -0.774265                      0.168655   \n","1                             0.122720                      0.652646   \n","2                             0.167354                      1.267813   \n","3                             1.137018                     -0.169060   \n","4                            -0.342773                      0.034109   \n","...                                ...                           ...   \n","120294                        0.242324                      0.345130   \n","120295                       -0.061498                     -0.375194   \n","120296                       -0.200532                     -0.735962   \n","120297                       -0.272936                     -1.094466   \n","120298                        0.461278                      0.062285   \n","\n","        oneDrop_metabolomics_Histidine  \\\n","0                            -1.016457   \n","1                             0.621709   \n","2                            -0.874707   \n","3                             0.863078   \n","4                            -0.589745   \n","...                                ...   \n","120294                       -0.261630   \n","120295                        1.805369   \n","120296                       -0.212039   \n","120297                        1.705383   \n","120298                       -0.052809   \n","\n","        oneDrop_metabolomics_Total.Concentration.of.Branched.Chain.Amino.Acids..Leucine...Isoleucine...Valine.  \\\n","0                                               -0.442465                                                        \n","1                                                0.150410                                                        \n","2                                                0.025624                                                        \n","3                                                1.844256                                                        \n","4                                               -0.844637                                                        \n","...                                                   ...                                                        \n","120294                                           0.698133                                                        \n","120295                                           1.318727                                                        \n","120296                                          -0.325872                                                        \n","120297                                           1.684035                                                        \n","120298                                           0.570204                                                        \n","\n","        oneDrop_metabolomics_Isoleucine  ...  MEL_at_base  CRC_at_base  \\\n","0                              0.200798  ...            0            0   \n","1                             -0.006246  ...            0            0   \n","2                             -0.043533  ...            0            0   \n","3                              2.094369  ...            0            0   \n","4                             -0.657025  ...            0            0   \n","...                                 ...  ...          ...          ...   \n","120294                         0.225687  ...            0            0   \n","120295                         1.213963  ...            0            0   \n","120296                        -0.267634  ...            0            0   \n","120297                         1.512493  ...            0            1   \n","120298                         0.314674  ...            0            0   \n","\n","        PC_at_base  BC_at_base  PD_at_base  OP_at_base  CAT_at_base  \\\n","0                0           0           0           0            0   \n","1                0           0           0           0            0   \n","2                0           0           0           0            0   \n","3                0           0           0           0            0   \n","4                0           0           0           0            0   \n","...            ...         ...         ...         ...          ...   \n","120294           0           0           0           0            0   \n","120295           0           0           0           0            0   \n","120296           0           0           0           0            0   \n","120297           0           0           0           0            0   \n","120298           0           0           0           0            0   \n","\n","        POAG_at_base  HT_at_base  testtrain  \n","0                  0           0      train  \n","1                  0           0      train  \n","2                  0           1      train  \n","3                  0           0      train  \n","4                  0           0      train  \n","...              ...         ...        ...  \n","120294             0           0      train  \n","120295             0           0      train  \n","120296             0           0      train  \n","120297             0           0      train  \n","120298             0           0      train  \n","\n","[120299 rows x 1625 columns]"]},"execution_count":9,"metadata":{},"output_type":"execute_result"}],"source":["def load_dataset(remote_path):\n","    import os\n","    import pandas as pd\n","\n","    local_filename = os.path.basename(remote_path)  # extracts just 'dfTrain.csv'\n","    !dx download \"{remote_path}\" --overwrite\n","    dataset = pd.read_csv(local_filename)\n","    return dataset\n","\n","dfTrain = load_dataset(\"SADIQ/30-04/dfTrain.csv\")\n","dfTest = load_dataset(\"SADIQ/30-04/dfTest.csv\")\n","dfTrain "]},{"cell_type":"code","execution_count":10,"metadata":{"trusted":true},"outputs":[],"source":["import pandas as pd\n","\n","dfTrain['followup_bin'] = pd.qcut(dfTrain['CVD_followup'], q=10, duplicates='drop')\n","'SEED CHANGE'\n","\n","sampled_df = dfTrain.groupby('followup_bin', group_keys=False).apply(\n","    lambda x: x.sample(frac=5000/len(dfTrain), random_state=1234)\n",")\n","\n","sampled_df = sampled_df.drop(columns='followup_bin')"]},{"cell_type":"code","execution_count":11,"metadata":{"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["Original dfTrain:\n","count    120299.000000\n","mean         12.828773\n","std           2.811520\n","min           0.002738\n","25%          12.668036\n","50%          13.596167\n","75%          14.362765\n","max          15.865845\n","Name: CVD_followup, dtype: float64\n","\n","Sampled df:\n","count    4999.000000\n","mean       12.836918\n","std         2.783891\n","min         0.019165\n","25%        12.665298\n","50%        13.596167\n","75%        14.364134\n","max        15.542779\n","Name: CVD_followup, dtype: float64\n"]}],"source":["print(\"Original dfTrain:\")\n","print(dfTrain['CVD_followup'].describe())\n","\n","print(\"\\nSampled df:\")\n","print(sampled_df['CVD_followup'].describe())"]},{"cell_type":"code","execution_count":12,"metadata":{"tags":[],"trusted":true},"outputs":[{"data":{"text/plain":["1625"]},"execution_count":12,"metadata":{},"output_type":"execute_result"}],"source":["len(sampled_df.columns)"]},{"cell_type":"code","execution_count":13,"metadata":{"tags":[],"trusted":true},"outputs":[{"data":{"text/plain":["1626"]},"execution_count":13,"metadata":{},"output_type":"execute_result"}],"source":["len(dfTrain.columns)"]},{"cell_type":"code","execution_count":14,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["dfTrain = sampled_df"]},{"cell_type":"code","execution_count":15,"metadata":{"tags":[],"trusted":true},"outputs":[{"data":{"text/plain":["1625"]},"execution_count":15,"metadata":{},"output_type":"execute_result"}],"source":["len(dfTrain.columns)"]},{"cell_type":"code","execution_count":16,"metadata":{"tags":[],"trusted":true},"outputs":[{"data":{"text/plain":["4999"]},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":["len(dfTrain)"]},{"cell_type":"code","execution_count":17,"metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["120495\n","118376\n","10706\n"]}],"source":["print(len(dfTest))\n","\n","# Drop rows where CVD_followup is between 10 and 12 AND CVD_status is True\n","dfTest = dfTest.drop(\n","    dfTest[\n","        (dfTest['CVD_followup'] >= 10) &\n","        (dfTest['CVD_followup'] <= 12) &\n","        (dfTest['CVD_status'] == True)\n","    ].index\n",")\n","\n","print(len(dfTest))\n","\n","print((dfTest['CVD_status'] == True).sum())\n"]},{"cell_type":"code","execution_count":19,"metadata":{"trusted":true},"outputs":[],"source":["models = [\n","    \"MRMR_COX_Sociodemographics.pkl\",\n","    \"MRMR_COX_Sociodemographics_oneDrop.pkl\",\n","    \"MRMR_COX_FULL_MODEL_+QRISKSociodemographics_oneDrop_qrisk.pkl\",\n","    \n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history.pkl\",\n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history_oneDrop.pkl\",\n","    \"MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_oneDrop_qrisk.pkl\",\n","    \n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors.pkl\",\n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_oneDrop.pkl\",\n","    \"MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_oneDrop_qrisk.pkl\",\n","    \n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors.pkl\",\n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_oneDrop.pkl\",\n","    \"MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_oneDrop_qrisk.pkl\",\n","    \n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history.pkl\",\n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_oneDrop.pkl\",\n","    \"MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_oneDrop_qrisk.pkl\",\n","    \n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment.pkl\",\n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_oneDrop.pkl\",\n","    \"MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_oneDrop_qrisk.pkl\",\n","    \n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors.pkl\",\n","    \"MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop.pkl\",\n","    \"MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop_qrisk.pkl\"\n","    \n","    \n","]"]},{"cell_type":"code","execution_count":20,"metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[2K[===========================================================>] Completed 140,774 of 140,774 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics.pkll\n","\u001b[2K[===========================================================>] Completed 151,715 of 151,715 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_oneDrop.pkll\n","\u001b[2K[===========================================================>] Completed 153,413 of 153,413 bytes (100%) /opt/notebooks/MRMR_COX_FULL_MODEL_+QRISKSociodemographics_oneDrop_qrisk.pkll\n","\u001b[2K[===========================================================>] Completed 155,785 of 155,785 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history.pkll\n","\u001b[2K[===========================================================>] Completed 166,726 of 166,726 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history_oneDrop.pkll\n","\u001b[2K[===========================================================>] Completed 168,424 of 168,424 bytes (100%) /opt/notebooks/MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_oneDrop_qrisk.pkll\n","\u001b[2K[===========================================================>] Completed 162,650 of 162,650 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors.pkll\n","\u001b[2K[===========================================================>] Completed 173,588 of 173,588 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_oneDrop.pkll\n","\u001b[2K[===========================================================>] Completed 175,286 of 175,286 bytes (100%) /opt/notebooks/MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_oneDrop_qrisk.pkll\n","\u001b[2K[===========================================================>] Completed 163,254 of 163,254 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors.pkll\n","\u001b[2K[===========================================================>] Completed 174,192 of 174,192 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_oneDrop.pkll\n","\u001b[2K[===========================================================>] Completed 175,890 of 175,890 bytes (100%) /opt/notebooks/MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_oneDrop_qrisk.pkll\n","\u001b[2K[===========================================================>] Completed 167,512 of 167,512 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history.pkll\n","\u001b[2K[===========================================================>] Completed 178,450 of 178,450 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_oneDrop.pkll\n","\u001b[2K[===========================================================>] Completed 180,148 of 180,148 bytes (100%) /opt/notebooks/MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_oneDrop_qrisk.pkll\n","\u001b[2K[===========================================================>] Completed 198,709 of 198,709 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment.pkll\n","\u001b[2K[===========================================================>] Completed 209,647 of 209,647 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_oneDrop.pkll\n","\u001b[2K[===========================================================>] Completed 211,345 of 211,345 bytes (100%) /opt/notebooks/MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_oneDrop_qrisk.pkll\n","\u001b[2K[===========================================================>] Completed 203,675 of 203,675 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors.pkll\n","\u001b[2K[===========================================================>] Completed 214,613 of 214,613 bytes (100%) /opt/notebooks/MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop.pkll\n","\u001b[2K[===========================================================>] Completed 216,311 of 216,311 bytes (100%) /opt/notebooks/MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop_qrisk.pkll\n","21\n"]}],"source":["modelsPickleList = []\n","for i in range(0,len(models)):\n","    modelsPickleList.append(load_model(models[i]))\n","    \n","print(len(modelsPickleList))\n","    "]},{"cell_type":"code","execution_count":21,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def evaluateBranch(cleanModel, cleanModelColumns,row, bloodModel, bloodModelColumns,ind1,ind2):\n","    predictedOutput = predictIndividual(cleanModel, cleanModelColumns, row)[0]\n","    if predictedOutput > ind2:\n","        'high risk '\n","        predictedOutput = predictIndividual(bloodModel, bloodModelColumns, row)[0]\n","        return predictedOutput\n","    elif predictedOutput < ind1:\n","        'low risk'\n","        return predictedOutput\n","    else:\n","        return 'not right'\n","    "]},{"cell_type":"code","execution_count":22,"metadata":{"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["Original dfTrain:\n","count    4999.000000\n","mean       12.836918\n","std         2.783891\n","min         0.019165\n","25%        12.665298\n","50%        13.596167\n","75%        14.364134\n","max        15.542779\n","Name: CVD_followup, dtype: float64\n","\n","Sampled df:\n","count    4999.000000\n","mean       12.836918\n","std         2.783891\n","min         0.019165\n","25%        12.665298\n","50%        13.596167\n","75%        14.364134\n","max        15.542779\n","Name: CVD_followup, dtype: float64\n"]}],"source":["print(\"Original dfTrain:\")\n","print(dfTrain['CVD_followup'].describe())\n","\n","print(\"\\nSampled df:\")\n","print(sampled_df['CVD_followup'].describe())"]},{"cell_type":"code","execution_count":23,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["dfTestFeaturesCols = dfTest.columns"]},{"cell_type":"code","execution_count":24,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["\n","######### feature selection class ###############################################\n","##################################################################################\n","class DiagnosticOptimization:\n","\n","    # Constructor, called at the start of this class\n","    #def __init__(self,dataset, proteins,min_values, max_values,class_Labels, evaluate_values,categories,\n","    #             population, generations, folds):\n","    def __init__(self,dataset, proteins,class_Labels,testset_dataset, testset_class_Labels, evaluate_values,population, generations):\n","        # variables for these parameters\n","        self.dataset = dataset\n","        self.testset_dataset = testset_dataset\n","        self.proteins = proteins\n","        self.class_Labels = class_Labels # labels or classes as list of list from pre process step\n","        self.testset_class_Labels = testset_class_Labels # labels or classes as list of list from pre process step\n","        self.powerSetClass = [] # initialise powerset class labels\n","        self.evaluate_values = evaluate_values # count of parameters to be optimised\n","        \n","        self.min_values = [-2.5,5,-2.5,5,-2.5,5,-2.5,5,-2.5,5,-2.5,5,-2.5,5,-2.5,0] # min value for parameters to be optimised\n","        self.max_values = [1,9,1,9,1,9,1,9,1,9,1,9,1,9,1,5] # max value for parameters to be optimised\n","                \n","        self.population = int(population) # population size in GA\n","        self.generations = int(generations) # no of generations in GA\n","        self.twoCross_prob = 0.45 # 2-point cross over probability\n","        self.arithCross_prob = 0.45 # arithmetic cross over probability\n","        self.mutation_prob = [0.001,0.2] # mutation probabilityy, min = 0.001 and max = 0.2\n","        self.gaussMut_varPro = [0.1,0.5] # gaussian variance proportion, min=0.1 and max = 0.5\n","        self.avgSimThr = 0.90 # average similariy between the members of the population and its best member\n","                \n","        \n","        \n","        \n","        \n","        self.output_message = '' # output message from feature selection\n","        self.goal_header = \"Questions_proportion,Single_blood_proportion,Risk_scores_proportion,True_positive_rate,False_negative_rate,AUC,Cindex,Weighted_Sum\"\n","        # goal significance for goals position 0 thru 8; includes distance. function getParetoFront() ParetoIndex to be aligned with weightage in goal_significance here\n","        self.goal_significance = [1.0, 1.0, 1.0,1.0,1.0,1.0, 1.0] # weightage for metrics \n","        self.convergeThreshold = 0.001 # convergence threshold for GA\n","        \n","   \n","        # variables for goal values (post pareto front goal tuning) \n","        self.max_eval_per_generation=[0]*self.generations # best values per goal per generation\n","        self.average_eval_per_generation=[0]*self.generations # weighhted avg value per generation across all goals\n","        self.sum_ranked_eval_per_generation=[0]*self.generations # Sum of weighted average goal per generation\n","        self.average_ranked_eval_per_generation=[0]*self.generations # Avg of weighted average goal per generation\n","        self.max_sol_per_generation=[0]*self.generations # best sol per generation\n","        self.Pareto_Wheel_Output = '' # output message from Pareto front and roulette wheel\n","    \n","    # Accessor functions for parameters\n","    def setPopulation(self, value):\n","        self.population = value\n","\n","    def getPopulation(self):\n","        return self.population\n","\n","    def setGenerations(self, value):\n","        self.generations = value\n","\n","    def getGenerations(self):\n","        return self.generations\n","\n","\n","\n","    #This function initializes a population of solutions (float vectors) for optimisation parameters\n","    # This function is called by biomarker_feature_selection()\n","    def initialize(self):\n","        population = self.getPopulation() #population:(integer) it is the number of individual solutions which evolve in parallel\n","        min_threshold = 0.3\n","        min_values=list()\n","        max_values= list() \n","        \n","        for i in range(15):\n","            if i%2==0:\n","                #stop criterion low risk #SADIQ to confirm this is in months or in years (how it appears) In principle we want to have a stopping criterion if the model predict very high score (very close event)\n","                min_values.extend([-2.5])\n","                max_values.extend([1])\n","                #max_values.extend([16])\n","            else:\n","                #stop criterion high risk #SADIQ to confirm this is in months or in years (how it appears) In principle we want to have a stopping criterion if the model predict very low score (very distant event)\n","                min_values.extend([5])\n","                max_values.extend([9])\n","        min_values.extend([0])\n","        max_values.extend([5])\n","        \n","        # initialise population count (size) as list of list with each individual =  parameters + feature length \n","        individuals=[[0 for x in range(len(min_values))] for x in range(population)] \n","        \n","        # generate random values for each individual in the population\n","        for i in range(len(individuals)): # iterate on population size i.e. count of individuals\n","            for j in range(0, len(min_values)): # iterate on size of each individual\n","                individuals[i][j] = random.uniform(min_values[j],max_values[j]) # assign values between min and max at random\n","        \n","        self.output_message += \"Population initialised for %s\"%len(individuals) + \" individuals \\n\"\n","                        \n","        return individuals\n","    \n","    #This function initializes a population of solutions (float vectors) with front1 solutions of previous run saved as FinalSolutionsAll.csv\n","    # this step needs to be manually activated and is not automated to pick previous run final solution into next run\n","    # This function is called by biomarker_feature_selection()\n","    \n","    # function to split data by class; used by wilconxon ranksum for a specific feature index\n","    def splitData_perFeat_byClass(self, feature_index,lblClassCount,lbl):\n","        \n","        listClassData = [] # initialise list of list to hold data by class for a given feature index\n","        classIndex = 0\n","        # split the dataset into class labels for each feature\n","        for k in lblClassCount.keys():\n","            listClassData.append([])\n","            for j in range(len(self.dataset[0])): # segregate observations per protein i.e. feature\n","                if int(self.class_Labels[lbl][j])==int(k):\n","                    listClassData[classIndex].append(float(self.dataset[feature_index][j]))\n","            classIndex = classIndex +1 # increment the counter for class\n","        \n","        return listClassData # returns list of list    \n","    # function to get individual class and ovreall ROC for multiclass problem\n","    def roc_multiclass(self,X_train,X_test,y_train, y_test,classifier_type,clf):\n","        \n","        # get Int labels as self.dictCountLabels has floats; \n","        #gamma_val,C_val,TreeCount\n","        # Dictionary of labels and associated count of samples\n","        dictIntLabels = {int(k):v for k,v in self.dictCountLabels.items()}\n","        \n","        # extract keys from dictionary and hold as list of classes\n","        class_label=[]\n","        class_support = []\n","        for key,values in dictIntLabels.items():\n","            class_label.append(key)\n","            class_support.append(values) # support for each class\n","                \n","        # Binarize the output\n","        ynew_test = label_binarize(y_test, classes=sorted(class_label))\n","        ynew_train = label_binarize(y_train, classes=sorted(class_label))\n","        n_classes = ynew_train.shape[1]\n","        \n","        clf.fit(X_train, ynew_train) # fit on train data\n","        \n","        y_score = clf.decision_function(X_test) # for SVM; for RF instead used predict_proba(); can be automated with if condition on clf type\n","        \n","        # Compute ROC AUC for each class\n","        fpr = dict()\n","        tpr = dict()\n","        roc_auc = dict()\n","        for i in range(n_classes):\n","            fpr[i], tpr[i], _ = roc_curve(ynew_test[:, i], y_score[:, i])\n","            roc_auc[i] = auc(fpr[i], tpr[i])\n","        \n","        # Calculate simple average i.e. micro roc\n","        i=0 # Counter to fetech suport as class labels not necessarilt 0,1,2\n","        roc_average = []\n","        for key, values in roc_auc.items():\n","            roc_average.append(values)\n","            i=i+1\n","\n","        return (sum(roc_average)/i),roc_average # return average and roc for each class as list\n","\n","\n","    '''# function to calculate remaining goal values (as per the fitness function) for each of the individual solutions in a given population\n","    # goal_values inialised to zero and calculated in this function\n","    # This function is called by evaluate_indivdiuals() for each individual solution. final_selected is features post intersection with GA.\n","    # classifier_type - is the value of the 1st position of the individual solution that decides the classifier type'''\n","    def train_Classifier(self, dataTran,final_selected,classifier_type, C_val, Gamma_val,Classchain,TreeCount,ftwo_scorer,finalFlag,multilabel):\n","        inputs = [] # initialise X of the model\n","        Colindices = [] # feature indices selected\n","        outputs = np.ravel(np.array(deepcopy(self.class_Labels)).T) # assign to a variable so that synthetic data labels can be apepnded\n","        # Create classification model dataset aligned with features selected\n","        # X i.e. input to be of the form shape [n_samples, n_features]\n","        # dataTRan is np.ndarray with shape as required by X\n","        for feature in range(len(final_selected)):\n","            if final_selected[feature]==1: # append only those feature that are 1 i.e. selected\n","                Colindices.append(feature) # get the indices of features selected\n","        \n","        self.output_message += \"Feature indices in final selection : %s\"%Colindices + \"\\n\"\n","        # append all rows i.e. observations for the feature selected\n","        inputs = dataTran[:,Colindices]  # type = numpy array\n","        \n","        #######################################################################\n","        ###################### Choose the classifier ##########################\n","        print (\"training classifiers\")\n","        \n","        SV_Tree = 0.0001 # assign insignificantly low value so that goal value is not zero\n","        Dist_Metric = 0.0001\n","        if classifier_type <1.0: #linear svm\n","            clf = OneVsRestClassifier(svm.SVC(C=C_val, kernel='linear', random_state=SEED))\n","        elif classifier_type >=1.0 and classifier_type <2.0: #RBF svm\n","            #for multi-class use OneVsRestClassifier as SVC does ovo and use descion function \n","            #https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html #sklearn.svm.SVC\n","            # if class_weight=\"balanced\" is used then use oversampling; \n","            #default gamma is being used; to optimize via GA use Gamma_val\n","            clf = OneVsRestClassifier(svm.SVC(C=C_val, gamma= 'scale', kernel='rbf',random_state=SEED)) #, random_state=SEED\n","            #for binary SVC can be used\n","            #clf = svm.SVC(C=C_val, gamma= 'scale', kernel='rbf', random_state=SEED)\n","            \n","        else: # random forest, n_jobs=-1 for parallelsim; \n","            #although RF inherently multi-class, OneVsRestClassifier may b better for convex loss\n","            # additional parameter required for tree depth or use variable Gamma_val if unused for SVM with defualt usage \n","            clf = OneVsRestClassifier(RandomForestClassifier(n_estimators = int(TreeCount), oob_score=True, random_state=SEED,n_jobs=-1))\n","            SV_Tree = float((self.max_trees - int(TreeCount))/self.max_trees) # don't round so that insignificantly low values are kept\n","            \n","        #### flag to check if this is for final model saving or evaluation of solutions during GA\n","        if finalFlag == True: # if final model being trained then no CV required\n","            inputs, outputs = self.sampling_function(outputs,inputs,multilabel) # perform oversampling of imbalanced class\n","            \n","            clf.fit(inputs,outputs) # train on all data\n","            print (\"final model fitting done\")\n","            return clf, inputs, outputs\n","        \n","        #######################################################################################\n","        ###################### Run CV and get the performance scores ##########################\n","        skf = StratifiedKFold(n_splits=self.getFolds()) #  stratified to preserve the percentage of samples for each class.\n","        sv_list=[]\n","        dist_list = []\n","        \n","        # initailise performance metrics to compute agreegate at fold level\n","        accuracy_fold = [] # this measures accuracy as all labels predicted correct \n","        \n","        y_true = [] # to append the held-out cases for final ROC computation\n","        y_pred_Class1prob = [] # to append the held-out cases for final ROC computation\n","        y_pred_Class=[]\n","        npv_split=[]\n","        specificity_fold = []\n","        precision_fold = []\n","        recall_fold = []\n","        roc_fold = []\n","        tn_all=0\n","        fp_all=0\n","        fn_all=0\n","        tp_all=0\n","        \n","        cv_scores = [] # initialise list of performance scores averaged over k-folds\n","        \n","        # CV Loop coded only for binary classification problem\n","        for train_index, test_index in skf.split(inputs, outputs): # takes 20% as test data\n","            X_train, X_test = inputs[train_index,:], inputs[test_index,:] # inputs is ndarray so use np.ndarray slicing\n","            y_train, y_test = np.array(outputs)[train_index], np.array(outputs)[test_index]\n","            \n","            X_train, y_train = self.sampling_function(y_train,X_train,multilabel) # MTD oversampling of all classes\n","            \n","            clf.fit(X_train, y_train) # fit on train data\n","            \n","            # 'OneVsRestClassifier' object has no attribute 'n_support_'\n","            if classifier_type <2.0: sv_list.append(0.0001)\n","            \n","            y_pred = clf.predict(X_test)\n","            if classifier_type <2.0: # svm\n","                y_score = clf.decision_function(X_test)\n","                y_pred_Class1prob.append(clf.decision_function(X_test)) # get deciison function\n","            else: # RF\n","                y_score = clf.predict_proba(X_test) # for RF use probaility;\n","                y_pred_Class1prob.append(y_score[:,1])\n","                \n","            y_true.append(y_test)\n","            y_pred_Class.append(y_pred)\n","            \n","            accuracy_fold.append(accuracy_score(y_test, y_pred)) # to get distance or accuracy based on hamming loss\n","            specificity_fold.append(recall_score(y_test, y_pred, pos_label=0, average='binary')) # true negative rate or specificity\n","            precision_fold.append(precision_score(y_test, y_pred, average='binary'))\n","            recall_fold.append(recall_score(y_test, y_pred, average='binary'))\n","            if classifier_type <2.0: # svm\n","                roc_fold.append(roc_auc_score(y_test,y_score)) # y_score[:,1]\n","            else: # RF\n","                roc_fold.append(roc_auc_score(y_test,y_score[:,1])) # y_score[:,1]\n","            \n","            #npv calculate\n","            tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel() \n","            tn_all = tn_all+tn\n","            fp_all = fp_all+fp\n","            fn_all = fn_all+fn\n","            tp_all = tp_all+tp\n","            \n","            if (tn>=0 and fn>0) or (tn>0 and fn>=0):\n","                npv= tn/(tn+fn)\n","            else: npv = 0.0\n","            npv_split.append(npv)\n","            \n","            err = 0.0\n","            for err_index in range(len(X_test)):\n","                if y_pred[err_index] != y_test[err_index]: \n","                    err = err + 1.0\n","            # store the k-fold error and accuracy of the ensemble model\n","            dist_list.append(1.0 - float(err/len(X_test)))\n","        \n","        if classifier_type <2.0: SV_Tree = np.mean(np.array(sv_list))\n","        Dist_Metric = np.mean(np.array(dist_list))\n","        \n","        # take average of metrics at fold level and append in sequence \n","        # [Accuracy,NPV, Specificity, PPV,Recall,Roc_auc]\n","        cv_scores.append(np.mean(np.array(accuracy_fold)))\n","        cv_scores.append(np.mean(np.array(npv_split)))\n","        cv_scores.append(np.mean(np.array(specificity_fold)))\n","        cv_scores.append(np.mean(np.array(precision_fold)))\n","        cv_scores.append(np.mean(np.array(recall_fold)))\n","        cv_scores.append(np.mean(np.array(roc_fold)))\n","        \n","        return cv_scores, SV_Tree,Dist_Metric\n","    \n","    '''#This function is used to calculate the optimization goals and is called by optimization()\n","    # Inputs are single individual solution, pVal for features calculated frm wilcoxon & transpose data for Kbest & MI feature selection\n","    # goal_values inialised to zero for single sol and calculated in this function'''\n","    \n","    def evaluate_individuals(self,ind, dataset): \n","\n","        Questions_proportion = 0\n","        Single_blood_proportion = 0 \n","        Risk_scores_proportion = 0 \n","        predicted_output = [0 for _ in range(len(dataset))]\n","\n","        actualValuesStatus = []\n","        actualValuesFollowup = []\n","        \n","        for i in range(len(dataset)):\n","            patient = dataset.iloc[[i]]\n","            actualValuesStatus.append(dataset['CVD_status'].iloc[i])\n","            actualValuesFollowup.append(dataset['CVD_followup'].iloc[i])\n","\n","            predicted = evaluateBranch(\n","                modelsPickleList[0],\n","                cleaned_model_to_csvs[\"Sociodemographics\"],\n","                patient,\n","                modelsPickleList[2],\n","                cleaned_model_to_csvs[\"Sociodemographics_oneDrop_qrisk\"],\n","                ind[0],\n","                ind[1]\n","            )\n","            \n","            if isinstance(predicted, float) and predicted < ind[0]:\n","                predicted_output[i] = predicted\n","                Questions_proportion += 97\n","\n","\n","            elif isinstance(predicted, float) and predicted > ind[1]:\n","                predicted_output[i] = predicted\n","                Questions_proportion += 97\n","                Single_blood_proportion += 185\n","                Risk_scores_proportion += 44\n","\n","            else:\n","                predicted = evaluateBranch(\n","                    modelsPickleList[3],\n","                    cleaned_model_to_csvs[\"Sociodemographics_Health and medical history\"],\n","                    patient,\n","                    modelsPickleList[5],\n","                    cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_oneDrop_qrisk\"],\n","                    ind[2],\n","                    ind[3]\n","                )\n","\n","                if isinstance(predicted, float) and predicted < ind[2]:\n","                    predicted_output[i] = predicted\n","                    Questions_proportion += 242\n","\n","                elif isinstance(predicted, float) and predicted > ind[3]:\n","                    predicted_output[i] = predicted\n","                    Questions_proportion += 242\n","                    Single_blood_proportion += 185\n","                    Risk_scores_proportion += 44\n","\n","                else:\n","                    predicted = evaluateBranch(\n","                        modelsPickleList[6],\n","                        cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors\"],\n","                        patient,\n","                        modelsPickleList[8],\n","                        cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_oneDrop_qrisk\"],\n","                        ind[4],\n","                        ind[5]\n","                    )\n","\n","                    if isinstance(predicted, float) and predicted < ind[4]:\n","                        predicted_output[i] = predicted\n","                        Questions_proportion += 312\n","\n","\n","                    elif isinstance(predicted, float) and predicted > ind[5]:\n","                        predicted_output[i] = predicted\n","                        Questions_proportion += 312\n","                        Single_blood_proportion += 185\n","                        Risk_scores_proportion += 44\n","\n","                    else:\n","                        predicted = evaluateBranch(\n","                            modelsPickleList[9],\n","                            cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors\"],\n","                            patient,\n","                            modelsPickleList[11],\n","                            cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_oneDrop_qrisk\"],\n","                            ind[6],\n","                            ind[7]\n","                        )\n","\n","                        if isinstance(predicted, float) and predicted < ind[6]:\n","                            predicted_output[i] = predicted\n","                            Questions_proportion += 319\n","\n","\n","                        elif isinstance(predicted, float) and predicted > ind[7]:\n","                            predicted_output[i] = predicted\n","                            Questions_proportion += 319\n","                            Single_blood_proportion += 185\n","                            Risk_scores_proportion += 44\n","\n","                        else:\n","                            predicted = evaluateBranch(\n","                                modelsPickleList[12],\n","                                cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history\"],\n","                                patient,\n","                                modelsPickleList[14],\n","                                cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_oneDrop_qrisk\"],\n","                                ind[8],\n","                                ind[9]\n","                            )\n","\n","                            if isinstance(predicted, float) and predicted < ind[8]:\n","                                predicted_output[i] = predicted\n","                                Questions_proportion += 376\n","\n","\n","                            elif isinstance(predicted, float) and predicted > ind[9]:\n","                                predicted_output[i] = predicted\n","                                Questions_proportion += 376\n","                                Single_blood_proportion += 185\n","                                Risk_scores_proportion += 44\n","\n","                            else:\n","                                predicted = evaluateBranch(\n","                                    modelsPickleList[15],\n","                                    cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment\"],\n","                                    patient,\n","                                    modelsPickleList[17],\n","                                    cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment_oneDrop_qrisk\"],\n","                                    ind[10],\n","                                    ind[11]\n","                                )\n","\n","                                if isinstance(predicted, float) and predicted < ind[10]:\n","                                    predicted_output[i] = predicted\n","                                    Questions_proportion += 703\n","\n","                                elif isinstance(predicted, float) and predicted > ind[11]:\n","                                    predicted_output[i] = predicted\n","                                    Questions_proportion += 703\n","                                    Single_blood_proportion += 185\n","                                    Risk_scores_proportion += 44\n","\n","                                #\"Make it into else to make sure catch everything\"\n","                                #if not isinstance(predicted_output[i], float) or predicted_output[i] == 0:\n","                                else:\n","                                    predicted = evaluateBranch(\n","                                        modelsPickleList[18],\n","                                        cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment_Psychosocial factors\"],\n","                                        patient,\n","                                        modelsPickleList[19],\n","                                        cleaned_model_to_csvs[\"Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop_qrisk\"],\n","                                        ind[12],\n","                                        ind[13]\n","                                    )\n","\n","                                    if isinstance(predicted, float) and predicted < ind[12]:\n","                                        predicted_output[i] = predicted\n","                                        Questions_proportion += 756\n","\n","                                    elif isinstance(predicted, float) and predicted > ind[13]:\n","                                        predicted_output[i] = predicted\n","                                        Questions_proportion += 756\n","                                        Single_blood_proportion += 185\n","                                        Risk_scores_proportion += 44\n","\n","                                    else:\n","                                        predicted = evaluateBranch(\n","                                        modelsPickleList[19],\n","                                        cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment_Psychosocial factors_oneDrop\"],\n","                                        patient,\n","                                        modelsPickleList[20],\n","                                        cleaned_model_to_csvs[\"Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop_qrisk\"],\n","                                        ind[14],\n","                                        ind[15]\n","                                        )\n","                                        if isinstance(predicted, float) and predicted < ind[14]:\n","                                            predicted_output[i] = predicted\n","                                            Questions_proportion += 756\n","                                            Single_blood_proportion += 185\n","                                            \n","                                        else:\n","                                            predicted = predictIndividual(modelsPickleList[20], cleaned_model_to_csvs[\"Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop_qrisk\"], patient)[0]\n","                                            predicted_output[i] = predicted\n","                            \n","                                            Questions_proportion += 756\n","                                            Single_blood_proportion += 185\n","                                            Risk_scores_proportion += 44\n","            \n","            # Normalize proportions\n","        Questions_proportion /= (756 * len(dataset))\n","        Single_blood_proportion /= (185 * len(dataset))\n","        Risk_scores_proportion /= (44 * len(dataset))\n","\n","        #Risk score as a median is the average of the lowest risk score (-3) and the highest (+9) so 3\n","        y_pred_binary = np.array(predicted_output) >= ind[15]\n","        y_true = np.array(actualValuesStatus)\n","\n","        TP = np.sum((y_pred_binary == 1) & (y_true == 1))\n","        TN = np.sum((y_pred_binary == 0) & (y_true == 0))\n","        FP = np.sum((y_pred_binary == 1) & (y_true == 0))\n","        FN = np.sum((y_pred_binary == 0) & (y_true == 1))\n","\n","        True_positive_rate = TP / (TP + FN) if (TP + FN) else 0\n","        True_negative_rate = TN / (TN + FP) if (TN + FP) else 0\n","\n","        try:\n","            AUC = roc_auc_score(y_true, predicted_output)\n","        except ValueError:\n","            AUC = 0  \n","\n","        try:\n","            from sksurv.metrics import concordance_index_censored\n","            y_surv = np.array([(bool(e), t) for e, t in zip(y_true, actualValuesFollowup)],\n","                              dtype=[('event', bool), ('time', float)])\n","            Cindex = concordance_index_censored(y_surv['event'], y_surv['time'], predicted_output)[0]\n","        except:\n","            Cindex = 0\n","\n","        print(f\"  1 - Questions proportion: {1 - Questions_proportion:.3f}\")\n","        print(f\"  1 - Single blood proportion: {1 - Single_blood_proportion:.3f}\")\n","        print(f\"  1 - Risk scores proportion: {1 - Risk_scores_proportion:.3f}\")\n","        print(f\"  True Positive Rate (TPR): {True_positive_rate:.3f}\")\n","        print(f\"  True Negative Rate (TNR): {True_negative_rate:.3f}\")\n","        print(f\"  AUC: {AUC:.3f}\")\n","        print(f\"  C-index: {Cindex:.3f}\")\n","\n","\n","        goal_values = [0]*8\n","        goal_values[0] = 1 - Questions_proportion\n","        goal_values[1] = 1 - Single_blood_proportion\n","        goal_values[2] = 1 - Risk_scores_proportion\n","        goal_values[3] = True_positive_rate\n","        goal_values[4] = True_negative_rate\n","        goal_values[5] = AUC\n","        goal_values[6] = Cindex\n","        goal_values[7] = (\n","            self.goal_significance[0]*goal_values[0] +\n","            self.goal_significance[1]*goal_values[1] +\n","            self.goal_significance[2]*goal_values[2] +\n","            self.goal_significance[3]*goal_values[3] +\n","            self.goal_significance[4]*goal_values[4] +\n","            self.goal_significance[5]*goal_values[5] +\n","            self.goal_significance[6]*goal_values[6]\n","        ) / sum(self.goal_significance)\n","\n","        return goal_values\n","        \n","\n","    #This function is used by non_Dominated_Sol() to compare \n","    # if a sol i.e. goal_value dominates another sol i.e. goal_value\n","    def dominate(self, solution1, solution2):\n","        check=2 # initialise solution comparison check to 2 i.e. equal\n","        ffs=len(solution1)\n","        dominate1=1\n","        equal1=1\n","        f=0\n","        while f<ffs and dominate1==1:\n","            if solution1[f]>solution2[f]:\n","                equal1=0\n","            elif solution1[f]==solution2[f]:\n","                do_nothing=1\n","            else:\n","                dominate1=0\n","            f=f+1\n","        \n","        if dominate1==1 and equal1==0:\n","            check=1\n","        elif dominate1==1 and equal1==1:\n","            check=2\n","        else:\n","            dominate2=1\n","            equal2=1\n","            f=0\n","            while f<ffs and dominate2==1:\n","                if solution2[f]>solution1[f]:\n","                    equal2=0\n","                elif solution2[f]==solution1[f]:\n","                    do_nothing=1\n","                else:\n","                    dominate2=0\n","                f=f+1\n","            \n","            if dominate2==1 and equal2==0:\n","                check=3\n","        \n","        return check\n","\n","    # function to bucket solutions of a given generation into diff. pareto fronts  \n","    # This function is called by optimization()\n","    # adapted from script written for Missense SNPs ensemble_multiclassifier.py by Konstantinos Theofilatos \n","    # A dominates B iff it is better than or equal to B on all dimensions and strictly better on at least one dimension\n","    #def non_Dominated_Sol(self,population,goal_values,individuals):\n","    # functioon inherited from ensemble_multiclassifier.py for Missense\n","    def getParetoFront(self,population,goal_values,stopIndx):\n","        assigned=0\n","        fronts=[0]*population # initialise pareto front for population size\n","        front=1\n","        #  create copy of goal_values so that original goal are not changed. Exclude weighted average\n","        eval_temp=copy.deepcopy(np.array(goal_values)) # change asarray to array to create copy\n","        eval_temp = eval_temp.T # take transpose so that goals are in rows and observations in columns\n","        \n","        #  reduce the metrics to PPV, sensitivity and ROC for Pareto Front. Align with choice of goal_significance weightage\n","        pareto_Index=[0,1,2,3,4,5,6] # no feature restriction, PPV, ROC, sensitivity\n","        \n","        eval_temp = [eval_temp[i] for i in pareto_Index]\n","        print (\"Parento front computation\")\n","        #print (eval_temp)\n","        \n","        number_of_solutions = copy.deepcopy(population) # deep copy of length of population\n","        \n","        #ref: Mishra, K. K., & Harit, S. (2010). A fast algorithm for finding the non-dominated set in multi objective optimization. International Journal of Computer Applications, 1(25), 35-39. \n","        while assigned < population: # iterate untill non-dominated solutions assigned is less than population size\n","            non_dominated_solutions = [0]*number_of_solutions\n","            index=[i for i in range(len(eval_temp[0]))] # generate list of index based on count of goals\n","            self.Pareto_Wheel_Output += \"Index for Ordered List %s: \"%(index)  + \"\\n\"\n","            \n","            eval_temp_index=list(zip(eval_temp[0],index)) # create list of tuples with goal val of sol. & index \n","            # sort the goal values in decrease order i.e. highest comes 1st\n","            self.Pareto_Wheel_Output += \"Tuple of goal and list index %s: \"%(eval_temp_index)  + \"\\n\"\n","            ordered_list= sorted(range(len(eval_temp_index)), key=lambda k: eval_temp_index[k],reverse=True) \n","            \n","            non_dominated_solutions[0]=ordered_list[0]\t# assign the 1st goal values as 1st non-dominanted sol.\t\t\n","            number_of_non_dominated_solutions=1\t\t# initiatlise the non-dominated sol count\n","            \n","            for i in range(1,number_of_solutions): # iterate from 1 to pop size as 1st i.e. 0 index is already selected\n","                n=0\n","                condition=0\n","                condition2=1\n","                while n<number_of_non_dominated_solutions and condition==0: # compare ordered list of solutions pairs \n","                    solution1=[0]*(len(eval_temp)) # excludes weighted average goal & clasification model\n","                    solution2=[0]*(len(eval_temp)) # excludes weighted average goal & classification model\n","                    \n","                    for j in range(len(eval_temp)): # iterate on no of goals; compare solutions by goals\n","                        solution1[j]=eval_temp[j][ordered_list[i]]\n","                        solution2[j]=eval_temp[j][ordered_list[n]]\t\t\t\t\n","                    \n","                    check=self.dominate(solution1,solution2) # compare the goal values in descending order\n","                    if check==3:\n","                        condition=1\n","                        condition2=0\n","                    elif check==1:\n","                        if number_of_non_dominated_solutions==1:\n","                            condition=1\n","                            non_dominated_solutions[0]=ordered_list[i]\n","                        else:\n","                            number_of_non_dominated_solutions=number_of_non_dominated_solutions-1\t\n","                            del non_dominated_solutions[n]\n","                            \n","                    n=n+1\n","                    \n","                if condition2==1:\n","                    non_dominated_solutions[number_of_non_dominated_solutions]=ordered_list[i]\n","                    number_of_non_dominated_solutions=number_of_non_dominated_solutions+1\n","            \n","            sorted_non_dominated_solutions=sorted(non_dominated_solutions, reverse=True) # index sorting of non-dominated sol\n","            \n","            for i in range(number_of_non_dominated_solutions):\n","                assigned=assigned+1\n","                fronts[sorted_non_dominated_solutions[i]]=front\n","                for j in range(len(eval_temp)): # sets the chosen goals to zero\n","                    eval_temp[j][sorted_non_dominated_solutions[i]]=0\t\t\n","            front=front+1\t\t\t\t\t\n","        \n","        self.Pareto_Wheel_Output += \"Calculated Pareto Frontiers %s \"%(fronts)  + \"\\n\\n\"\n","        print(\"Pareto Front Calculation Completed\")\t\t\n","        \n","        ### check if front = 0 for error capture\n","        if min(fronts) == 0:\n","            f = open('FrontERROR.txt','w') # redirect all print statements to file output\n","            print ('Pareto front zero detected',file=f)\n","            # assign front 0 to be max(front)+1 so that it is worst sol\n","            newFront = max(fronts)+1\n","            for i,f in enumerate(fronts):\n","                if f==0: \n","                    fronts[i]=newFront\n","            \n","        return fronts # returns list of pareto fronts for the observations i.e. population\n","\t \n","    #Tune fitness values by locating and using solution niches\n","    # adapted from script written for Missense SNPs ensemble_multiclassifier.py by Konstantinos Theofilatos \n","    # uses distance for solutions within a given pareto front to see if they are similar or diff\n","    # and assigns a new fitness score\n","    def tuneFitnessValue(self,fronts,sigma_share,evaluation_values,rep,population,individuals):\n","        ####### evaluation values has last values as weighted average\n","        min_values = self.min_values\n","        max_values = self.max_values\n","        \n","        print(\"Starting Fitness Tuning As Per Pareto Frontiers\")\n","        \n","        for i in range(1,max(fronts)+1): # start from 1st pareto front; index of fronts is starting with 1, increment by 1\n","            self.Pareto_Wheel_Output += \"Fitness Tuning As Per Pareto Front: %s \"%(i)  + \"\\n\"\n","            #CHANGE 2\n","            ind = [y for y, x in enumerate(fronts) if x == i]\n","\n","\n","            self.Pareto_Wheel_Output += \"Individuals in pareto front %s \"%(ind)  + \"\\n\"\n","            \n","            #Calculate max values per goal per pareto frontier\t\t\t\n","            max_significances=[0]*(len(evaluation_values)-1) # exclude weighted average\n","            for j in range(len(ind)):\n","                for goal in range(len(evaluation_values)-1):\n","                    if evaluation_values[goal][ind[j]]>=max_significances[goal]:\n","                        max_significances[goal]=evaluation_values[goal][ind[j]]\t\t\t\n","            \n","            self.Pareto_Wheel_Output += \"Max. Significance %s \"%(max_significances)  + \"\\n\"\n","            \n","            # compute distance based similarity for the solutions within a given front\n","            for j in range(len(ind)): # iterate for indivdiuals within a given front\n","                m=0\n","                for k in range(len(ind)):\n","                    d=0\n","                    for gene in range(len(individuals[0])): \n","                        d=d+((individuals[ind[j]][gene]-individuals[ind[k]][gene])/float(max_values[gene]-min_values[gene]))**2\n","                        \n","                    d=math.sqrt(d/(len(individuals[0])))\n","                    \n","                    if d<=sigma_share:\n","                        m=m+(1-((d/float(sigma_share))**2))\n","                    if m==0:\n","                        m=1\n","                self.Pareto_Wheel_Output += \"Value of m  %s \"%str(m)  + \"\\n\"\n","                for goal in range(len(evaluation_values)-1):\n","                    evaluation_values[goal][ind[j]]=float(max_significances[goal])/m\n","        \n","        ## recalculate weighted sum of goals using goal_significance\n","        # goals are in rows and samples in columns\n","        for i in range(len(evaluation_values[0])): # compute weighted sum for each sample\n","            evaluation_values[-1][i]=0 # reset weighted average i.e. last row to zero\n","            for j in range(len(evaluation_values)-1): # exclude last row i.e. weighted sum in iteration\n","                evaluation_values[-1][i]=evaluation_values[-1][i]+evaluation_values[j][i]*self.goal_significance[j]\n","            evaluation_values[-1][i]=evaluation_values[-1][i]/float(len(evaluation_values)-1)\n","        \n","        ## Update Sum of weighted avg goal and avg of weighhted avg goal for the generation\n","        # to be used for visualisation\n","        sum_ranked=0\n","        for i in range(population):\n","            sum_ranked=sum_ranked+evaluation_values[-1][i]\n","        \n","        self.sum_ranked_eval_per_generation[rep]=sum_ranked\n","        self.average_ranked_eval_per_generation[rep]=sum_ranked/population\n","        \n","        print(\"Fitness Tuning As Per Pareto Frontiers Completed\")\n","        \n","        return evaluation_values\n","    \n","    #roulett wheel based selection of ranked indivdiuals\n","    # adapted from script written for Missense SNPs ensemble_multiclassifier.py by Konstantinos Theofilatos \n","    # evaluation values are of shape goals X samples\n","    def roulett_Wheel(self, evaluation_values_tuned,population,individuals,rep,bestSolIndex):\n","        \n","        min_values = self.min_values\n","        selected_individuals=[[0 for x in range(len(min_values))] for x in range(population)]\n","        indv_wheel = copy.deepcopy(individuals) # inclulded this to avoid any link between individuals and selected indv.\n","        # calculate the cumulative proportions i.e. selection probability based on weighted avg goal\n","        sum_prop=[0]*(population+1)\n","        for i in range(1, population+1): \n","            sum_prop[i]=sum_prop[i-1]+evaluation_values_tuned[-1][i-1]/float(self.sum_ranked_eval_per_generation[rep])\n","        \n","        self.Pareto_Wheel_Output += \"Cummuative proportions based on weighted avg post tuning of Fitness Value:  %s \"%(sum_prop)  + \"\\n\"\n","        \n","        for i in range(1, population): # get proportions from index 1 thru pop size as 0 is for best ind.\n","            random_number=random.uniform(0,1) # generate probability randomly\n","            for j in range(0,population): # select those indv. for which random no is within cummulative prop \n","                if random_number>=sum_prop[j] and random_number<sum_prop[j+1]:\n","                    selected_individuals[i]=copy.deepcopy(indv_wheel[j]) # assign via deepcopy and not original\n","        \n","        ## assign best sol based on max of weighted sum of goals to index 0\n","        selected_individuals[0]= copy.deepcopy(indv_wheel[bestSolIndex]) # assign via deepcopy to avoid index 0 editing\n","        \n","        print(\"Roulette Wheel Selection Completed\")\n","        \n","        return selected_individuals\n","\n","    # function to apply two-point cross over\n","    # adapted from script written for Missense SNPs ensemble_multiclassifier.py by Konstantinos Theofilatos \n","    def GACross_Over(self,population,cross_indv):\n","        \n","\n","        \n","        temp_individuals=[[0 for x in range(len(self.min_values))] for x in range(population)] \n","        \n","        cross_indv_copy = copy.deepcopy(cross_indv) # take deep copy to avoid index 0 editing\n","        features = copy.deepcopy(len(cross_indv_copy[0])) # take deep copy to avoid index 0 editing\n","        \n","        best_sol_before = sum(cross_indv[0]) # index 0 holds best sol\n","        \n","        for i in range(1,population-1,2): # 1st is best sol so preseve it\n","            random_number=random.uniform(0,1) # generate prob randomly \n","            if random_number < self.twoCross_prob: \n","                cross_point1=0\n","                cross_point2=0\n","                while cross_point1==cross_point2:\n","                    cross_point1 = math.ceil((features-4)*random.uniform(0,1)) # minus 4 to leave last positions\n","                    if cross_point1 < math.floor((2*features-1)/3): \n","                        width = math.ceil(random.uniform(0,1)*(math.floor(features-1)/3 -2))\n","                        cross_point2 = cross_point1 + width\n","                    else:\n","                        width=math.ceil(random.uniform(0,1))*(math.floor(features/3 -1)-2-(cross_point1-math.floor(2*features/3)))\n","                        cross_point2=cross_point1+width\n","                if cross_point1>cross_point2:\n","                    temp_cross_point=cross_point1\n","                    cross_point1=cross_point2\n","                    cross_point2=temp_cross_point\n","                width=int(width)\n","                cross_point1=int(cross_point1)\n","                cross_point2=int(cross_point2)\n","                for j in range(cross_point1,cross_point2+1):\n","                    temp_individuals[i][j]=copy.deepcopy(cross_indv_copy[i+1][j])\n","                for j in range(cross_point1,cross_point2+1):\n","                    cross_indv_copy[i+1][j]=copy.deepcopy(cross_indv_copy[i][j])\n","                for j in range(cross_point1,cross_point2+1):\n","                    cross_indv_copy[i][j]=copy.deepcopy(temp_individuals[i][j])\n","            \n","            elif random_number>= self.twoCross_prob and random_number < (self.twoCross_prob + self.arithCross_prob): \n","                # arithmetic cross over\n","                alpha=random.uniform(0,1)\t\t\t\n","                for j in range(0,features): \n","                    temp_individuals[i][j]=copy.deepcopy(alpha*cross_indv_copy[i][j]+(1-alpha)*cross_indv_copy[i+1][j])\n","                    temp_individuals[i+1][j]=copy.deepcopy((1-alpha)*cross_indv_copy[i][j]+(alpha)*cross_indv_copy[i+1][j])\n","                for k in range(0,features):\t\t\t\t\n","                    cross_indv_copy[i][k]=copy.deepcopy(temp_individuals[i][k])\n","                    cross_indv_copy[i+1][k]=copy.deepcopy(temp_individuals[i+1][k])\t\t\n","        \n","        best_sol_after = sum(cross_indv_copy[0]) # index 0 holds best sol; take after sum from copy version \n","        \n","        ### check if best solution is preserved\n","        if (best_sol_before - best_sol_after) != 0.0:\n","            fcross = open('CrossOverERROR.txt','w') # redirect all print statements to file output\n","            self.output_message += \"Cross over of best sol detected \" + \"\\n\"\n","            print ('Cross over for best solution detected. Sol sum delta %s'%(best_sol_before - best_sol_after),file=fcross)\n","        \n","        print(\"Crossover completed\")\n","        return copy.deepcopy(cross_indv_copy)\n","    \n","    # function to apply adaptive mutation; Use deep copy for assignment to avoid index 0 editing\n","    # adapted from script written for Missense SNPs ensemble_multiclassifier.py by Konstantinos Theofilatos \n","    # script edited for adaptive mutation rate\n","    def adaptiveMutation(self,population,mutate_indv, rep):\n","        print(\"Doing Adaptive Mutation\")\n","        min_values = self.min_values\n","        max_values = self.max_values\n","        \n","        best_sol_before = sum(mutate_indv[0]) # index 0 holds best sol\n","        mutate_indv_copy = copy.deepcopy(mutate_indv) # take deep copy so that index 0 is not chnaged\n","        \n","        # get the mutation probability and gaussian variance proportion based on average similarity \n","        # ref: Rapakoulia, T., Theofilatos, K., Kleftogiannis, D., Likothanasis, S., Tsakalidis, A., & Mavroudi, S. (2014). \n","        # EnsembleGASVR: a novel ensemble method for classifying missense single nucleotide polymorphisms. Bioinformatics, 30(16), 2324-2333. \n","        \n","        # compute average similarity between the members of the population and its best member i.e. at index 0\n","        a = np.array(mutate_indv_copy[:0:-1]) # take all elements from bottom to 1, exclude 0\n","        b = np.array([mutate_indv_copy[0]]) # best solution\n","        CDist= distance.cdist(a, b, 'chebyshev')\n","        AvgSimilarity = np.mean(CDist)/ (np.max(CDist) - np.min(CDist))\n","        self.Pareto_Wheel_Output += \"Mutation Cal. Avg. Similarity %s \"%(AvgSimilarity)  + \"\\n\"\n","        \n","        numerator = float(self.mutation_prob[1]) - (1/self.population) # 0.2-(1/Pop size)\n","        delta = float(rep)*numerator/float(self.generations)\n","        if round(AvgSimilarity,2) >= round(self.avgSimThr,2): # high similaity, so increase the diversity with higher mutation\n","            mutationRate = float(self.mutation_prob[1]) + delta # 0.2+ delta\n","            gaussVarProp = float(self.gaussMut_varPro[1]) + delta # 0.5+ delta\n","        else:\n","            mutationRate = float(self.mutation_prob[1]) - delta # 0.2 - delta\n","            gaussVarProp = float(self.gaussMut_varPro[1]) - delta # 0.5 - delta\n","            if mutationRate < float(self.mutation_prob[0]): # reset if lesser than lower bound\n","                mutationRate = float(self.mutation_prob[0])\n","            if gaussVarProp < float(self.gaussMut_varPro[0]): # reset if lesser than lower bound\n","                gaussVarProp = float(self.gaussMut_varPro[0])\n","        \n","        self.Pareto_Wheel_Output += \"Mutation Rate chosen %s \"%(mutationRate)  + \"\\n\"\n","        self.Pareto_Wheel_Output += \"Gaussian Variance Proportion chosen %s \"%(gaussVarProp)  + \"\\n\"\n","        \n","        feature = copy.deepcopy(len(mutate_indv_copy[0]))\n","        #apply mutation operator\n","        for i in range(1,population): # preserve the 1st sol as its best\n","            for j in range(0,feature):\n","                random_number=random.uniform(0,1)\n","                if random_number<mutationRate:\n","                    # Gaussian distribution. mean zero and standard deviation = (max-min)*variance proportion\n","                    mutate_indv_copy[i][j]=copy.deepcopy(mutate_indv_copy[i][j]+random.gauss(0,gaussVarProp*(max_values[j]-min_values[j])))\n","            \n","            #Correct values out of boundaries\n","            for j in range(0, len(min_values)):\n","                if mutate_indv_copy[i][j] < min_values[j]:\n","                    mutate_indv_copy[i][j] = copy.deepcopy(min_values[j])\n","                if mutate_indv_copy[i][j] > max_values[j]:\n","                    mutate_indv_copy[i][j] = copy.deepcopy(max_values[j])\n","        \n","        best_sol_after = sum(mutate_indv_copy[0]) # index 0 holds best sol; take sum of index 0 from copy to check if any change\n","        \n","        ### check if best solution is preserved\n","        if (best_sol_before - best_sol_after) != 0.0:\n","            fmutate = open('MutationERROR.txt','w') # redirect all print statements to file output\n","            self.output_message += \"Mutation of best sol in Gen %s\"%rep + \"\\n\"\n","            print ('Mutation of best solution detected. Sol sum delta %s'%(best_sol_before - best_sol_after),file=fmutate)\n","        \n","        return copy.deepcopy(mutate_indv_copy)\n","    \n","    ########################################################\n","    def labelPowerset (self):\n","        # take transpose of list of list i.e. the class labels outcome Y of shape [n_labels, n_sample]\n","        yNew = list(map(list, zip(*self.class_Labels))) # yNew shape [n_samples, n_labels]\n","        # convert to list of strings\n","        yNew = [[str(int(j)) for j in i] for i in yNew] # change int to str\n","        self.powerSetClass = list(map(''.join, yNew)) # list of strings with powerset of class labels given the dataset\n","    \n","    \n","\n","    \n","    '''#This function pre-computes the wilconxon ranksum, JMIM and MRMR for given set of k values.\n","    # it then passes diff. populations of teh generation for evaluation to evaluate_individuals()\n","    # Pareto optimal solution search and cross-over + mutation for GA is also done here'''\n","    def optimization(self):\n","        \n","        \n","        self.output_message += \"Parameters selected Gen: %s , Pop: %s\"%(self.generations, \n","                                                                                       self.population) + \"\\n\" # capture the parameters \n","\n","        individuals = self.initialize() #Initialize Population of Individual Solutions\n","        \n","        #individuals = self.initializeBestCSV() # initialise with front1 solutions from previous run\n","        self.output_message += \"Initial population formulated. Sample indivdiuals length : %s\"%len(individuals[0]) + \", %s\"%len(individuals[4]) + \"\\n\"\n","        self.output_message += \"Sample individual : %s\"%(individuals[4]) + \"\\n\"\n","        \n","        print(\"Individuals Initiatised\")\n","        \n","        # variables for boundary condition / hold output of each generation\n","        generations = self.getGenerations()\n","                \n","        # make custom score f2 for cross_validate\n","        \n","        sigma_share=0.5/(float(len(individuals[0]))**(0.1)) # sigma for distance threshold\n","        \n","        #Apply Evolutionary Process to Optimize Individual Solutions\n","        print(\"Starting GA based optimisation\")\t\n","        premature_termination=0 \n","        front1_feat_perGen = [] # list of lists to hold front1 features per  generation\n","        front1_sol_perGen = [] # list of lists to hold front1 sol per  generation\n","        front1_index = 0 # index to write front1 sol per generartion\n","        final_selected_best = [] # list of lists to hold best sol features per generation\n","        best_index = 0 # index to write best sol per generartion\n","        average_per_generation = []\n","        for rep in range(generations): # iterate on generations\n","            print(\"\\nGeneration:\"+str(rep+1))\n","            # initialise goal values to zero, list of list with length = population size; calcualted by evaluate_individuals()\n","            # position 0 - Questions_proportion\n","            # position 1 - Single_blood_proportion\n","            # position 2 - Risk_scores_proportion\n","            # position 3 - True_positive_rate = Tp/(Tp+Fn)\n","            # position 4 - False_negative_rate = Tn /(Tn+Fp)\n","            # position 5 - AUC\n","            # position 6 - Cindex\n","            # position 7 - Weighted_Sum\n","\n","                \n","        \n","            \n","            ############ Evaluate population of solutions. \n","            self.output_message += \"\\n\"\n","            self.output_message += \"Generation : %s\"%str(rep) + \"\\n\"\n","            post_evaluate = False\n","            # evaluate individuals of a given population in paralle\n","            results = Parallel(n_jobs=7, backend=\"multiprocessing\")(\n","            delayed(self.evaluate_individuals)(ind, dfTrain) for ind in individuals)\n","\n","            self.goal_values = np.asarray(results)          # shape: (population, 8)\n","            goal_values = self.goal_values\n","            average_performance = 0 # average performance for convergence. initialise to zero\n","            convergence = 0.0\n","            avgAll = goal_values.mean(axis=0)               # average of each goal\n","            best_performance = goal_values.max(axis=0)      # max of each goal\n","            bestSolIndex = goal_values[:, 7].argmax()\n","            best_avg_perf = goal_values[bestSolIndex][7]\n","            average_performance=avgAll[7]\n","            self.output_message += \"best sol index %s\"%bestSolIndex + \"\\n\"\n","            average_per_generation.append(average_performance)\n","            \n","            \n","            convergence = math.fabs(best_avg_perf - average_performance)\n","            print (\"convergence value %s\"%convergence)\n","            #Convergence criterion is checked in order to stop the evolution if the population is deemd us converged\n","            if convergence < self.convergeThreshold * average_performance:\n","                premature_termination=1\t\t\t\n","            \n","            #Estimate pareto fronts for the solutions of this generation based on all goals except weighted average\n","            stopIndx = (len(goal_values[0])-1) # goals to slice for pareto front computation, excludes weighted average\n","            print (\"Pareto stopIndx %s\"%stopIndx)\n","            self.Pareto_Wheel_Output += \"Generation : %s\"%str(rep+1) + \"\\n\"\n","            fronts = self.getParetoFront(len(individuals),goal_values,stopIndx)\n","            \n","            ##### save front1 solutions, features are intersect of GA individuals and filter method\n","            post_evaluate = True # when true this flag avoid classifier fitting within evaluate\n","            for_front1 = copy.deepcopy(individuals)\n","            for i,ind in enumerate(for_front1[:]): # iterate on solutions to pic sol with front = 1\n","                if fronts[i] == 1:\n","                    front1_sol_perGen.append([])\n","                    front1_sol_perGen[front1_index].append(rep) # add generation number\n","                    front1_sol_perGen[front1_index].extend(ind)\n","                    front1_feat_perGen.append([])\n","                    front1_feat_perGen[front1_index].append(rep) # add generation number\n","                    front1_feat_perGen[front1_index].extend(goal_values)\n","                    front1_index = front1_index + 1\n","            \n","            # get max per goal; and best sol\n","            self.max_eval_per_generation[rep] = list(map(max, zip(*[i for i in goal_values if i is not None])))\n","            self.output_message += \"best performance goals :\" + \"\\n\"\n","            self.output_message += str(self.max_eval_per_generation[rep])\n","            self.output_message += \"\\n\"\n","            \n","            self.max_sol_per_generation[rep] = copy.deepcopy(individuals[bestSolIndex]) # use index of bestSolIndex,there can be more than one index with max weighted average\n","            self.output_message += \"best solution :\" + \"\\n\"\n","            self.output_message += str(self.max_sol_per_generation[rep])\n","            self.output_message += \"\\n\"\n","            self.output_message += str(individuals[bestSolIndex])\n","            self.output_message += \"\\n\"\n","                        \n","            #Check if its last generation, do not apply pareto, roulette wheel, cross over and mutation\n","            if rep == (generations-1) or premature_termination==1:\n","                if premature_termination ==0:\n","                    print (\"This was last generation without convergence\")\n","                    self.output_message += \"No Convergence\" + \"\\n\"\n","                else:\n","                    print (\"Solution converged\")\n","                    self.output_message += \"Convergence at Generation : %s\"%str(rep+1) + \"\\n\"\n","                \n","                #np.savetxt(pathFeature + os.sep +\"goals_Final.csv\", goal_values, delimiter=\",\", header = self.goal_header)\n","                \n","            \n","            # compute fitness value guided by grouping of sol in each pareto fronts and recompute weighthed average\n","            # so retain weighted average goal\n","            stopIndxAvg = (len(goal_values[0])) # last column is classification model so exclude that\n","            \n","            evaluation_values = copy.deepcopy(np.asarray(goal_values)) \n","            evaluation_values = evaluation_values.T\n","            evaluation_values_tuned = self.tuneFitnessValue(fronts,sigma_share,evaluation_values,rep,len(individuals),individuals)\n","            \n","            # routleet wheel selection on tuned evaluation values by patreto front\n","            wheel_indv = copy.deepcopy(individuals) #take a deep copy so that best sol is not altered\n","            selected_individuals = self.roulett_Wheel(evaluation_values_tuned,len(individuals),wheel_indv,rep,bestSolIndex)\n","            \n","            # do cross-over \n","            cross_indv = copy.deepcopy(selected_individuals) # take deep copy so that best ind is not altered; index 0 solution\n","            cross_individuals = self.GACross_Over(len(individuals),cross_indv)\n","            # do muttaion\n","            mutate_indv = copy.deepcopy(cross_individuals) # take deep copy so that best ind is not altered; index 0 solution\n","            mutated_individuals = self.adaptiveMutation(len(individuals),mutate_indv,rep)\n","            \n","            \n","            individuals=copy.deepcopy(mutated_individuals)\n","        \n","        # last solution is the final solution \n","        try:\n","            dfFinalSol = pd.DataFrame(individuals)\n","            #SADIQ Got rid of the header\n","            dfFinalSol.to_csv(path_or_buf = pathFeature + os.sep+\"FinalSolutionsAll.csv\", index=False, header=False)\n","        except ValueError as e:\n","            dfFinalSol = pd.DataFrame(individuals)\n","            dfFinalSol.to_csv(path_or_buf = pathFeature + os.sep+\"FinalSolutionsAll.csv\", index=False, header=False)\n","        \n","        front1FinalIndv = []\n","        front1FinalGoal = []\n","        print(\"goal values\")\n","        print(goal_values)\n","        index = 0\n","        for_final_front1 = copy.deepcopy(individuals)\n","        for i,ind in enumerate(for_final_front1[:]):\n","                if fronts[i] == 1:\n","                    front1FinalIndv.append([])\n","                    front1FinalIndv[index].extend(ind)\n","                    \n","                    front1FinalGoal.append([])\n","                    front1FinalGoal[index].extend(goal_values[i])\n","\n","                    index = index + 1\n","        \n","        solHeader = ['Min_Threshold1','Max_Threshold1','Min_Threshold2','Max_Threshold2','Min_Threshold3','Max_Threshold3','Min_Threshold4','Max_Threshold4','Min_Threshold5','Max_Threshold5','Min_Threshold6','Max_Threshold6','Min_Threshold7','Max_Threshold7','Min_Threshold8','Classification_Threshold']\n","        try:\n","            df = pd.DataFrame(front1FinalIndv)\n","            df.to_csv(path_or_buf = pathFeature + os.sep+\"FinalSolFront1.csv\", index=False, header=solHeader)\n","        except ValueError as e:\n","            df = pd.DataFrame(front1FinalIndv)\n","            df.to_csv(path_or_buf = pathFeature + os.sep+\"FinalSolFront1.csv\", index=False, header=False)\n","        ##### ValueError: Writing 0 cols but got 428 aliases\n","        np.savetxt(pathFeature + os.sep +\"goals_FinalFront1.csv\", front1FinalGoal, delimiter=\",\", header = self.goal_header)\n","        df=pd.DataFrame(self.max_eval_per_generation)\n","        df.to_csv(path_or_buf = pathFeature + os.sep+ 'Max Solution Performance Per Generation.csv', index=False, header=False)\n","        df=pd.DataFrame(average_per_generation)\n","        df.to_csv(path_or_buf = pathFeature + os.sep+ 'Average Performance Per Generation.csv' , index=False, header=False)\n","        return dfTrain, front1FinalIndv\n","        \n","    ##################################################################################\n","    # function to finalise and save FINAL Model for prediction \n","    def FinaliseAndSaveModel(self,dataTran, front1FinalIndv,front1FinalFeature):\n","        print (\"Fitting and saving final model(s) for prediction\")\n","        ###### Finalizing the model : train the model on the entire training dataset available\n","        finalFlag = True # set the flag for final solution True so that CV is not done when training the classifiers\n","        skip_count = 0 # check how many models with 0 features skipped\n","        chain_dictionary = [] # initialise list for writing final front1 classifier chain. List allows same chain to be associated with more than one model\n","        # start the loop post optimisation length of optimisation parameters\n","        for ind in range(len(front1FinalIndv)): # iterate on all sol selected\n","            # train the classifier ; front1FinalIndv[ind][0] = classifier_type; front1FinalIndv[ind][1] = C_val; \n","            # front1FinalIndv[ind][2] = Gamma_val; front1FinalIndv[ind][7] = Treelevel; front1FinalIndv[ind][0] = TreeCount,ftwo_scorer = 0\n","            if len(front1FinalFeature[ind])>0: # check for zero features\n","            \n","                if len(MULTI_LABELS) == 1: # for Single label\n","                    try:\n","                        clf,X, Y = self.train_Classifier(dataTran,front1FinalFeature[ind],front1FinalIndv[ind][0],front1FinalIndv[ind][1],\n","                                             front1FinalIndv[ind][2], front1FinalIndv[ind][7], front1FinalIndv[ind][8],0,finalFlag,copy.deepcopy(MULTI_LABELS))\n","                        clf.fit(X,Y)\n","                        dump(clf, pathModel + os.sep + str(ind)+'finalSingleModel.pkl')\n","                        \n","                    except ValueError as e: # ValueError: Found array with 0 feature(s) while a minimum of 1 is required.\n","                        skip_count = skip_count + 1\n","                        continue\n","                \n","                else: # train classifier chain for multi label\n","                    try:\n","                        models_list, inputs, outputs  = self.train_Classifier_MultiLabel(dataTran,front1FinalFeature[ind],front1FinalIndv[ind][0],front1FinalIndv[ind][1],\n","                                             front1FinalIndv[ind][2], front1FinalIndv[ind][7], front1FinalIndv[ind][8],0,finalFlag,copy.deepcopy(MULTI_LABELS))\n","                        dump(models_list, pathModel + os.sep + str(ind)+'finalChainOfModels.pkl')\n","                        \n","                    except ValueError as e: # ValueError: Found array with 0 feature(s) while a minimum of 1 is required.\n","                        skip_count = skip_count + 1\n","                        continue\n","                    \n","                    # writing the classifier chain as dictionary to be used by prediction module\n","                    chain_dictionary.append([])\n","                    chain_dictionary[ind].extend(self.dict_chain[int(front1FinalIndv[ind][7])])\n","        \n","        if len(MULTI_LABELS) == 1:\n","            chain_dictionary.append(MULTI_LABELS)\n","        \n","        df = pd.DataFrame(chain_dictionary)\n","        df.to_csv(path_or_buf = pathModel + os.sep+\"classifierChain.csv\",index=False,header=False)\n","        \n","        print (\"Final model(s) saving completed and %s with 0 features skipped\"%skip_count)\n","\n","def readArguments(argv):\n","    # Get parameters based on command line input \n","    parser = argparse.ArgumentParser()\n","    #parser.add_argument('file', metavar='DATA_FILENAME', help='txt file name with input data') # default type is str\n","    parser.add_argument('-pp','--population', nargs='?', type = int, default=300, help='Population size, default is 300')\n","    parser.add_argument('-g','--generations', nargs='?', type = int, default=100, help='Count of max generations, default is 100')\n","    parser.add_argument('-f','--folds', nargs='?', type = int, default=5, help='Count of folds for cross validation, default is 5')\n","    \n","    args = parser.parse_args() # get the mandatory and default arguments\n","    \n","    return args\n","\n","################### main called when ensemble_cvd.py is run\n","def mainFunction(varDict):\n","    parser = argparse.ArgumentParser()\n","    \n","    #min_values=[0.01,0.0001,0.01,0.01,0.00005,1.0,2.0, 0.0,10.0]\n","    #max_values=[2.9,10.0,1.0,3.9,0.05,20.0,10.0, 1.0,300.0]\n","\n","    parser.add_argument('-pp','--population', nargs='?', type = int, default=300, help='Population size, default is 300')\n","    parser.add_argument('-g','--generations', nargs='?', type = int, default=100, help='Count of max generations, default is 100')\n","    #parser.add_argument('-f','--folds', nargs='?', type = int, default=5, help='Count of folds for cross validation, default is 5')\n","   \n","    categories=[\"Early life factors_Early life factors\",\"Family history_Family history\",\"Health and medical history_Breathing\",\"Health and medical history_Cancer screening\",\"Health and medical history_Chest pain\",\"Health and medical history_Claudication and peripheral artery disease\",\"Health and medical history_Eyesight\",\"Health and medical history_General health\",\"Health and medical history_Hearing\",\"Health and medical history_Medical conditions\",\"Health and medical history_Medication\",\"Health and medical history_Mouth\",\"Health and medical history_Operations\",\"Health and medical history_Pain\",\"Lifestyle and environment_Alcohol\",\"Lifestyle and environment_Diet\",\"Lifestyle and environment_Electronic device use\",\"Lifestyle and environment_Physical activity\",\"Lifestyle and environment_Sexual factors\",\"Lifestyle and environment_Sleep\",\"Lifestyle and environment_Smoking\",\"Lifestyle and environment_Sun exposure\",\"Psychosocial factors_Mental health\",\"Psychosocial factors_Socal support\",\"Sex-specific factors\",\"Sociodemographics_Early life factors\",\"Sociodemographics_Education\",\"Sociodemographics_Employment\",\"Sociodemographics_Ethnicity\",\"Sociodemographics_Household\",\"Sociodemographics_Other sociodemographic factors\"]\n","    \n","   \n","    ########################## perform adaptive GA based feature selection\n","    #size of chromosome 8X2\n","    evaluate_values = 16\n","        \n","    # pass dataset and dataTran to feature selection as ranksum uses dataformat of shape dataset while others use dataTran format\n","    optimizer = DiagnosticOptimization(dfTrain,dfTestFeaturesCols, dfTrain[\"CVD_followup\"], dfTest, dfTest[\"CVD_followup\"], evaluate_values,varDict['-pp'],varDict['-g'])\n","    trainingData,front1FinalIndv = optimizer.optimization() \n","    dataset = dfTest\n","    goal_values = np.zeros([len(front1FinalIndv), 8])\n","    iter=0\n","    for ind in front1FinalIndv:\n","        Questions_proportion = 0\n","        Single_blood_proportion = 0 \n","        Risk_scores_proportion = 0 \n","        predicted_output = [0 for _ in range(len(dataset))]\n","\n","        actualValuesStatus = []\n","        actualValuesFollowup = []\n","\n","        for i in range(len(dfTest)):\n","            patient = dataset.iloc[[i]]\n","            actualValuesStatus.append(dataset['CVD_status'].iloc[i])\n","            actualValuesFollowup.append(dataset['CVD_followup'].iloc[i])\n","\n","            predicted = evaluateBranch(\n","                modelsPickleList[0],\n","                cleaned_model_to_csvs[\"Sociodemographics\"],\n","                patient,\n","                modelsPickleList[2],\n","                cleaned_model_to_csvs[\"Sociodemographics_oneDrop_qrisk\"],\n","                ind[0],\n","                100\n","            )\n","\n","            if isinstance(predicted, float) and predicted < ind[0]:\n","                predicted_output[i] = predicted\n","                Questions_proportion += 97\n","\n","\n","            elif isinstance(predicted, float) and predicted > ind[1]:\n","                predicted_output[i] = predicted\n","                Questions_proportion += 97\n","                Single_blood_proportion += 185\n","                Risk_scores_proportion += 44\n","\n","            else:\n","                predicted = evaluateBranch(\n","                    modelsPickleList[3],\n","                    cleaned_model_to_csvs[\"Sociodemographics_Health and medical history\"],\n","                    patient,\n","                    modelsPickleList[5],\n","                    cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_oneDrop_qrisk\"],\n","                    ind[2],\n","                    ind[3]\n","                )\n","\n","                if isinstance(predicted, float) and predicted < ind[2]:\n","                    predicted_output[i] = predicted\n","                    Questions_proportion += 242\n","\n","                elif isinstance(predicted, float) and predicted > ind[3]:\n","                    predicted_output[i] = predicted\n","                    Questions_proportion += 242\n","                    Single_blood_proportion += 185\n","                    Risk_scores_proportion += 44\n","\n","                else:\n","                    predicted = evaluateBranch(\n","                        modelsPickleList[6],\n","                        cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors\"],\n","                        patient,\n","                        modelsPickleList[8],\n","                        cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_oneDrop_qrisk\"],\n","                        ind[4],\n","                        ind[5]\n","                    )\n","\n","                    if isinstance(predicted, float) and predicted < ind[4]:\n","                        predicted_output[i] = predicted\n","                        Questions_proportion += 312\n","\n","\n","                    elif isinstance(predicted, float) and predicted > ind[5]:\n","                        predicted_output[i] = predicted\n","                        Questions_proportion += 312\n","                        Single_blood_proportion += 185\n","                        Risk_scores_proportion += 44\n","\n","                    else:\n","                        predicted = evaluateBranch(\n","                            modelsPickleList[9],\n","                            cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors\"],\n","                            patient,\n","                            modelsPickleList[11],\n","                            cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_oneDrop_qrisk\"],\n","                            ind[6],\n","                            ind[7]\n","                        )\n","\n","                        if isinstance(predicted, float) and predicted < ind[6]:\n","                            predicted_output[i] = predicted\n","                            Questions_proportion += 319\n","\n","\n","                        elif isinstance(predicted, float) and predicted > ind[7]:\n","                            predicted_output[i] = predicted\n","                            Questions_proportion += 319\n","                            Single_blood_proportion += 185\n","                            Risk_scores_proportion += 44\n","\n","                        else:\n","                            predicted = evaluateBranch(\n","                                modelsPickleList[12],\n","                                cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history\"],\n","                                patient,\n","                                modelsPickleList[14],\n","                                cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_oneDrop_qrisk\"],\n","                                ind[8],\n","                                ind[9]\n","                            )\n","\n","                            if isinstance(predicted, float) and predicted < ind[8]:\n","                                predicted_output[i] = predicted\n","                                Questions_proportion += 376\n","\n","\n","                            elif isinstance(predicted, float) and predicted > ind[9]:\n","                                predicted_output[i] = predicted\n","                                Questions_proportion += 376\n","                                Single_blood_proportion += 185\n","                                Risk_scores_proportion += 44\n","\n","                            else:\n","                                predicted = evaluateBranch(\n","                                    modelsPickleList[15],\n","                                    cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment\"],\n","                                    patient,\n","                                    modelsPickleList[17],\n","                                    cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment_oneDrop_qrisk\"],\n","                                    ind[10],\n","                                    ind[11]\n","                                )\n","\n","                                if isinstance(predicted, float) and predicted < ind[10]:\n","                                    predicted_output[i] = predicted\n","                                    Questions_proportion += 703\n","\n","                                elif isinstance(predicted, float) and predicted > ind[11]:\n","                                    predicted_output[i] = predicted\n","                                    Questions_proportion += 703\n","                                    Single_blood_proportion += 185\n","                                    Risk_scores_proportion += 44\n","\n","                                #\"Make it into else to make sure catch everything\"\n","                                #if not isinstance(predicted_output[i], float) or predicted_output[i] == 0:\n","                                else:\n","                                    predicted = evaluateBranch(\n","                                        modelsPickleList[18],\n","                                        cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment_Psychosocial factors\"],\n","                                        patient,\n","                                        modelsPickleList[19],\n","                                        cleaned_model_to_csvs[\"Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop_qrisk\"],\n","                                        ind[12],\n","                                        ind[13]\n","                                    )\n","\n","                                    if isinstance(predicted, float) and predicted < ind[12]:\n","                                        predicted_output[i] = predicted\n","                                        Questions_proportion += 756\n","\n","                                    elif isinstance(predicted, float) and predicted > ind[13]:\n","                                        predicted_output[i] = predicted\n","                                        Questions_proportion += 756\n","                                        Single_blood_proportion += 185\n","                                        Risk_scores_proportion += 44\n","\n","                                    else:\n","                                        predicted = evaluateBranch(\n","                                        modelsPickleList[19],\n","                                        cleaned_model_to_csvs[\"Sociodemographics_Health and medical history_Sex-specific factors_Early life factors_Family history_Lifestyle and environment_Psychosocial factors_oneDrop\"],\n","                                        patient,\n","                                        modelsPickleList[20],\n","                                        cleaned_model_to_csvs[\"Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop_qrisk\"],\n","                                        ind[14],\n","                                        ind[15]\n","                                        )\n","                                        if isinstance(predicted, float) and predicted < ind[14]:\n","                                            predicted_output[i] = predicted\n","                                            Questions_proportion += 756\n","                                            Single_blood_proportion += 185\n","                                        else:\n","                                            predicted = predictIndividual(modelsPickleList[20], cleaned_model_to_csvs[\"Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop_qrisk\"], patient)[0]\n","                                            predicted_output[i] = predicted\n","                            \n","                                            Questions_proportion += 756\n","                                            Single_blood_proportion += 185\n","                                            Risk_scores_proportion += 44\n","        \n","\n","\n","        # Normalize proportions\n","        Questions_proportion /= (756 * len(dataset))\n","        Single_blood_proportion /= (185 * len(dataset))\n","        Risk_scores_proportion /= (44 * len(dataset))\n","\n","        #Risk score as a median is the average of the lowest risk score (-3) and the highest (+9) so 3\n","        y_pred_binary = np.array(predicted_output) >= ind[15]\n","        y_true = np.array(actualValuesStatus)\n","\n","        TP = np.sum((y_pred_binary == 1) & (y_true == 1))\n","        TN = np.sum((y_pred_binary == 0) & (y_true == 0))\n","        FP = np.sum((y_pred_binary == 1) & (y_true == 0))\n","        FN = np.sum((y_pred_binary == 0) & (y_true == 1))\n","\n","        True_positive_rate = TP / (TP + FN) if (TP + FN) else 0\n","        True_negative_rate = TN / (TN + FP) if (TN + FP) else 0\n","\n","        try:\n","            AUC = roc_auc_score(y_true, predicted_output)\n","        except ValueError:\n","            AUC = 0  \n","\n","        try:\n","            from sksurv.metrics import concordance_index_censored\n","            y_surv = np.array([(bool(e), t) for e, t in zip(y_true, actualValuesFollowup)],\n","                              dtype=[('event', bool), ('time', float)])\n","            Cindex = concordance_index_censored(y_surv['event'], y_surv['time'], predicted_output)[0]\n","        except:\n","            Cindex = 0\n","\n","        print(f\"Individual {ind}:\")\n","        print(f\"  1 - Questions proportion: {1 - Questions_proportion:.3f}\")\n","        print(f\"  1 - Single blood proportion: {1 - Single_blood_proportion:.3f}\")\n","        print(f\"  1 - Risk scores proportion: {1 - Risk_scores_proportion:.3f}\")\n","        print(f\"  True Positive Rate (TPR): {True_positive_rate:.3f}\")\n","        print(f\"  True Negative Rate (TNR): {True_negative_rate:.3f}\")\n","        print(f\"  AUC: {AUC:.3f}\")\n","        print(f\"  C-index: {Cindex:.3f}\")\n","\n","\n","        # Set goal values\n","        goal_values[iter][0] = 1 - Questions_proportion\n","        goal_values[iter][1] = 1 - Single_blood_proportion\n","        goal_values[iter][2] = 1 - Risk_scores_proportion\n","        goal_values[iter][3] = True_positive_rate\n","        goal_values[iter][4] = True_negative_rate\n","        goal_values[iter][5] = AUC\n","        goal_values[iter][6] = Cindex\n","        iter+=1\n","        df = pd.DataFrame(goal_values)\n","        df.to_csv(path_or_buf = pathFeature + \"testSetGoalVals.csv\", index=False, header=False) # pval data\n","\n","\n","\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["Individuals Initiatised\n","Starting GA based optimisation\n","\n","Generation:1\n","  1 - Questions proportion: 0.057\n","  1 - Single blood proportion: 0.105\n","  1 - Risk scores proportion: 0.106\n","  True Positive Rate (TPR): 0.717\n","  True Negative Rate (TNR): 0.703\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.021\n","  1 - Single blood proportion: 0.043\n","  1 - Risk scores proportion: 0.056\n","  True Positive Rate (TPR): 0.026\n","  True Negative Rate (TNR): 0.998\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.000\n","  1 - Risk scores proportion: 0.000\n","  True Positive Rate (TPR): 0.010\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.042\n","  1 - Single blood proportion: 0.210\n","  1 - Risk scores proportion: 0.210\n","  True Positive Rate (TPR): 0.972\n","  True Negative Rate (TNR): 0.214\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.045\n","  1 - Single blood proportion: 0.060\n","  1 - Risk scores proportion: 0.060\n","  True Positive Rate (TPR): 0.621\n","  True Negative Rate (TNR): 0.771\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.044\n","  1 - Single blood proportion: 0.074\n","  1 - Risk scores proportion: 0.099\n","  True Positive Rate (TPR): 0.945\n","  True Negative Rate (TNR): 0.312\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.010\n","  1 - Single blood proportion: 0.016\n","  1 - Risk scores proportion: 0.019\n","  True Positive Rate (TPR): 0.735\n","  True Negative Rate (TNR): 0.680\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.087\n","  1 - Single blood proportion: 0.452\n","  1 - Risk scores proportion: 0.452\n","  True Positive Rate (TPR): 0.330\n","  True Negative Rate (TNR): 0.936\n","  AUC: 0.768\n","  C-index: 0.764\n","  1 - Questions proportion: 0.049\n","  1 - Single blood proportion: 0.267\n","  1 - Risk scores proportion: 0.267\n","  True Positive Rate (TPR): 0.036\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.021\n","  1 - Single blood proportion: 0.039\n","  1 - Risk scores proportion: 0.384\n","  True Positive Rate (TPR): 0.113\n","  True Negative Rate (TNR): 0.986\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.073\n","  1 - Risk scores proportion: 0.180\n","  True Positive Rate (TPR): 0.285\n","  True Negative Rate (TNR): 0.946\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.035\n","  1 - Single blood proportion: 0.218\n","  1 - Risk scores proportion: 0.228\n","  True Positive Rate (TPR): 0.911\n","  True Negative Rate (TNR): 0.429\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.031\n","  1 - Risk scores proportion: 0.038\n","  True Positive Rate (TPR): 0.615\n","  True Negative Rate (TNR): 0.777\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.024\n","  1 - Single blood proportion: 0.050\n","  1 - Risk scores proportion: 0.050\n","  True Positive Rate (TPR): 0.960\n","  True Negative Rate (TNR): 0.271\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.041\n","  1 - Single blood proportion: 0.305\n","  1 - Risk scores proportion: 0.305\n","  True Positive Rate (TPR): 0.253\n","  True Negative Rate (TNR): 0.952\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.053\n","  1 - Single blood proportion: 0.105\n","  1 - Risk scores proportion: 0.256\n","  True Positive Rate (TPR): 0.751\n","  True Negative Rate (TNR): 0.650\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.037\n","  1 - Single blood proportion: 0.098\n","  1 - Risk scores proportion: 0.098\n","  True Positive Rate (TPR): 0.008\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.443\n","  1 - Risk scores proportion: 0.443\n","  True Positive Rate (TPR): 0.670\n","  True Negative Rate (TNR): 0.733\n","  AUC: 0.765\n","  C-index: 0.761\n","  1 - Questions proportion: 0.012\n","  1 - Single blood proportion: 0.288\n","  1 - Risk scores proportion: 0.288\n","  True Positive Rate (TPR): 0.678\n","  True Negative Rate (TNR): 0.731\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.035\n","  1 - Single blood proportion: 0.091\n","  1 - Risk scores proportion: 0.326\n","  True Positive Rate (TPR): 0.093\n","  True Negative Rate (TNR): 0.991\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.054\n","  1 - Single blood proportion: 0.116\n","  1 - Risk scores proportion: 0.118\n","  True Positive Rate (TPR): 0.065\n","  True Negative Rate (TNR): 0.996\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.033\n","  1 - Single blood proportion: 0.431\n","  1 - Risk scores proportion: 0.431\n","  True Positive Rate (TPR): 0.006\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.026\n","  1 - Single blood proportion: 0.062\n","  1 - Risk scores proportion: 0.062\n","  True Positive Rate (TPR): 0.055\n","  True Negative Rate (TNR): 0.996\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.062\n","  1 - Single blood proportion: 0.177\n","  1 - Risk scores proportion: 0.177\n","  True Positive Rate (TPR): 0.018\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.061\n","  1 - Single blood proportion: 0.490\n","  1 - Risk scores proportion: 0.503\n","  True Positive Rate (TPR): 0.012\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.770\n","  C-index: 0.766\n","  1 - Questions proportion: 0.083\n","  1 - Single blood proportion: 0.428\n","  1 - Risk scores proportion: 0.428\n","  True Positive Rate (TPR): 0.492\n","  True Negative Rate (TNR): 0.856\n","  AUC: 0.769\n","  C-index: 0.765\n","  1 - Questions proportion: 0.011\n","  1 - Single blood proportion: 0.124\n","  1 - Risk scores proportion: 0.362\n","  True Positive Rate (TPR): 0.105\n","  True Negative Rate (TNR): 0.990\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.019\n","  1 - Single blood proportion: 0.034\n","  1 - Risk scores proportion: 0.035\n","  True Positive Rate (TPR): 0.413\n","  True Negative Rate (TNR): 0.898\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.046\n","  1 - Single blood proportion: 0.078\n","  1 - Risk scores proportion: 0.220\n","  True Positive Rate (TPR): 0.024\n","  True Negative Rate (TNR): 0.998\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.008\n","  1 - Single blood proportion: 0.055\n","  1 - Risk scores proportion: 0.067\n","  True Positive Rate (TPR): 0.992\n","  True Negative Rate (TNR): 0.145\n","  AUC: 0.778\n","  C-index: 0.773\n","convergence value 0.10457191144225908\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:2\n","  1 - Questions proportion: 0.021\n","  1 - Single blood proportion: 0.267\n","  1 - Risk scores proportion: 0.267\n","  True Positive Rate (TPR): 0.036\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.142\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.223\n","  True Negative Rate (TNR): 0.962\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.443\n","  1 - Risk scores proportion: 0.443\n","  True Positive Rate (TPR): 0.670\n","  True Negative Rate (TNR): 0.733\n","  AUC: 0.765\n","  C-index: 0.761\n","  1 - Questions proportion: 0.042\n","  1 - Single blood proportion: 0.226\n","  1 - Risk scores proportion: 0.226\n","  True Positive Rate (TPR): 0.512\n","  True Negative Rate (TNR): 0.844\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.049\n","  1 - Single blood proportion: 0.267\n","  1 - Risk scores proportion: 0.267\n","  True Positive Rate (TPR): 0.036\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.017\n","  1 - Single blood proportion: 0.030\n","  1 - Risk scores proportion: 0.117\n","  True Positive Rate (TPR): 0.814\n","  True Negative Rate (TNR): 0.556\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.018\n","  1 - Single blood proportion: 0.036\n","  1 - Risk scores proportion: 0.036\n","  True Positive Rate (TPR): 0.921\n","  True Negative Rate (TNR): 0.419\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.064\n","  1 - Single blood proportion: 0.156\n","  1 - Risk scores proportion: 0.157\n","  True Positive Rate (TPR): 0.672\n","  True Negative Rate (TNR): 0.738\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.049\n","  1 - Single blood proportion: 0.431\n","  1 - Risk scores proportion: 0.431\n","  True Positive Rate (TPR): 0.006\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.773\n","  C-index: 0.768\n","  1 - Questions proportion: 0.023\n","  1 - Single blood proportion: 0.319\n","  1 - Risk scores proportion: 0.319\n","  True Positive Rate (TPR): 0.010\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.017\n","  1 - Single blood proportion: 0.029\n","  1 - Risk scores proportion: 0.030\n","  True Positive Rate (TPR): 0.415\n","  True Negative Rate (TNR): 0.896\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.023\n","  1 - Single blood proportion: 0.087\n","  1 - Risk scores proportion: 0.087\n","  True Positive Rate (TPR): 0.138\n","  True Negative Rate (TNR): 0.983\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.033\n","  1 - Single blood proportion: 0.431\n","  1 - Risk scores proportion: 0.431\n","  True Positive Rate (TPR): 0.006\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.034\n","  1 - Single blood proportion: 0.081\n","  1 - Risk scores proportion: 0.081\n","  True Positive Rate (TPR): 0.008\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.007\n","  1 - Single blood proportion: 0.095\n","  1 - Risk scores proportion: 0.182\n","  True Positive Rate (TPR): 0.285\n","  True Negative Rate (TNR): 0.946\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.013\n","  1 - Single blood proportion: 0.288\n","  1 - Risk scores proportion: 0.288\n","  True Positive Rate (TPR): 0.678\n","  True Negative Rate (TNR): 0.731\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.048\n","  1 - Single blood proportion: 0.130\n","  1 - Risk scores proportion: 0.380\n","  True Positive Rate (TPR): 0.937\n","  True Negative Rate (TNR): 0.351\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.000\n","  1 - Risk scores proportion: 0.001\n","  True Positive Rate (TPR): 0.024\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.005\n","  1 - Single blood proportion: 0.051\n","  1 - Risk scores proportion: 0.052\n","  True Positive Rate (TPR): 0.038\n","  True Negative Rate (TNR): 0.998\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.012\n","  1 - Single blood proportion: 0.049\n","  1 - Risk scores proportion: 0.053\n","  True Positive Rate (TPR): 0.010\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.058\n","  1 - Single blood proportion: 0.100\n","  1 - Risk scores proportion: 0.101\n","  True Positive Rate (TPR): 0.020\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.032\n","  1 - Single blood proportion: 0.431\n","  1 - Risk scores proportion: 0.431\n","  True Positive Rate (TPR): 0.615\n","  True Negative Rate (TNR): 0.778\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.047\n","  1 - Single blood proportion: 0.452\n","  1 - Risk scores proportion: 0.491\n","  True Positive Rate (TPR): 0.330\n","  True Negative Rate (TNR): 0.936\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.035\n","  1 - Risk scores proportion: 0.053\n","  True Positive Rate (TPR): 0.006\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.074\n","  1 - Single blood proportion: 0.103\n","  1 - Risk scores proportion: 0.107\n","  True Positive Rate (TPR): 0.075\n","  True Negative Rate (TNR): 0.989\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.021\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.036\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.772\n","  C-index: 0.768\n","  1 - Questions proportion: 0.057\n","  1 - Single blood proportion: 0.200\n","  1 - Risk scores proportion: 0.200\n","  True Positive Rate (TPR): 0.717\n","  True Negative Rate (TNR): 0.703\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.089\n","  1 - Single blood proportion: 0.392\n","  1 - Risk scores proportion: 0.392\n","  True Positive Rate (TPR): 0.387\n","  True Negative Rate (TNR): 0.910\n","  AUC: 0.771\n","  C-index: 0.766\n","  1 - Questions proportion: 0.015\n","  1 - Single blood proportion: 0.212\n","  1 - Risk scores proportion: 0.212\n","  True Positive Rate (TPR): 0.040\n","  True Negative Rate (TNR): 0.996\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.012\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.949\n","  True Negative Rate (TNR): 0.297\n","  AUC: 0.772\n","  C-index: 0.767\n","convergence value 0.10049983401704249\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:3\n","  1 - Questions proportion: 0.012\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.949\n","  True Negative Rate (TNR): 0.297\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.198\n","  1 - Risk scores proportion: 0.198\n","  True Positive Rate (TPR): 0.010\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.021\n","  1 - Single blood proportion: 0.155\n","  1 - Risk scores proportion: 0.155\n","  True Positive Rate (TPR): 0.188\n","  True Negative Rate (TNR): 0.971\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.005\n","  1 - Single blood proportion: 0.003\n","  1 - Risk scores proportion: 0.003\n","  True Positive Rate (TPR): 0.638\n","  True Negative Rate (TNR): 0.763\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.017\n","  1 - Single blood proportion: 0.198\n","  1 - Risk scores proportion: 0.198\n","  True Positive Rate (TPR): 0.717\n","  True Negative Rate (TNR): 0.703\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.064\n","  1 - Single blood proportion: 0.101\n","  1 - Risk scores proportion: 0.102\n","  True Positive Rate (TPR): 0.994\n","  True Negative Rate (TNR): 0.074\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.057\n","  1 - Single blood proportion: 0.091\n","  1 - Risk scores proportion: 0.093\n","  True Positive Rate (TPR): 0.998\n","  True Negative Rate (TNR): 0.053\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.288\n","  1 - Risk scores proportion: 0.288\n","  True Positive Rate (TPR): 0.678\n","  True Negative Rate (TNR): 0.731\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.038\n","  1 - Single blood proportion: 0.536\n","  1 - Risk scores proportion: 0.536\n","  True Positive Rate (TPR): 0.010\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.771\n","  C-index: 0.767\n","  1 - Questions proportion: 0.020\n","  1 - Single blood proportion: 0.039\n","  1 - Risk scores proportion: 0.040\n","  True Positive Rate (TPR): 0.138\n","  True Negative Rate (TNR): 0.983\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.047\n","  1 - Single blood proportion: 0.564\n","  1 - Risk scores proportion: 0.573\n","  True Positive Rate (TPR): 0.330\n","  True Negative Rate (TNR): 0.936\n","  AUC: 0.771\n","  C-index: 0.767\n","  1 - Questions proportion: 0.032\n","  1 - Single blood proportion: 0.229\n","  1 - Risk scores proportion: 0.420\n","  True Positive Rate (TPR): 0.010\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.015\n","  1 - Single blood proportion: 0.024\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.330\n","  True Negative Rate (TNR): 0.936\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.001\n","  1 - Risk scores proportion: 0.044\n","  True Positive Rate (TPR): 0.348\n","  True Negative Rate (TNR): 0.925\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.037\n","  1 - Risk scores proportion: 0.062\n","  True Positive Rate (TPR): 0.067\n","  True Negative Rate (TNR): 0.993\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.072\n","  1 - Single blood proportion: 0.288\n","  1 - Risk scores proportion: 0.288\n","  True Positive Rate (TPR): 0.678\n","  True Negative Rate (TNR): 0.731\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.012\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.949\n","  True Negative Rate (TNR): 0.297\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.007\n","  1 - Single blood proportion: 0.036\n","  1 - Risk scores proportion: 0.053\n","  True Positive Rate (TPR): 0.006\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.009\n","  1 - Single blood proportion: 0.175\n","  1 - Risk scores proportion: 0.175\n","  True Positive Rate (TPR): 0.018\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.018\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.980\n","  True Negative Rate (TNR): 0.203\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.000\n","  1 - Risk scores proportion: 0.001\n","  True Positive Rate (TPR): 0.024\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.045\n","  1 - Single blood proportion: 0.452\n","  1 - Risk scores proportion: 0.477\n","  True Positive Rate (TPR): 0.330\n","  True Negative Rate (TNR): 0.936\n","  AUC: 0.774\n","  C-index: 0.770\n","  1 - Questions proportion: 0.004\n","  1 - Single blood proportion: 0.180\n","  1 - Risk scores proportion: 0.199\n","  True Positive Rate (TPR): 0.506\n","  True Negative Rate (TNR): 0.848\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.025\n","  1 - Single blood proportion: 0.043\n","  1 - Risk scores proportion: 0.046\n","  True Positive Rate (TPR): 0.040\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.064\n","  1 - Single blood proportion: 0.156\n","  1 - Risk scores proportion: 0.157\n","  True Positive Rate (TPR): 0.672\n","  True Negative Rate (TNR): 0.738\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.001\n","  1 - Risk scores proportion: 0.001\n","  True Positive Rate (TPR): 0.024\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.041\n","  1 - Single blood proportion: 0.253\n","  1 - Risk scores proportion: 0.253\n","  True Positive Rate (TPR): 0.105\n","  True Negative Rate (TNR): 0.989\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.008\n","  1 - Single blood proportion: 0.013\n","  1 - Risk scores proportion: 0.013\n","  True Positive Rate (TPR): 0.512\n","  True Negative Rate (TNR): 0.844\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.058\n","  1 - Single blood proportion: 0.100\n","  1 - Risk scores proportion: 0.101\n","  True Positive Rate (TPR): 0.020\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.014\n","  1 - Single blood proportion: 0.499\n","  1 - Risk scores proportion: 0.501\n","  True Positive Rate (TPR): 0.842\n","  True Negative Rate (TNR): 0.512\n","  AUC: 0.773\n","  C-index: 0.768\n","convergence value 0.11185273143716934\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:4\n","  1 - Questions proportion: 0.047\n","  1 - Single blood proportion: 0.564\n","  1 - Risk scores proportion: 0.573\n","  True Positive Rate (TPR): 0.330\n","  True Negative Rate (TNR): 0.936\n","  AUC: 0.771\n","  C-index: 0.767\n","  1 - Questions proportion: 0.059\n","  1 - Single blood proportion: 0.100\n","  1 - Risk scores proportion: 0.101\n","  True Positive Rate (TPR): 0.020\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.015\n","  1 - Single blood proportion: 0.499\n","  1 - Risk scores proportion: 0.500\n","  True Positive Rate (TPR): 0.978\n","  True Negative Rate (TNR): 0.224\n","  AUC: 0.773\n","  C-index: 0.768\n","  1 - Questions proportion: 0.015\n","  1 - Single blood proportion: 0.024\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.330\n","  True Negative Rate (TNR): 0.936\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.000\n","  1 - Risk scores proportion: 0.001\n","  True Positive Rate (TPR): 0.024\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.000\n","  1 - Risk scores proportion: 0.001\n","  True Positive Rate (TPR): 0.022\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.006\n","  1 - Single blood proportion: 0.198\n","  1 - Risk scores proportion: 0.198\n","  True Positive Rate (TPR): 0.970\n","  True Negative Rate (TNR): 0.253\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.082\n","  1 - Single blood proportion: 0.314\n","  1 - Risk scores proportion: 0.314\n","  True Positive Rate (TPR): 0.012\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.773\n","  C-index: 0.769\n","  1 - Questions proportion: 0.020\n","  1 - Single blood proportion: 0.268\n","  1 - Risk scores proportion: 0.268\n","  True Positive Rate (TPR): 0.012\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.023\n","  1 - Single blood proportion: 0.357\n","  1 - Risk scores proportion: 0.357\n","  True Positive Rate (TPR): 0.573\n","  True Negative Rate (TNR): 0.815\n","  AUC: 0.779\n","  C-index: 0.774\n","  1 - Questions proportion: 0.014\n","  1 - Single blood proportion: 0.499\n","  1 - Risk scores proportion: 0.501\n","  True Positive Rate (TPR): 0.842\n","  True Negative Rate (TNR): 0.512\n","  AUC: 0.773\n","  C-index: 0.768\n","  1 - Questions proportion: 0.012\n","  1 - Single blood proportion: 0.330\n","  1 - Risk scores proportion: 0.330\n","  True Positive Rate (TPR): 0.449\n","  True Negative Rate (TNR): 0.883\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.013\n","  1 - Single blood proportion: 0.175\n","  1 - Risk scores proportion: 0.175\n","  True Positive Rate (TPR): 0.040\n","  True Negative Rate (TNR): 0.996\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.018\n","  1 - Single blood proportion: 0.134\n","  1 - Risk scores proportion: 0.134\n","  True Positive Rate (TPR): 0.154\n","  True Negative Rate (TNR): 0.979\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.002\n","  1 - Single blood proportion: 0.155\n","  1 - Risk scores proportion: 0.155\n","  True Positive Rate (TPR): 0.018\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.076\n","  1 - Single blood proportion: 0.501\n","  1 - Risk scores proportion: 0.501\n","  True Positive Rate (TPR): 0.735\n","  True Negative Rate (TNR): 0.661\n","  AUC: 0.760\n","  C-index: 0.756\n","  1 - Questions proportion: 0.035\n","  1 - Single blood proportion: 0.435\n","  1 - Risk scores proportion: 0.458\n","  True Positive Rate (TPR): 0.336\n","  True Negative Rate (TNR): 0.933\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.075\n","  1 - Single blood proportion: 0.142\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.330\n","  True Negative Rate (TNR): 0.936\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.005\n","  1 - Single blood proportion: 0.288\n","  1 - Risk scores proportion: 0.288\n","  True Positive Rate (TPR): 0.678\n","  True Negative Rate (TNR): 0.731\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.004\n","  1 - Single blood proportion: 0.004\n","  1 - Risk scores proportion: 0.212\n","  True Positive Rate (TPR): 0.107\n","  True Negative Rate (TNR): 0.989\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.049\n","  1 - Single blood proportion: 0.079\n","  1 - Risk scores proportion: 0.079\n","  True Positive Rate (TPR): 0.984\n","  True Negative Rate (TNR): 0.133\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.288\n","  1 - Risk scores proportion: 0.288\n","  True Positive Rate (TPR): 0.678\n","  True Negative Rate (TNR): 0.731\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.041\n","  1 - Single blood proportion: 0.253\n","  1 - Risk scores proportion: 0.253\n","  True Positive Rate (TPR): 0.105\n","  True Negative Rate (TNR): 0.989\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.039\n","  1 - Single blood proportion: 0.536\n","  1 - Risk scores proportion: 0.536\n","  True Positive Rate (TPR): 0.032\n","  True Negative Rate (TNR): 0.998\n","  AUC: 0.771\n","  C-index: 0.767\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.000\n","  1 - Risk scores proportion: 0.000\n","  True Positive Rate (TPR): 0.026\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.065\n","  1 - Single blood proportion: 0.156\n","  1 - Risk scores proportion: 0.157\n","  True Positive Rate (TPR): 0.672\n","  True Negative Rate (TNR): 0.738\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.034\n","  1 - Single blood proportion: 0.155\n","  1 - Risk scores proportion: 0.156\n","  True Positive Rate (TPR): 0.749\n","  True Negative Rate (TNR): 0.652\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.001\n","  1 - Risk scores proportion: 0.001\n","  True Positive Rate (TPR): 0.018\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.018\n","  1 - Single blood proportion: 0.516\n","  1 - Risk scores proportion: 0.533\n","  True Positive Rate (TPR): 0.285\n","  True Negative Rate (TNR): 0.944\n","  AUC: 0.772\n","  C-index: 0.768\n","convergence value 0.11564430283407007\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:5\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.051\n","  1 - Single blood proportion: 0.536\n","  1 - Risk scores proportion: 0.536\n","  True Positive Rate (TPR): 0.026\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.770\n","  C-index: 0.765\n","  1 - Questions proportion: 0.009\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.038\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.772\n","  C-index: 0.768\n","  1 - Questions proportion: 0.082\n","  1 - Single blood proportion: 0.145\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.330\n","  True Negative Rate (TNR): 0.936\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.019\n","  1 - Single blood proportion: 0.268\n","  1 - Risk scores proportion: 0.268\n","  True Positive Rate (TPR): 0.010\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.017\n","  1 - Single blood proportion: 0.053\n","  1 - Risk scores proportion: 0.236\n","  True Positive Rate (TPR): 0.164\n","  True Negative Rate (TNR): 0.978\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.003\n","  1 - Risk scores proportion: 0.003\n","  True Positive Rate (TPR): 0.018\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.039\n","  1 - Single blood proportion: 0.536\n","  1 - Risk scores proportion: 0.536\n","  True Positive Rate (TPR): 0.115\n","  True Negative Rate (TNR): 0.986\n","  AUC: 0.771\n","  C-index: 0.767\n","  1 - Questions proportion: 0.069\n","  1 - Single blood proportion: 0.306\n","  1 - Risk scores proportion: 0.306\n","  True Positive Rate (TPR): 0.577\n","  True Negative Rate (TNR): 0.806\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.014\n","  1 - Single blood proportion: 0.023\n","  1 - Risk scores proportion: 0.030\n","  True Positive Rate (TPR): 0.020\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.013\n","  1 - Single blood proportion: 0.177\n","  1 - Risk scores proportion: 0.177\n","  True Positive Rate (TPR): 0.132\n","  True Negative Rate (TNR): 0.984\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.004\n","  1 - Single blood proportion: 0.310\n","  1 - Risk scores proportion: 0.310\n","  True Positive Rate (TPR): 0.563\n","  True Negative Rate (TNR): 0.823\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.012\n","  1 - Single blood proportion: 0.141\n","  1 - Risk scores proportion: 0.278\n","  True Positive Rate (TPR): 0.986\n","  True Negative Rate (TNR): 0.179\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.008\n","  1 - Single blood proportion: 0.045\n","  1 - Risk scores proportion: 0.084\n","  True Positive Rate (TPR): 0.008\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.099\n","  1 - Single blood proportion: 0.446\n","  1 - Risk scores proportion: 0.446\n","  True Positive Rate (TPR): 0.045\n","  True Negative Rate (TNR): 0.996\n","  AUC: 0.765\n","  C-index: 0.761\n","  1 - Questions proportion: 0.041\n","  1 - Single blood proportion: 0.374\n","  1 - Risk scores proportion: 0.374\n","  True Positive Rate (TPR): 0.858\n","  True Negative Rate (TNR): 0.512\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.013\n","  1 - Single blood proportion: 0.499\n","  1 - Risk scores proportion: 0.500\n","  True Positive Rate (TPR): 0.962\n","  True Negative Rate (TNR): 0.265\n","  AUC: 0.773\n","  C-index: 0.768\n","  1 - Questions proportion: 0.036\n","  1 - Single blood proportion: 0.061\n","  1 - Risk scores proportion: 0.061\n","  True Positive Rate (TPR): 0.951\n","  True Negative Rate (TNR): 0.305\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.060\n","  1 - Single blood proportion: 0.499\n","  1 - Risk scores proportion: 0.501\n","  True Positive Rate (TPR): 0.887\n","  True Negative Rate (TNR): 0.404\n","  AUC: 0.766\n","  C-index: 0.762\n","  1 - Questions proportion: 0.062\n","  1 - Single blood proportion: 0.263\n","  1 - Risk scores proportion: 0.263\n","  True Positive Rate (TPR): 0.016\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.003\n","  1 - Single blood proportion: 0.148\n","  1 - Risk scores proportion: 0.148\n","  True Positive Rate (TPR): 0.018\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.000\n","  1 - Risk scores proportion: 0.001\n","  True Positive Rate (TPR): 0.457\n","  True Negative Rate (TNR): 0.876\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.020\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.804\n","  True Negative Rate (TNR): 0.555\n","  AUC: 0.770\n","  C-index: 0.766\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.004\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.140\n","  True Negative Rate (TNR): 0.983\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.008\n","  1 - Single blood proportion: 0.055\n","  1 - Risk scores proportion: 0.228\n","  True Positive Rate (TPR): 0.229\n","  True Negative Rate (TNR): 0.960\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.013\n","  1 - Single blood proportion: 0.132\n","  1 - Risk scores proportion: 0.320\n","  True Positive Rate (TPR): 0.055\n","  True Negative Rate (TNR): 0.995\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.070\n","  1 - Single blood proportion: 0.164\n","  1 - Risk scores proportion: 0.164\n","  True Positive Rate (TPR): 0.895\n","  True Negative Rate (TNR): 0.463\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.013\n","  1 - Single blood proportion: 0.175\n","  1 - Risk scores proportion: 0.175\n","  True Positive Rate (TPR): 0.057\n","  True Negative Rate (TNR): 0.995\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.045\n","  1 - Single blood proportion: 0.253\n","  1 - Risk scores proportion: 0.253\n","  True Positive Rate (TPR): 0.105\n","  True Negative Rate (TNR): 0.989\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.013\n","  1 - Risk scores proportion: 0.013\n","  True Positive Rate (TPR): 0.024\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.773\n","convergence value 0.12037337005394533\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:6\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.339\n","  1 - Risk scores proportion: 0.339\n","  True Positive Rate (TPR): 0.006\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.041\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.980\n","  True Negative Rate (TNR): 0.157\n","  AUC: 0.770\n","  C-index: 0.765\n","  1 - Questions proportion: 0.041\n","  1 - Single blood proportion: 0.253\n","  1 - Risk scores proportion: 0.253\n","  True Positive Rate (TPR): 0.105\n","  True Negative Rate (TNR): 0.989\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.060\n","  1 - Single blood proportion: 0.499\n","  1 - Risk scores proportion: 0.499\n","  True Positive Rate (TPR): 0.887\n","  True Negative Rate (TNR): 0.403\n","  AUC: 0.766\n","  C-index: 0.762\n","  1 - Questions proportion: 0.077\n","  1 - Single blood proportion: 0.142\n","  1 - Risk scores proportion: 0.238\n","  True Positive Rate (TPR): 0.229\n","  True Negative Rate (TNR): 0.960\n","  AUC: 0.773\n","  C-index: 0.768\n","  1 - Questions proportion: 0.015\n","  1 - Single blood proportion: 0.041\n","  1 - Risk scores proportion: 0.041\n","  True Positive Rate (TPR): 0.020\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.013\n","  1 - Single blood proportion: 0.499\n","  1 - Risk scores proportion: 0.500\n","  True Positive Rate (TPR): 0.962\n","  True Negative Rate (TNR): 0.265\n","  AUC: 0.773\n","  C-index: 0.768\n","  1 - Questions proportion: 0.005\n","  1 - Single blood proportion: 0.038\n","  1 - Risk scores proportion: 0.262\n","  True Positive Rate (TPR): 0.206\n","  True Negative Rate (TNR): 0.967\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.069\n","  1 - Single blood proportion: 0.115\n","  1 - Risk scores proportion: 0.374\n","  True Positive Rate (TPR): 0.043\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.077\n","  1 - Single blood proportion: 0.255\n","  1 - Risk scores proportion: 0.255\n","  True Positive Rate (TPR): 0.105\n","  True Negative Rate (TNR): 0.989\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.005\n","  1 - Risk scores proportion: 0.018\n","  True Positive Rate (TPR): 0.020\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.004\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.772\n","  C-index: 0.768\n","  1 - Questions proportion: 0.053\n","  1 - Single blood proportion: 0.156\n","  1 - Risk scores proportion: 0.156\n","  True Positive Rate (TPR): 0.897\n","  True Negative Rate (TNR): 0.462\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.075\n","  1 - Single blood proportion: 0.184\n","  1 - Risk scores proportion: 0.184\n","  True Positive Rate (TPR): 0.132\n","  True Negative Rate (TNR): 0.984\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.070\n","  1 - Risk scores proportion: 0.070\n","  True Positive Rate (TPR): 0.457\n","  True Negative Rate (TNR): 0.876\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.070\n","  1 - Single blood proportion: 0.164\n","  1 - Risk scores proportion: 0.164\n","  True Positive Rate (TPR): 0.895\n","  True Negative Rate (TNR): 0.463\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.003\n","  1 - Single blood proportion: 0.019\n","  1 - Risk scores proportion: 0.252\n","  True Positive Rate (TPR): 0.130\n","  True Negative Rate (TNR): 0.984\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.007\n","  1 - Single blood proportion: 0.107\n","  1 - Risk scores proportion: 0.303\n","  True Positive Rate (TPR): 0.055\n","  True Negative Rate (TNR): 0.996\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.306\n","  1 - Risk scores proportion: 0.306\n","  True Positive Rate (TPR): 0.988\n","  True Negative Rate (TNR): 0.107\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.001\n","  1 - Risk scores proportion: 0.001\n","  True Positive Rate (TPR): 0.540\n","  True Negative Rate (TNR): 0.835\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.070\n","  1 - Single blood proportion: 0.536\n","  1 - Risk scores proportion: 0.536\n","  True Positive Rate (TPR): 0.889\n","  True Negative Rate (TNR): 0.444\n","  AUC: 0.767\n","  C-index: 0.763\n","  1 - Questions proportion: 0.018\n","  1 - Single blood proportion: 0.244\n","  1 - Risk scores proportion: 0.244\n","  True Positive Rate (TPR): 0.816\n","  True Negative Rate (TNR): 0.554\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.019\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.804\n","  True Negative Rate (TNR): 0.555\n","  AUC: 0.770\n","  C-index: 0.766\n","  1 - Questions proportion: 0.004\n","  1 - Single blood proportion: 0.105\n","  1 - Risk scores proportion: 0.105\n","  True Positive Rate (TPR): 0.026\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.013\n","  1 - Single blood proportion: 0.148\n","  1 - Risk scores proportion: 0.157\n","  True Positive Rate (TPR): 0.055\n","  True Negative Rate (TNR): 0.995\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.081\n","  1 - Single blood proportion: 0.306\n","  1 - Risk scores proportion: 0.306\n","  True Positive Rate (TPR): 0.579\n","  True Negative Rate (TNR): 0.806\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.072\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.010\n","  1 - Single blood proportion: 0.154\n","  1 - Risk scores proportion: 0.156\n","  True Positive Rate (TPR): 0.038\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.778\n","  C-index: 0.772\n","convergence value 0.10784941744872156\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:7\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.463\n","  1 - Risk scores proportion: 0.463\n","  True Positive Rate (TPR): 0.146\n","  True Negative Rate (TNR): 0.982\n","  AUC: 0.773\n","  C-index: 0.769\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.435\n","  1 - Risk scores proportion: 0.435\n","  True Positive Rate (TPR): 0.079\n","  True Negative Rate (TNR): 0.993\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.074\n","  1 - Single blood proportion: 0.170\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.038\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.009\n","  1 - Single blood proportion: 0.122\n","  1 - Risk scores proportion: 0.142\n","  True Positive Rate (TPR): 0.043\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.023\n","  1 - Single blood proportion: 0.280\n","  1 - Risk scores proportion: 0.309\n","  True Positive Rate (TPR): 0.186\n","  True Negative Rate (TNR): 0.972\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.033\n","  1 - Single blood proportion: 0.066\n","  1 - Risk scores proportion: 0.066\n","  True Positive Rate (TPR): 0.229\n","  True Negative Rate (TNR): 0.960\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.069\n","  1 - Single blood proportion: 0.133\n","  1 - Risk scores proportion: 0.133\n","  True Positive Rate (TPR): 0.016\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.062\n","  1 - Single blood proportion: 0.162\n","  1 - Risk scores proportion: 0.162\n","  True Positive Rate (TPR): 0.038\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.142\n","  1 - Risk scores proportion: 0.374\n","  True Positive Rate (TPR): 0.043\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.773\n","  C-index: 0.768\n","  1 - Questions proportion: 0.060\n","  1 - Single blood proportion: 0.147\n","  1 - Risk scores proportion: 0.209\n","  True Positive Rate (TPR): 0.962\n","  True Negative Rate (TNR): 0.245\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.115\n","  1 - Risk scores proportion: 0.267\n","  True Positive Rate (TPR): 0.832\n","  True Negative Rate (TNR): 0.534\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.075\n","  1 - Single blood proportion: 0.184\n","  1 - Risk scores proportion: 0.184\n","  True Positive Rate (TPR): 0.132\n","  True Negative Rate (TNR): 0.984\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.035\n","  1 - Single blood proportion: 0.415\n","  1 - Risk scores proportion: 0.415\n","  True Positive Rate (TPR): 0.844\n","  True Negative Rate (TNR): 0.523\n","  AUC: 0.771\n","  C-index: 0.767\n","  1 - Questions proportion: 0.090\n","  1 - Single blood proportion: 0.554\n","  1 - Risk scores proportion: 0.554\n","  True Positive Rate (TPR): 0.731\n","  True Negative Rate (TNR): 0.662\n","  AUC: 0.764\n","  C-index: 0.760\n","  1 - Questions proportion: 0.069\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.808\n","  True Negative Rate (TNR): 0.528\n","  AUC: 0.762\n","  C-index: 0.758\n","  1 - Questions proportion: 0.062\n","  1 - Single blood proportion: 0.515\n","  1 - Risk scores proportion: 0.515\n","  True Positive Rate (TPR): 0.986\n","  True Negative Rate (TNR): 0.119\n","  AUC: 0.762\n","  C-index: 0.758\n","  1 - Questions proportion: 0.037\n","  1 - Single blood proportion: 0.068\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.040\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.054\n","  1 - Single blood proportion: 0.389\n","  1 - Risk scores proportion: 0.389\n","  True Positive Rate (TPR): 0.949\n","  True Negative Rate (TNR): 0.268\n","  AUC: 0.773\n","  C-index: 0.769\n","  1 - Questions proportion: 0.002\n","  1 - Single blood proportion: 0.146\n","  1 - Risk scores proportion: 0.146\n","  True Positive Rate (TPR): 0.953\n","  True Negative Rate (TNR): 0.285\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.056\n","  1 - Single blood proportion: 0.405\n","  1 - Risk scores proportion: 0.405\n","  True Positive Rate (TPR): 0.947\n","  True Negative Rate (TNR): 0.283\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.038\n","  1 - Single blood proportion: 0.253\n","  1 - Risk scores proportion: 0.253\n","  True Positive Rate (TPR): 0.010\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.063\n","  1 - Single blood proportion: 0.499\n","  1 - Risk scores proportion: 0.499\n","  True Positive Rate (TPR): 0.887\n","  True Negative Rate (TNR): 0.403\n","  AUC: 0.766\n","  C-index: 0.762\n","  1 - Questions proportion: 0.057\n","  1 - Single blood proportion: 0.105\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.213\n","  True Negative Rate (TNR): 0.965\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.115\n","  1 - Risk scores proportion: 0.267\n","  True Positive Rate (TPR): 0.899\n","  True Negative Rate (TNR): 0.455\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.125\n","  1 - Risk scores proportion: 0.305\n","  True Positive Rate (TPR): 0.055\n","  True Negative Rate (TNR): 0.996\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.004\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.572\n","  True Positive Rate (TPR): 0.668\n","  True Negative Rate (TNR): 0.736\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.002\n","  1 - Single blood proportion: 0.011\n","  1 - Risk scores proportion: 0.011\n","  True Positive Rate (TPR): 0.140\n","  True Negative Rate (TNR): 0.983\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.112\n","  1 - Single blood proportion: 0.536\n","  1 - Risk scores proportion: 0.537\n","  True Positive Rate (TPR): 0.229\n","  True Negative Rate (TNR): 0.960\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.368\n","  1 - Risk scores proportion: 0.368\n","  True Positive Rate (TPR): 0.030\n","  True Negative Rate (TNR): 0.998\n","  AUC: 0.777\n","  C-index: 0.772\n","convergence value 0.10219680496993411\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:8\n","  1 - Questions proportion: 0.100\n","  1 - Single blood proportion: 0.567\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.071\n","  True Negative Rate (TNR): 0.993\n","  AUC: 0.761\n","  C-index: 0.757\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.074\n","  1 - Single blood proportion: 0.168\n","  1 - Risk scores proportion: 0.169\n","  True Positive Rate (TPR): 0.091\n","  True Negative Rate (TNR): 0.991\n","  AUC: 0.774\n","  C-index: 0.768\n","  1 - Questions proportion: 0.019\n","  1 - Single blood proportion: 0.052\n","  1 - Risk scores proportion: 0.057\n","  True Positive Rate (TPR): 0.113\n","  True Negative Rate (TNR): 0.988\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.018\n","  1 - Single blood proportion: 0.071\n","  1 - Risk scores proportion: 0.147\n","  True Positive Rate (TPR): 0.073\n","  True Negative Rate (TNR): 0.993\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.029\n","  1 - Single blood proportion: 0.107\n","  1 - Risk scores proportion: 0.107\n","  True Positive Rate (TPR): 0.016\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.030\n","  1 - Single blood proportion: 0.058\n","  1 - Risk scores proportion: 0.194\n","  True Positive Rate (TPR): 0.953\n","  True Negative Rate (TNR): 0.286\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.010\n","  1 - Single blood proportion: 0.120\n","  1 - Risk scores proportion: 0.140\n","  True Positive Rate (TPR): 0.043\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.034\n","  1 - Single blood proportion: 0.415\n","  1 - Risk scores proportion: 0.415\n","  True Positive Rate (TPR): 0.844\n","  True Negative Rate (TNR): 0.523\n","  AUC: 0.771\n","  C-index: 0.767\n","  1 - Questions proportion: 0.059\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.806\n","  True Negative Rate (TNR): 0.530\n","  AUC: 0.762\n","  C-index: 0.758\n","  1 - Questions proportion: 0.071\n","  1 - Single blood proportion: 0.141\n","  1 - Risk scores proportion: 0.141\n","  True Positive Rate (TPR): 0.038\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.007\n","  1 - Single blood proportion: 0.072\n","  1 - Risk scores proportion: 0.072\n","  True Positive Rate (TPR): 0.024\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.004\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.340\n","  True Negative Rate (TNR): 0.928\n","  AUC: 0.772\n","  C-index: 0.768\n","  1 - Questions proportion: 0.010\n","  1 - Single blood proportion: 0.035\n","  1 - Risk scores proportion: 0.035\n","  True Positive Rate (TPR): 0.081\n","  True Negative Rate (TNR): 0.993\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.011\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.545\n","  True Negative Rate (TNR): 0.835\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.572\n","  True Positive Rate (TPR): 0.668\n","  True Negative Rate (TNR): 0.736\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.036\n","  1 - Single blood proportion: 0.435\n","  1 - Risk scores proportion: 0.435\n","  True Positive Rate (TPR): 0.079\n","  True Negative Rate (TNR): 0.993\n","  AUC: 0.772\n","  C-index: 0.768\n","  1 - Questions proportion: 0.069\n","  1 - Single blood proportion: 0.133\n","  1 - Risk scores proportion: 0.133\n","  True Positive Rate (TPR): 0.016\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.072\n","  1 - Single blood proportion: 0.142\n","  1 - Risk scores proportion: 0.142\n","  True Positive Rate (TPR): 0.140\n","  True Negative Rate (TNR): 0.983\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.056\n","  1 - Single blood proportion: 0.095\n","  1 - Risk scores proportion: 0.200\n","  True Positive Rate (TPR): 0.962\n","  True Negative Rate (TNR): 0.242\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.020\n","  1 - Single blood proportion: 0.033\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.040\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.034\n","  1 - Single blood proportion: 0.241\n","  1 - Risk scores proportion: 0.241\n","  True Positive Rate (TPR): 0.010\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.038\n","  1 - Single blood proportion: 0.253\n","  1 - Risk scores proportion: 0.262\n","  True Positive Rate (TPR): 0.190\n","  True Negative Rate (TNR): 0.971\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.012\n","  1 - Single blood proportion: 0.216\n","  1 - Risk scores proportion: 0.220\n","  True Positive Rate (TPR): 0.008\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.054\n","  1 - Single blood proportion: 0.369\n","  1 - Risk scores proportion: 0.385\n","  True Positive Rate (TPR): 0.024\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.054\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.765\n","  C-index: 0.761\n","  1 - Questions proportion: 0.099\n","  1 - Single blood proportion: 0.565\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.682\n","  True Negative Rate (TNR): 0.724\n","  AUC: 0.760\n","  C-index: 0.756\n","  1 - Questions proportion: 0.076\n","  1 - Single blood proportion: 0.142\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.213\n","  True Negative Rate (TNR): 0.965\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.555\n","  1 - Risk scores proportion: 0.555\n","  True Positive Rate (TPR): 0.727\n","  True Negative Rate (TNR): 0.673\n","  AUC: 0.770\n","  C-index: 0.765\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.115\n","  1 - Risk scores proportion: 0.243\n","  True Positive Rate (TPR): 0.996\n","  True Negative Rate (TNR): 0.055\n","  AUC: 0.776\n","  C-index: 0.770\n","convergence value 0.11430563758409501\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:9\n","  1 - Questions proportion: 0.048\n","  1 - Single blood proportion: 0.565\n","  1 - Risk scores proportion: 0.566\n","  True Positive Rate (TPR): 0.909\n","  True Negative Rate (TNR): 0.392\n","  AUC: 0.771\n","  C-index: 0.767\n","  1 - Questions proportion: 0.099\n","  1 - Single blood proportion: 0.565\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.682\n","  True Negative Rate (TNR): 0.724\n","  AUC: 0.760\n","  C-index: 0.756\n","  1 - Questions proportion: 0.027\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.770\n","  C-index: 0.765\n","  1 - Questions proportion: 0.072\n","  1 - Single blood proportion: 0.142\n","  1 - Risk scores proportion: 0.142\n","  True Positive Rate (TPR): 0.010\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.072\n","  1 - Single blood proportion: 0.142\n","  1 - Risk scores proportion: 0.142\n","  True Positive Rate (TPR): 0.294\n","  True Negative Rate (TNR): 0.944\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.014\n","  1 - Single blood proportion: 0.120\n","  1 - Risk scores proportion: 0.125\n","  True Positive Rate (TPR): 0.059\n","  True Negative Rate (TNR): 0.994\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.081\n","  1 - Single blood proportion: 0.185\n","  1 - Risk scores proportion: 0.187\n","  True Positive Rate (TPR): 0.476\n","  True Negative Rate (TNR): 0.866\n","  AUC: 0.774\n","  C-index: 0.768\n","  1 - Questions proportion: 0.074\n","  1 - Single blood proportion: 0.536\n","  1 - Risk scores proportion: 0.536\n","  True Positive Rate (TPR): 0.486\n","  True Negative Rate (TNR): 0.861\n","  AUC: 0.767\n","  C-index: 0.762\n","  1 - Questions proportion: 0.072\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.806\n","  True Negative Rate (TNR): 0.530\n","  AUC: 0.762\n","  C-index: 0.758\n","  1 - Questions proportion: 0.054\n","  1 - Single blood proportion: 0.127\n","  1 - Risk scores proportion: 0.127\n","  True Positive Rate (TPR): 0.204\n","  True Negative Rate (TNR): 0.968\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.086\n","  1 - Single blood proportion: 0.369\n","  1 - Risk scores proportion: 0.385\n","  True Positive Rate (TPR): 0.006\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.037\n","  1 - Single blood proportion: 0.065\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.043\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.008\n","  1 - Single blood proportion: 0.105\n","  1 - Risk scores proportion: 0.106\n","  True Positive Rate (TPR): 0.573\n","  True Negative Rate (TNR): 0.810\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.010\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.638\n","  True Negative Rate (TNR): 0.756\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.007\n","  1 - Single blood proportion: 0.012\n","  1 - Risk scores proportion: 0.396\n","  True Positive Rate (TPR): 0.036\n","  True Negative Rate (TNR): 0.997\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.069\n","  1 - Single blood proportion: 0.369\n","  1 - Risk scores proportion: 0.370\n","  True Positive Rate (TPR): 0.982\n","  True Negative Rate (TNR): 0.146\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.006\n","  1 - Single blood proportion: 0.035\n","  1 - Risk scores proportion: 0.038\n","  True Positive Rate (TPR): 0.036\n","  True Negative Rate (TNR): 0.996\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.025\n","  1 - Single blood proportion: 0.211\n","  1 - Risk scores proportion: 0.315\n","  True Positive Rate (TPR): 0.006\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.033\n","  1 - Single blood proportion: 0.173\n","  1 - Risk scores proportion: 0.292\n","  True Positive Rate (TPR): 0.024\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.088\n","  1 - Single blood proportion: 0.369\n","  1 - Risk scores proportion: 0.369\n","  True Positive Rate (TPR): 0.140\n","  True Negative Rate (TNR): 0.983\n","  AUC: 0.770\n","  C-index: 0.765\n","  1 - Questions proportion: 0.074\n","  1 - Single blood proportion: 0.168\n","  1 - Risk scores proportion: 0.169\n","  True Positive Rate (TPR): 0.091\n","  True Negative Rate (TNR): 0.991\n","  AUC: 0.773\n","  C-index: 0.768\n","  1 - Questions proportion: 0.075\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.340\n","  True Negative Rate (TNR): 0.928\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.020\n","  1 - Single blood proportion: 0.066\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.330\n","  True Negative Rate (TNR): 0.935\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.059\n","  1 - Single blood proportion: 0.101\n","  1 - Risk scores proportion: 0.216\n","  True Positive Rate (TPR): 0.020\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.008\n","  1 - Single blood proportion: 0.064\n","  1 - Risk scores proportion: 0.395\n","  True Positive Rate (TPR): 0.036\n","  True Negative Rate (TNR): 0.996\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.023\n","  1 - Single blood proportion: 0.062\n","  1 - Risk scores proportion: 0.096\n","  True Positive Rate (TPR): 0.014\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.115\n","  1 - Risk scores proportion: 0.243\n","  True Positive Rate (TPR): 0.996\n","  True Negative Rate (TNR): 0.055\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.030\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.660\n","  True Negative Rate (TNR): 0.741\n","  AUC: 0.768\n","  C-index: 0.764\n","  1 - Questions proportion: 0.012\n","  1 - Single blood proportion: 0.216\n","  1 - Risk scores proportion: 0.220\n","  True Positive Rate (TPR): 0.008\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.777\n","  C-index: 0.772\n","convergence value 0.1105978479403143\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:10\n","  1 - Questions proportion: 0.064\n","  1 - Single blood proportion: 0.429\n","  1 - Risk scores proportion: 0.436\n","  True Positive Rate (TPR): 0.045\n","  True Negative Rate (TNR): 0.996\n","  AUC: 0.773\n","  C-index: 0.769\n","  1 - Questions proportion: 0.041\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.753\n","  True Negative Rate (TNR): 0.625\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.099\n","  1 - Single blood proportion: 0.565\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.682\n","  True Negative Rate (TNR): 0.724\n","  AUC: 0.760\n","  C-index: 0.756\n","  1 - Questions proportion: 0.018\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.711\n","  True Negative Rate (TNR): 0.701\n","  AUC: 0.771\n","  C-index: 0.766\n","  1 - Questions proportion: 0.059\n","  1 - Single blood proportion: 0.101\n","  1 - Risk scores proportion: 0.101\n","  True Positive Rate (TPR): 0.018\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.016\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.769\n","  True Negative Rate (TNR): 0.609\n","  AUC: 0.771\n","  C-index: 0.767\n","  1 - Questions proportion: 0.059\n","  1 - Single blood proportion: 0.101\n","  1 - Risk scores proportion: 0.216\n","  True Positive Rate (TPR): 0.020\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.027\n","  1 - Single blood proportion: 0.561\n","  1 - Risk scores proportion: 0.561\n","  True Positive Rate (TPR): 0.868\n","  True Negative Rate (TNR): 0.457\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.045\n","  1 - Single blood proportion: 0.265\n","  1 - Risk scores proportion: 0.265\n","  True Positive Rate (TPR): 0.263\n","  True Negative Rate (TNR): 0.949\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.003\n","  1 - Single blood proportion: 0.020\n","  1 - Risk scores proportion: 0.402\n","  True Positive Rate (TPR): 0.024\n","  True Negative Rate (TNR): 0.998\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.062\n","  1 - Single blood proportion: 0.562\n","  1 - Risk scores proportion: 0.562\n","  True Positive Rate (TPR): 0.834\n","  True Negative Rate (TNR): 0.490\n","  AUC: 0.762\n","  C-index: 0.758\n","  1 - Questions proportion: 0.019\n","  1 - Single blood proportion: 0.051\n","  1 - Risk scores proportion: 0.051\n","  True Positive Rate (TPR): 0.109\n","  True Negative Rate (TNR): 0.988\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.077\n","  1 - Single blood proportion: 0.168\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.016\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.004\n","  1 - Single blood proportion: 0.008\n","  1 - Risk scores proportion: 0.410\n","  True Positive Rate (TPR): 0.024\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.340\n","  True Negative Rate (TNR): 0.928\n","  AUC: 0.772\n","  C-index: 0.768\n","  1 - Questions proportion: 0.088\n","  1 - Single blood proportion: 0.369\n","  1 - Risk scores proportion: 0.369\n","  True Positive Rate (TPR): 0.140\n","  True Negative Rate (TNR): 0.983\n","  AUC: 0.770\n","  C-index: 0.765\n","  1 - Questions proportion: 0.032\n","  1 - Single blood proportion: 0.475\n","  1 - Risk scores proportion: 0.475\n","  True Positive Rate (TPR): 0.980\n","  True Negative Rate (TNR): 0.172\n","  AUC: 0.771\n","  C-index: 0.766\n","  1 - Questions proportion: 0.028\n","  1 - Single blood proportion: 0.125\n","  1 - Risk scores proportion: 0.125\n","  True Positive Rate (TPR): 0.368\n","  True Negative Rate (TNR): 0.920\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.085\n","  1 - Single blood proportion: 0.186\n","  1 - Risk scores proportion: 0.188\n","  True Positive Rate (TPR): 0.476\n","  True Negative Rate (TNR): 0.866\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.065\n","  1 - Single blood proportion: 0.130\n","  1 - Risk scores proportion: 0.130\n","  True Positive Rate (TPR): 0.340\n","  True Negative Rate (TNR): 0.929\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.018\n","  1 - Single blood proportion: 0.246\n","  1 - Risk scores proportion: 0.246\n","  True Positive Rate (TPR): 0.466\n","  True Negative Rate (TNR): 0.872\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.028\n","  1 - Single blood proportion: 0.373\n","  1 - Risk scores proportion: 0.390\n","  True Positive Rate (TPR): 0.980\n","  True Negative Rate (TNR): 0.194\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.041\n","  1 - Single blood proportion: 0.511\n","  1 - Risk scores proportion: 0.511\n","  True Positive Rate (TPR): 0.632\n","  True Negative Rate (TNR): 0.766\n","  AUC: 0.770\n","  C-index: 0.766\n","  1 - Questions proportion: 0.035\n","  1 - Single blood proportion: 0.059\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.800\n","  True Negative Rate (TNR): 0.581\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.014\n","  1 - Single blood proportion: 0.023\n","  1 - Risk scores proportion: 0.257\n","  True Positive Rate (TPR): 0.992\n","  True Negative Rate (TNR): 0.130\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.011\n","  1 - Single blood proportion: 0.062\n","  1 - Risk scores proportion: 0.160\n","  True Positive Rate (TPR): 0.128\n","  True Negative Rate (TNR): 0.985\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.059\n","  1 - Single blood proportion: 0.101\n","  1 - Risk scores proportion: 0.216\n","  True Positive Rate (TPR): 0.018\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.775\n","  C-index: 0.770\n","  1 - Questions proportion: 0.043\n","  1 - Single blood proportion: 0.214\n","  1 - Risk scores proportion: 0.214\n","  True Positive Rate (TPR): 0.887\n","  True Negative Rate (TNR): 0.473\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.016\n","  1 - Single blood proportion: 0.211\n","  1 - Risk scores proportion: 0.315\n","  True Positive Rate (TPR): 0.012\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.072\n","  1 - Single blood proportion: 0.142\n","  1 - Risk scores proportion: 0.142\n","  True Positive Rate (TPR): 0.010\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.774\n","  C-index: 0.769\n","convergence value 0.10672153898264158\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:11\n","  1 - Questions proportion: 0.099\n","  1 - Single blood proportion: 0.536\n","  1 - Risk scores proportion: 0.536\n","  True Positive Rate (TPR): 0.109\n","  True Negative Rate (TNR): 0.988\n","  AUC: 0.760\n","  C-index: 0.756\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.016\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.772\n","  C-index: 0.768\n","  1 - Questions proportion: 0.099\n","  1 - Single blood proportion: 0.565\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.682\n","  True Negative Rate (TNR): 0.724\n","  AUC: 0.760\n","  C-index: 0.756\n","  1 - Questions proportion: 0.076\n","  1 - Single blood proportion: 0.562\n","  1 - Risk scores proportion: 0.567\n","  True Positive Rate (TPR): 0.986\n","  True Negative Rate (TNR): 0.121\n","  AUC: 0.762\n","  C-index: 0.758\n","  1 - Questions proportion: 0.010\n","  1 - Single blood proportion: 0.268\n","  1 - Risk scores proportion: 0.268\n","  True Positive Rate (TPR): 0.800\n","  True Negative Rate (TNR): 0.572\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.028\n","  1 - Single blood proportion: 0.125\n","  1 - Risk scores proportion: 0.125\n","  True Positive Rate (TPR): 0.953\n","  True Negative Rate (TNR): 0.295\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.007\n","  1 - Single blood proportion: 0.053\n","  1 - Risk scores proportion: 0.065\n","  True Positive Rate (TPR): 0.931\n","  True Negative Rate (TNR): 0.375\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.479\n","  1 - Risk scores proportion: 0.479\n","  True Positive Rate (TPR): 0.344\n","  True Negative Rate (TNR): 0.927\n","  AUC: 0.773\n","  C-index: 0.769\n","  1 - Questions proportion: 0.048\n","  1 - Single blood proportion: 0.433\n","  1 - Risk scores proportion: 0.438\n","  True Positive Rate (TPR): 0.312\n","  True Negative Rate (TNR): 0.938\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.182\n","  1 - Risk scores proportion: 0.197\n","  True Positive Rate (TPR): 0.360\n","  True Negative Rate (TNR): 0.922\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.024\n","  1 - Single blood proportion: 0.328\n","  1 - Risk scores proportion: 0.351\n","  True Positive Rate (TPR): 0.233\n","  True Negative Rate (TNR): 0.957\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.044\n","  1 - Single blood proportion: 0.077\n","  1 - Risk scores proportion: 0.078\n","  True Positive Rate (TPR): 0.020\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.014\n","  1 - Single blood proportion: 0.181\n","  1 - Risk scores proportion: 0.304\n","  True Positive Rate (TPR): 0.012\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.007\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.648\n","  True Negative Rate (TNR): 0.752\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.093\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.140\n","  True Negative Rate (TNR): 0.983\n","  AUC: 0.758\n","  C-index: 0.755\n","  1 - Questions proportion: 0.085\n","  1 - Single blood proportion: 0.186\n","  1 - Risk scores proportion: 0.188\n","  True Positive Rate (TPR): 0.476\n","  True Negative Rate (TNR): 0.866\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.085\n","  1 - Single blood proportion: 0.157\n","  1 - Risk scores proportion: 0.161\n","  True Positive Rate (TPR): 0.476\n","  True Negative Rate (TNR): 0.866\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.117\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.081\n","  True Negative Rate (TNR): 0.993\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.027\n","  1 - Single blood proportion: 0.325\n","  1 - Risk scores proportion: 0.343\n","  True Positive Rate (TPR): 0.980\n","  True Negative Rate (TNR): 0.198\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.048\n","  1 - Single blood proportion: 0.432\n","  1 - Risk scores proportion: 0.432\n","  True Positive Rate (TPR): 0.409\n","  True Negative Rate (TNR): 0.898\n","  AUC: 0.769\n","  C-index: 0.764\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.116\n","  1 - Risk scores proportion: 0.116\n","  True Positive Rate (TPR): 0.389\n","  True Negative Rate (TNR): 0.909\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.075\n","  1 - Single blood proportion: 0.565\n","  1 - Risk scores proportion: 0.565\n","  True Positive Rate (TPR): 0.630\n","  True Negative Rate (TNR): 0.769\n","  AUC: 0.770\n","  C-index: 0.765\n","  1 - Questions proportion: 0.074\n","  1 - Single blood proportion: 0.511\n","  1 - Risk scores proportion: 0.530\n","  True Positive Rate (TPR): 0.692\n","  True Negative Rate (TNR): 0.717\n","  AUC: 0.760\n","  C-index: 0.756\n","  1 - Questions proportion: 0.017\n","  1 - Single blood proportion: 0.526\n","  1 - Risk scores proportion: 0.526\n","  True Positive Rate (TPR): 0.012\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.032\n","  1 - Single blood proportion: 0.475\n","  1 - Risk scores proportion: 0.475\n","  True Positive Rate (TPR): 0.980\n","  True Negative Rate (TNR): 0.172\n","  AUC: 0.771\n","  C-index: 0.766\n","  1 - Questions proportion: 0.006\n","  1 - Single blood proportion: 0.129\n","  1 - Risk scores proportion: 0.306\n","  True Positive Rate (TPR): 0.012\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.044\n","  1 - Single blood proportion: 0.331\n","  1 - Risk scores proportion: 0.435\n","  True Positive Rate (TPR): 0.034\n","  True Negative Rate (TNR): 0.996\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.022\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.899\n","  True Negative Rate (TNR): 0.413\n","  AUC: 0.772\n","  C-index: 0.768\n","  1 - Questions proportion: 0.009\n","  1 - Single blood proportion: 0.407\n","  1 - Risk scores proportion: 0.407\n","  True Positive Rate (TPR): 0.800\n","  True Negative Rate (TNR): 0.578\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.077\n","  1 - Single blood proportion: 0.168\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.016\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.772\n","  C-index: 0.767\n","convergence value 0.09275704141700547\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:12\n","  1 - Questions proportion: 0.102\n","  1 - Single blood proportion: 0.565\n","  1 - Risk scores proportion: 0.574\n","  True Positive Rate (TPR): 0.016\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.760\n","  C-index: 0.756\n","  1 - Questions proportion: 0.099\n","  1 - Single blood proportion: 0.565\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.682\n","  True Negative Rate (TNR): 0.724\n","  AUC: 0.760\n","  C-index: 0.756\n","  1 - Questions proportion: 0.075\n","  1 - Single blood proportion: 0.565\n","  1 - Risk scores proportion: 0.565\n","  True Positive Rate (TPR): 0.630\n","  True Negative Rate (TNR): 0.769\n","  AUC: 0.770\n","  C-index: 0.765\n","  1 - Questions proportion: 0.053\n","  1 - Single blood proportion: 0.567\n","  1 - Risk scores proportion: 0.567\n","  True Positive Rate (TPR): 0.630\n","  True Negative Rate (TNR): 0.770\n","  AUC: 0.772\n","  C-index: 0.768\n","  1 - Questions proportion: 0.073\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.566\n","  True Positive Rate (TPR): 0.682\n","  True Negative Rate (TNR): 0.725\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.079\n","  1 - Single blood proportion: 0.158\n","  1 - Risk scores proportion: 0.219\n","  True Positive Rate (TPR): 0.083\n","  True Negative Rate (TNR): 0.992\n","  AUC: 0.774\n","  C-index: 0.768\n","  1 - Questions proportion: 0.078\n","  1 - Single blood proportion: 0.159\n","  1 - Risk scores proportion: 0.233\n","  True Positive Rate (TPR): 0.014\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.774\n","  C-index: 0.768\n","  1 - Questions proportion: 0.077\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.140\n","  True Negative Rate (TNR): 0.983\n","  AUC: 0.758\n","  C-index: 0.754\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.182\n","  1 - Risk scores proportion: 0.197\n","  True Positive Rate (TPR): 0.008\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.776\n","  C-index: 0.770\n","  1 - Questions proportion: 0.048\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.572\n","  True Positive Rate (TPR): 0.312\n","  True Negative Rate (TNR): 0.938\n","  AUC: 0.771\n","  C-index: 0.766\n","  1 - Questions proportion: 0.071\n","  1 - Single blood proportion: 0.562\n","  1 - Risk scores proportion: 0.567\n","  True Positive Rate (TPR): 0.986\n","  True Negative Rate (TNR): 0.120\n","  AUC: 0.762\n","  C-index: 0.758\n","  1 - Questions proportion: 0.011\n","  1 - Single blood proportion: 0.162\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.648\n","  True Negative Rate (TNR): 0.752\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.005\n","  1 - Single blood proportion: 0.116\n","  1 - Risk scores proportion: 0.116\n","  True Positive Rate (TPR): 0.389\n","  True Negative Rate (TNR): 0.909\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.016\n","  1 - Single blood proportion: 0.129\n","  1 - Risk scores proportion: 0.306\n","  True Positive Rate (TPR): 0.014\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.001\n","  1 - Single blood proportion: 0.562\n","  1 - Risk scores proportion: 0.562\n","  True Positive Rate (TPR): 0.016\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.772\n","  C-index: 0.768\n","  1 - Questions proportion: 0.022\n","  1 - Single blood proportion: 0.061\n","  1 - Risk scores proportion: 0.281\n","  True Positive Rate (TPR): 0.233\n","  True Negative Rate (TNR): 0.957\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.018\n","  1 - Single blood proportion: 0.561\n","  1 - Risk scores proportion: 0.561\n","  True Positive Rate (TPR): 0.403\n","  True Negative Rate (TNR): 0.902\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.085\n","  1 - Single blood proportion: 0.325\n","  1 - Risk scores proportion: 0.343\n","  True Positive Rate (TPR): 0.988\n","  True Negative Rate (TNR): 0.061\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.080\n","  1 - Single blood proportion: 0.249\n","  1 - Risk scores proportion: 0.249\n","  True Positive Rate (TPR): 0.109\n","  True Negative Rate (TNR): 0.988\n","  AUC: 0.773\n","  C-index: 0.768\n","  1 - Questions proportion: 0.009\n","  1 - Single blood proportion: 0.475\n","  1 - Risk scores proportion: 0.475\n","  True Positive Rate (TPR): 0.978\n","  True Negative Rate (TNR): 0.224\n","  AUC: 0.773\n","  C-index: 0.769\n","  1 - Questions proportion: 0.085\n","  1 - Single blood proportion: 0.157\n","  1 - Risk scores proportion: 0.161\n","  True Positive Rate (TPR): 0.476\n","  True Negative Rate (TNR): 0.866\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.074\n","  1 - Single blood proportion: 0.475\n","  1 - Risk scores proportion: 0.475\n","  True Positive Rate (TPR): 0.986\n","  True Negative Rate (TNR): 0.078\n","  AUC: 0.761\n","  C-index: 0.758\n","  1 - Questions proportion: 0.096\n","  1 - Single blood proportion: 0.489\n","  1 - Risk scores proportion: 0.500\n","  True Positive Rate (TPR): 0.253\n","  True Negative Rate (TNR): 0.952\n","  AUC: 0.762\n","  C-index: 0.758\n","  1 - Questions proportion: 0.033\n","  1 - Single blood proportion: 0.073\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.016\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.085\n","  1 - Single blood proportion: 0.186\n","  1 - Risk scores proportion: 0.188\n","  True Positive Rate (TPR): 0.476\n","  True Negative Rate (TNR): 0.866\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.011\n","  1 - Risk scores proportion: 0.416\n","  True Positive Rate (TPR): 0.650\n","  True Negative Rate (TNR): 0.749\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.007\n","  1 - Single blood proportion: 0.069\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.008\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.075\n","  1 - Single blood proportion: 0.143\n","  1 - Risk scores proportion: 0.379\n","  True Positive Rate (TPR): 0.996\n","  True Negative Rate (TNR): 0.037\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.068\n","  1 - Single blood proportion: 0.562\n","  1 - Risk scores proportion: 0.567\n","  True Positive Rate (TPR): 0.986\n","  True Negative Rate (TNR): 0.121\n","  AUC: 0.762\n","  C-index: 0.758\n","  1 - Questions proportion: 0.031\n","  1 - Single blood proportion: 0.360\n","  1 - Risk scores proportion: 0.444\n","  True Positive Rate (TPR): 0.512\n","  True Negative Rate (TNR): 0.844\n","  AUC: 0.774\n","  C-index: 0.769\n","convergence value 0.08876977697375676\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:13\n","  1 - Questions proportion: 0.099\n","  1 - Single blood proportion: 0.565\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.682\n","  True Negative Rate (TNR): 0.724\n","  AUC: 0.760\n","  C-index: 0.756\n","  1 - Questions proportion: 0.094\n","  1 - Single blood proportion: 0.461\n","  1 - Risk scores proportion: 0.461\n","  True Positive Rate (TPR): 0.559\n","  True Negative Rate (TNR): 0.827\n","  AUC: 0.764\n","  C-index: 0.761\n","  1 - Questions proportion: 0.077\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.154\n","  True Negative Rate (TNR): 0.981\n","  AUC: 0.758\n","  C-index: 0.754\n","  1 - Questions proportion: 0.089\n","  1 - Single blood proportion: 0.302\n","  1 - Risk scores proportion: 0.302\n","  True Positive Rate (TPR): 0.182\n","  True Negative Rate (TNR): 0.974\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.020\n","  1 - Single blood proportion: 0.342\n","  1 - Risk scores proportion: 0.342\n","  True Positive Rate (TPR): 0.846\n","  True Negative Rate (TNR): 0.535\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.083\n","  1 - Single blood proportion: 0.147\n","  1 - Risk scores proportion: 0.151\n","  True Positive Rate (TPR): 0.451\n","  True Negative Rate (TNR): 0.882\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.028\n","  1 - Single blood proportion: 0.171\n","  1 - Risk scores proportion: 0.177\n","  True Positive Rate (TPR): 0.951\n","  True Negative Rate (TNR): 0.298\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.034\n","  1 - Single blood proportion: 0.489\n","  1 - Risk scores proportion: 0.500\n","  True Positive Rate (TPR): 0.253\n","  True Negative Rate (TNR): 0.952\n","  AUC: 0.773\n","  C-index: 0.768\n","  1 - Questions proportion: 0.005\n","  1 - Single blood proportion: 0.116\n","  1 - Risk scores proportion: 0.116\n","  True Positive Rate (TPR): 0.389\n","  True Negative Rate (TNR): 0.909\n","  AUC: 0.778\n","  C-index: 0.773\n","  1 - Questions proportion: 0.059\n","  1 - Single blood proportion: 0.107\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.008\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.774\n","  C-index: 0.769\n","  1 - Questions proportion: 0.017\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.978\n","  True Negative Rate (TNR): 0.225\n","  AUC: 0.772\n","  C-index: 0.768\n","  1 - Questions proportion: 0.085\n","  1 - Single blood proportion: 0.187\n","  1 - Risk scores proportion: 0.201\n","  True Positive Rate (TPR): 0.016\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.072\n","  1 - Single blood proportion: 0.475\n","  1 - Risk scores proportion: 0.485\n","  True Positive Rate (TPR): 0.698\n","  True Negative Rate (TNR): 0.714\n","  AUC: 0.762\n","  C-index: 0.758\n","  1 - Questions proportion: 0.072\n","  1 - Single blood proportion: 0.142\n","  1 - Risk scores proportion: 0.285\n","  True Positive Rate (TPR): 0.233\n","  True Negative Rate (TNR): 0.957\n","  AUC: 0.773\n","  C-index: 0.768\n","  1 - Questions proportion: 0.008\n","  1 - Single blood proportion: 0.561\n","  1 - Risk scores proportion: 0.561\n","  True Positive Rate (TPR): 0.036\n","  True Negative Rate (TNR): 0.998\n","  AUC: 0.772\n","  C-index: 0.767\n","  1 - Questions proportion: 0.055\n","  1 - Single blood proportion: 0.547\n","  1 - Risk scores proportion: 0.547\n","  True Positive Rate (TPR): 0.330\n","  True Negative Rate (TNR): 0.935\n","  AUC: 0.766\n","  C-index: 0.762\n","  1 - Questions proportion: 0.085\n","  1 - Single blood proportion: 0.157\n","  1 - Risk scores proportion: 0.157\n","  True Positive Rate (TPR): 0.476\n","  True Negative Rate (TNR): 0.866\n","  AUC: 0.775\n","  C-index: 0.769\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.216\n","  1 - Risk scores proportion: 0.216\n","  True Positive Rate (TPR): 0.032\n","  True Negative Rate (TNR): 0.998\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.005\n","  1 - Single blood proportion: 0.012\n","  1 - Risk scores proportion: 0.098\n","  True Positive Rate (TPR): 0.648\n","  True Negative Rate (TNR): 0.752\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.055\n","  1 - Single blood proportion: 0.108\n","  1 - Risk scores proportion: 0.108\n","  True Positive Rate (TPR): 0.966\n","  True Negative Rate (TNR): 0.226\n","  AUC: 0.776\n","  C-index: 0.771\n","  1 - Questions proportion: 0.002\n","  1 - Single blood proportion: 0.023\n","  1 - Risk scores proportion: 0.140\n","  True Positive Rate (TPR): 0.121\n","  True Negative Rate (TNR): 0.985\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.000\n","  1 - Single blood proportion: 0.011\n","  1 - Risk scores proportion: 0.416\n","  True Positive Rate (TPR): 0.650\n","  True Negative Rate (TNR): 0.749\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.015\n","  1 - Single blood proportion: 0.129\n","  1 - Risk scores proportion: 0.306\n","  True Positive Rate (TPR): 0.014\n","  True Negative Rate (TNR): 0.999\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.010\n","  1 - Single blood proportion: 0.287\n","  1 - Risk scores proportion: 0.287\n","  True Positive Rate (TPR): 0.498\n","  True Negative Rate (TNR): 0.852\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.074\n","  1 - Single blood proportion: 0.563\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.986\n","  True Negative Rate (TNR): 0.078\n","  AUC: 0.759\n","  C-index: 0.755\n","  1 - Questions proportion: 0.003\n","  1 - Single blood proportion: 0.562\n","  1 - Risk scores proportion: 0.562\n","  True Positive Rate (TPR): 0.073\n","  True Negative Rate (TNR): 0.993\n","  AUC: 0.771\n","  C-index: 0.767\n","  1 - Questions proportion: 0.013\n","  1 - Single blood proportion: 0.311\n","  1 - Risk scores proportion: 0.311\n","  True Positive Rate (TPR): 0.310\n","  True Negative Rate (TNR): 0.939\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.043\n","  1 - Single blood proportion: 0.536\n","  1 - Risk scores proportion: 0.549\n","  True Positive Rate (TPR): 0.779\n","  True Negative Rate (TNR): 0.585\n","  AUC: 0.771\n","  C-index: 0.766\n","  1 - Questions proportion: 0.019\n","  1 - Single blood proportion: 0.162\n","  1 - Risk scores proportion: 0.395\n","  True Positive Rate (TPR): 0.917\n","  True Negative Rate (TNR): 0.419\n","  AUC: 0.778\n","  C-index: 0.772\n","  1 - Questions proportion: 0.034\n","  1 - Single blood proportion: 0.360\n","  1 - Risk scores proportion: 0.444\n","  True Positive Rate (TPR): 0.512\n","  True Negative Rate (TNR): 0.844\n","  AUC: 0.774\n","  C-index: 0.769\n","convergence value 0.09658381647580011\n","Pareto stopIndx 7\n","Parento front computation\n","Pareto Front Calculation Completed\n","Starting Fitness Tuning As Per Pareto Frontiers\n","Fitness Tuning As Per Pareto Frontiers Completed\n","Roulette Wheel Selection Completed\n","Crossover completed\n","Doing Adaptive Mutation\n","\n","Generation:14\n","  1 - Questions proportion: 0.099\n","  1 - Single blood proportion: 0.565\n","  1 - Risk scores proportion: 0.568\n","  True Positive Rate (TPR): 0.682\n","  True Negative Rate (TNR): 0.724\n","  AUC: 0.760\n","  C-index: 0.756\n","  1 - Questions proportion: 0.002\n","  1 - Single blood proportion: 0.217\n","  1 - Risk scores proportion: 0.217\n","  True Positive Rate (TPR): 0.251\n","  True Negative Rate (TNR): 0.952\n","  AUC: 0.777\n","  C-index: 0.771\n","  1 - Questions proportion: 0.082\n","  1 - Single blood proportion: 0.562\n","  1 - Risk scores proportion: 0.563\n","  True Positive Rate (TPR): 0.073\n","  True Negative Rate (TNR): 0.993\n","  AUC: 0.755\n","  C-index: 0.752\n","  1 - Questions proportion: 0.007\n","  1 - Single blood proportion: 0.069\n","  1 - Risk scores proportion: 0.418\n","  True Positive Rate (TPR): 0.008\n","  True Negative Rate (TNR): 1.000\n","  AUC: 0.777\n","  C-index: 0.772\n","  1 - Questions proportion: 0.089\n","  1 - Single blood proportion: 0.342\n","  1 - Risk scores proportion: 0.342\n","  True Positive Rate (TPR): 0.836\n","  True Negative Rate (TNR): 0.539\n","  AUC: 0.771\n","  C-index: 0.766\n","  1 - Questions proportion: 0.066\n","  1 - Single blood proportion: 0.198\n","  1 - Risk scores proportion: 0.198\n","  True Positive Rate (TPR): 0.573\n","  True Negative Rate (TNR): 0.809\n","  AUC: 0.775\n","  C-index: 0.770\n"]}],"source":["#G-100 pp 30\n","varDict={\"-g\": 100, \"-pp\":30}\n","mainFunction(varDict) # pass the dictionary of arguments to main\n","\n","# run the code using command below\n","# python multi_objective_optimization.py -g 5 -p 10"]},{"cell_type":"code","execution_count":30,"metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop_qrisk.pkl\n","MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_oneDrop_qrisk.pkl\n","MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_oneDrop_qrisk.pkl\n","MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_oneDrop_qrisk.pkl\n","MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_Sex-specific_factors_oneDrop_qrisk.pkl\n","MRMR_COX_FULL_MODEL_+QRISKSociodemographics_Health_and_medical_history_oneDrop_qrisk.pkl\n","MRMR_COX_FULL_MODEL_+QRISKSociodemographics_oneDrop_qrisk.pkl\n","MRMR_COX_Sociodemographics.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_Psychosocial_factors_oneDrop.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_Lifestyle_and_environment_oneDrop.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_Family_history_oneDrop.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_Early_life_factors_oneDrop.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history_Sex-specific_factors_oneDrop.pkl\n","MRMR_COX_Sociodemographics_Health_and_medical_history_oneDrop.pkl\n","MRMR_COX_Sociodemographics_oneDrop.pkl\n","Output_FeatSelect\n","Output_Model\n","Output_PreProcess\n","dfTest.csv\n","dfTrain.csv\n"]}],"source":["!ls"]},{"cell_type":"code","execution_count":31,"metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["2025_06_02  2025_06_02testSetGoalVals.csv\n"]}],"source":["!ls Output_FeatSelect"]},{"cell_type":"code","execution_count":32,"metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["'Average Performance Per Generation.csv'\n"," FinalSolFront1.csv\n"," FinalSolutionsAll.csv\n","'Max Solution Performance Per Generation.csv'\n"," goals_FinalFront1.csv\n"]}],"source":["!ls Output_FeatSelect/2025_06_02"]},{"cell_type":"code","execution_count":34,"metadata":{"tags":[],"trusted":true},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Min_Threshold1</th>\n","      <th>Max_Threshold1</th>\n","      <th>Min_Threshold2</th>\n","      <th>Max_Threshold2</th>\n","      <th>Min_Threshold3</th>\n","      <th>Max_Threshold3</th>\n","      <th>Min_Threshold4</th>\n","      <th>Max_Threshold4</th>\n","      <th>Min_Threshold5</th>\n","      <th>Max_Threshold5</th>\n","      <th>Min_Threshold6</th>\n","      <th>Max_Threshold6</th>\n","      <th>Min_Threshold7</th>\n","      <th>Max_Threshold7</th>\n","      <th>Min_Threshold8</th>\n","      <th>Classification_Threshold</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>-2.484193</td>\n","      <td>5.000000</td>\n","      <td>-0.947774</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>6.276615</td>\n","      <td>1.000000</td>\n","      <td>5.620068</td>\n","      <td>1.000000</td>\n","      <td>8.681749</td>\n","      <td>1.000000</td>\n","      <td>8.748657</td>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>1.792720</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>-1.563720</td>\n","      <td>6.944407</td>\n","      <td>-1.888616</td>\n","      <td>8.209462</td>\n","      <td>0.864605</td>\n","      <td>5.762497</td>\n","      <td>-0.211366</td>\n","      <td>5.931304</td>\n","      <td>-1.340304</td>\n","      <td>5.000000</td>\n","      <td>-0.209072</td>\n","      <td>7.760002</td>\n","      <td>-2.500000</td>\n","      <td>5.000000</td>\n","      <td>0.022641</td>\n","      <td>3.074763</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>-2.500000</td>\n","      <td>7.055593</td>\n","      <td>-0.153862</td>\n","      <td>6.418866</td>\n","      <td>0.454577</td>\n","      <td>5.000000</td>\n","      <td>-1.251862</td>\n","      <td>5.000000</td>\n","      <td>0.040180</td>\n","      <td>6.080559</td>\n","      <td>-0.229976</td>\n","      <td>7.746080</td>\n","      <td>-1.302351</td>\n","      <td>5.637896</td>\n","      <td>0.485388</td>\n","      <td>3.112052</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>-0.799507</td>\n","      <td>7.957415</td>\n","      <td>-2.500000</td>\n","      <td>8.798075</td>\n","      <td>-1.517770</td>\n","      <td>6.145288</td>\n","      <td>0.116261</td>\n","      <td>8.241050</td>\n","      <td>-1.815650</td>\n","      <td>9.000000</td>\n","      <td>-0.417373</td>\n","      <td>5.157499</td>\n","      <td>-0.517976</td>\n","      <td>5.222527</td>\n","      <td>1.000000</td>\n","      <td>1.638921</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>-0.703108</td>\n","      <td>8.402900</td>\n","      <td>0.573783</td>\n","      <td>5.063458</td>\n","      <td>-0.895456</td>\n","      <td>5.655919</td>\n","      <td>1.000000</td>\n","      <td>9.000000</td>\n","      <td>-1.305070</td>\n","      <td>7.508770</td>\n","      <td>0.454413</td>\n","      <td>7.108318</td>\n","      <td>-2.500000</td>\n","      <td>5.127444</td>\n","      <td>-0.853306</td>\n","      <td>2.547014</td>\n","    </tr>\n","    <tr>\n","      <th>5</th>\n","      <td>1.000000</td>\n","      <td>6.624750</td>\n","      <td>-0.988286</td>\n","      <td>8.351893</td>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>0.012210</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>8.375118</td>\n","      <td>-0.068469</td>\n","      <td>9.000000</td>\n","      <td>-1.533015</td>\n","      <td>9.000000</td>\n","      <td>1.000000</td>\n","      <td>4.014643</td>\n","    </tr>\n","    <tr>\n","      <th>6</th>\n","      <td>-0.231252</td>\n","      <td>5.000000</td>\n","      <td>-2.500000</td>\n","      <td>7.556734</td>\n","      <td>-1.552548</td>\n","      <td>8.048886</td>\n","      <td>-1.412026</td>\n","      <td>9.000000</td>\n","      <td>-0.555866</td>\n","      <td>8.464159</td>\n","      <td>-1.252060</td>\n","      <td>5.605640</td>\n","      <td>-1.408749</td>\n","      <td>8.784983</td>\n","      <td>1.000000</td>\n","      <td>0.230816</td>\n","    </tr>\n","    <tr>\n","      <th>7</th>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>-0.947774</td>\n","      <td>5.000000</td>\n","      <td>0.388595</td>\n","      <td>5.000000</td>\n","      <td>0.568337</td>\n","      <td>9.000000</td>\n","      <td>1.000000</td>\n","      <td>7.277051</td>\n","      <td>-2.500000</td>\n","      <td>7.457110</td>\n","      <td>-2.173024</td>\n","      <td>5.309422</td>\n","      <td>-1.433755</td>\n","      <td>0.000000</td>\n","    </tr>\n","    <tr>\n","      <th>8</th>\n","      <td>-2.484193</td>\n","      <td>5.923599</td>\n","      <td>-2.500000</td>\n","      <td>9.000000</td>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>5.620068</td>\n","      <td>1.000000</td>\n","      <td>8.681749</td>\n","      <td>1.000000</td>\n","      <td>8.748657</td>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>0.000000</td>\n","    </tr>\n","    <tr>\n","      <th>9</th>\n","      <td>0.921640</td>\n","      <td>8.612685</td>\n","      <td>-1.406337</td>\n","      <td>8.929677</td>\n","      <td>1.000000</td>\n","      <td>8.649072</td>\n","      <td>0.731134</td>\n","      <td>8.850317</td>\n","      <td>-2.500000</td>\n","      <td>5.000000</td>\n","      <td>-1.154680</td>\n","      <td>8.576633</td>\n","      <td>0.885723</td>\n","      <td>5.355557</td>\n","      <td>1.000000</td>\n","      <td>2.341229</td>\n","    </tr>\n","    <tr>\n","      <th>10</th>\n","      <td>1.000000</td>\n","      <td>5.581058</td>\n","      <td>0.100928</td>\n","      <td>8.173586</td>\n","      <td>0.751805</td>\n","      <td>6.467111</td>\n","      <td>0.321589</td>\n","      <td>7.635728</td>\n","      <td>1.000000</td>\n","      <td>6.302015</td>\n","      <td>0.014918</td>\n","      <td>9.000000</td>\n","      <td>-1.816842</td>\n","      <td>7.375160</td>\n","      <td>-1.206248</td>\n","      <td>0.000000</td>\n","    </tr>\n","    <tr>\n","      <th>11</th>\n","      <td>0.919816</td>\n","      <td>9.000000</td>\n","      <td>-2.500000</td>\n","      <td>7.462543</td>\n","      <td>-1.712006</td>\n","      <td>9.000000</td>\n","      <td>-1.809982</td>\n","      <td>8.499060</td>\n","      <td>-1.682064</td>\n","      <td>5.000000</td>\n","      <td>-1.281227</td>\n","      <td>7.699881</td>\n","      <td>-2.111316</td>\n","      <td>5.412747</td>\n","      <td>0.588788</td>\n","      <td>4.774818</td>\n","    </tr>\n","    <tr>\n","      <th>12</th>\n","      <td>0.555150</td>\n","      <td>6.058563</td>\n","      <td>-1.783375</td>\n","      <td>7.013663</td>\n","      <td>-0.889651</td>\n","      <td>5.000000</td>\n","      <td>0.082096</td>\n","      <td>9.000000</td>\n","      <td>-2.108757</td>\n","      <td>6.558063</td>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>5.438562</td>\n","      <td>0.984333</td>\n","      <td>1.596120</td>\n","    </tr>\n","    <tr>\n","      <th>13</th>\n","      <td>-2.500000</td>\n","      <td>6.688769</td>\n","      <td>-1.346836</td>\n","      <td>6.546752</td>\n","      <td>-0.473439</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>7.554996</td>\n","      <td>-1.188209</td>\n","      <td>9.000000</td>\n","      <td>-2.289905</td>\n","      <td>6.641053</td>\n","      <td>-1.875815</td>\n","      <td>7.239179</td>\n","      <td>-1.498051</td>\n","      <td>1.343219</td>\n","    </tr>\n","    <tr>\n","      <th>14</th>\n","      <td>-1.739023</td>\n","      <td>5.000000</td>\n","      <td>-2.500000</td>\n","      <td>8.934220</td>\n","      <td>1.000000</td>\n","      <td>8.995979</td>\n","      <td>-1.757982</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>7.577163</td>\n","      <td>-1.206961</td>\n","      <td>5.236958</td>\n","      <td>-0.941809</td>\n","      <td>9.000000</td>\n","      <td>1.000000</td>\n","      <td>0.000000</td>\n","    </tr>\n","    <tr>\n","      <th>15</th>\n","      <td>-0.193415</td>\n","      <td>5.000000</td>\n","      <td>-1.763051</td>\n","      <td>8.911601</td>\n","      <td>1.000000</td>\n","      <td>7.707022</td>\n","      <td>-0.475890</td>\n","      <td>5.000000</td>\n","      <td>-0.408703</td>\n","      <td>7.566323</td>\n","      <td>1.000000</td>\n","      <td>9.000000</td>\n","      <td>-0.634160</td>\n","      <td>5.000000</td>\n","      <td>0.582720</td>\n","      <td>2.109472</td>\n","    </tr>\n","    <tr>\n","      <th>16</th>\n","      <td>-0.868076</td>\n","      <td>6.899075</td>\n","      <td>-2.357094</td>\n","      <td>6.899075</td>\n","      <td>-0.287576</td>\n","      <td>7.569593</td>\n","      <td>-0.334092</td>\n","      <td>6.986923</td>\n","      <td>-0.273360</td>\n","      <td>7.673489</td>\n","      <td>1.000000</td>\n","      <td>7.524166</td>\n","      <td>1.000000</td>\n","      <td>5.195959</td>\n","      <td>0.804769</td>\n","      <td>0.000000</td>\n","    </tr>\n","    <tr>\n","      <th>17</th>\n","      <td>-2.337470</td>\n","      <td>6.845520</td>\n","      <td>-1.321150</td>\n","      <td>7.812926</td>\n","      <td>0.817139</td>\n","      <td>6.639562</td>\n","      <td>0.282562</td>\n","      <td>7.477598</td>\n","      <td>-1.142909</td>\n","      <td>8.961613</td>\n","      <td>-2.215556</td>\n","      <td>6.188587</td>\n","      <td>-0.975260</td>\n","      <td>5.272863</td>\n","      <td>0.955138</td>\n","      <td>2.324567</td>\n","    </tr>\n","    <tr>\n","      <th>18</th>\n","      <td>-2.500000</td>\n","      <td>9.000000</td>\n","      <td>-1.013972</td>\n","      <td>8.222575</td>\n","      <td>1.000000</td>\n","      <td>7.940133</td>\n","      <td>-1.281260</td>\n","      <td>8.915542</td>\n","      <td>1.000000</td>\n","      <td>8.502546</td>\n","      <td>-1.326409</td>\n","      <td>8.398827</td>\n","      <td>-1.375298</td>\n","      <td>8.513964</td>\n","      <td>0.873299</td>\n","      <td>0.310507</td>\n","    </tr>\n","    <tr>\n","      <th>19</th>\n","      <td>-1.438735</td>\n","      <td>8.734394</td>\n","      <td>-2.500000</td>\n","      <td>8.854252</td>\n","      <td>1.000000</td>\n","      <td>7.128696</td>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>9.000000</td>\n","      <td>1.000000</td>\n","      <td>7.969859</td>\n","      <td>1.000000</td>\n","      <td>9.000000</td>\n","      <td>1.000000</td>\n","      <td>4.070031</td>\n","    </tr>\n","    <tr>\n","      <th>20</th>\n","      <td>-2.500000</td>\n","      <td>6.979361</td>\n","      <td>0.358294</td>\n","      <td>5.867188</td>\n","      <td>-2.325128</td>\n","      <td>5.000000</td>\n","      <td>-1.553361</td>\n","      <td>9.000000</td>\n","      <td>0.687667</td>\n","      <td>5.957286</td>\n","      <td>0.156503</td>\n","      <td>8.179476</td>\n","      <td>-1.060701</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","    </tr>\n","    <tr>\n","      <th>21</th>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>-1.614255</td>\n","      <td>7.443017</td>\n","      <td>1.000000</td>\n","      <td>7.216845</td>\n","      <td>0.083578</td>\n","      <td>5.000000</td>\n","      <td>-0.136759</td>\n","      <td>5.771946</td>\n","      <td>-1.977589</td>\n","      <td>6.975859</td>\n","      <td>-0.738587</td>\n","      <td>6.145567</td>\n","      <td>1.000000</td>\n","      <td>2.393012</td>\n","    </tr>\n","    <tr>\n","      <th>22</th>\n","      <td>1.000000</td>\n","      <td>6.269356</td>\n","      <td>-1.777364</td>\n","      <td>8.872206</td>\n","      <td>0.249281</td>\n","      <td>7.010816</td>\n","      <td>0.013728</td>\n","      <td>5.748915</td>\n","      <td>-2.306030</td>\n","      <td>5.063360</td>\n","      <td>0.096497</td>\n","      <td>5.275726</td>\n","      <td>-0.361210</td>\n","      <td>6.258981</td>\n","      <td>1.000000</td>\n","      <td>2.812965</td>\n","    </tr>\n","    <tr>\n","      <th>23</th>\n","      <td>-2.500000</td>\n","      <td>6.945668</td>\n","      <td>-2.500000</td>\n","      <td>5.856597</td>\n","      <td>1.000000</td>\n","      <td>6.253239</td>\n","      <td>-1.276809</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>0.156503</td>\n","      <td>8.003465</td>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>1.135637</td>\n","    </tr>\n","    <tr>\n","      <th>24</th>\n","      <td>1.000000</td>\n","      <td>9.000000</td>\n","      <td>-2.500000</td>\n","      <td>9.000000</td>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>5.000000</td>\n","      <td>1.000000</td>\n","      <td>9.000000</td>\n","      <td>1.000000</td>\n","      <td>7.953200</td>\n","      <td>1.000000</td>\n","      <td>9.000000</td>\n","      <td>1.000000</td>\n","      <td>1.059567</td>\n","    </tr>\n","    <tr>\n","      <th>25</th>\n","      <td>0.921640</td>\n","      <td>9.000000</td>\n","      <td>-2.500000</td>\n","      <td>8.929677</td>\n","      <td>0.740709</td>\n","      <td>7.151827</td>\n","      <td>-2.387043</td>\n","      <td>7.647966</td>\n","      <td>-2.500000</td>\n","      <td>5.000000</td>\n","      <td>0.145513</td>\n","      <td>9.000000</td>\n","      <td>-0.515695</td>\n","      <td>5.355557</td>\n","      <td>1.000000</td>\n","      <td>2.341229</td>\n","    </tr>\n","    <tr>\n","      <th>26</th>\n","      <td>0.555150</td>\n","      <td>6.058563</td>\n","      <td>-1.783375</td>\n","      <td>8.874329</td>\n","      <td>1.000000</td>\n","      <td>8.649072</td>\n","      <td>-1.420631</td>\n","      <td>7.864078</td>\n","      <td>-2.108757</td>\n","      <td>5.000000</td>\n","      <td>-0.461511</td>\n","      <td>8.153280</td>\n","      <td>1.000000</td>\n","      <td>5.438562</td>\n","      <td>-0.052988</td>\n","      <td>5.000000</td>\n","    </tr>\n","    <tr>\n","      <th>27</th>\n","      <td>-2.500000</td>\n","      <td>6.995078</td>\n","      <td>-1.681115</td>\n","      <td>8.474831</td>\n","      <td>0.315464</td>\n","      <td>6.898417</td>\n","      <td>1.000000</td>\n","      <td>5.785397</td>\n","      <td>-2.040477</td>\n","      <td>6.399901</td>\n","      <td>0.191731</td>\n","      <td>6.407787</td>\n","      <td>-2.500000</td>\n","      <td>6.102496</td>\n","      <td>0.711590</td>\n","      <td>2.711636</td>\n","    </tr>\n","    <tr>\n","      <th>28</th>\n","      <td>-2.486268</td>\n","      <td>5.494166</td>\n","      <td>-1.058608</td>\n","      <td>5.525169</td>\n","      <td>0.896542</td>\n","      <td>6.370591</td>\n","      <td>1.000000</td>\n","      <td>5.645055</td>\n","      <td>1.000000</td>\n","      <td>8.198363</td>\n","      <td>0.877842</td>\n","      <td>8.394869</td>\n","      <td>-2.500000</td>\n","      <td>5.166626</td>\n","      <td>0.956411</td>\n","      <td>1.931600</td>\n","    </tr>\n","    <tr>\n","      <th>29</th>\n","      <td>-1.585830</td>\n","      <td>7.137483</td>\n","      <td>-2.500000</td>\n","      <td>9.000000</td>\n","      <td>-1.007128</td>\n","      <td>8.184494</td>\n","      <td>1.000000</td>\n","      <td>7.327064</td>\n","      <td>-1.432210</td>\n","      <td>8.384512</td>\n","      <td>-2.500000</td>\n","      <td>9.000000</td>\n","      <td>-0.795443</td>\n","      <td>7.017673</td>\n","      <td>-0.513190</td>\n","      <td>0.000000</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["    Min_Threshold1  Max_Threshold1  Min_Threshold2  Max_Threshold2  \\\n","0        -2.484193        5.000000       -0.947774        5.000000   \n","1        -1.563720        6.944407       -1.888616        8.209462   \n","2        -2.500000        7.055593       -0.153862        6.418866   \n","3        -0.799507        7.957415       -2.500000        8.798075   \n","4        -0.703108        8.402900        0.573783        5.063458   \n","5         1.000000        6.624750       -0.988286        8.351893   \n","6        -0.231252        5.000000       -2.500000        7.556734   \n","7         1.000000        5.000000       -0.947774        5.000000   \n","8        -2.484193        5.923599       -2.500000        9.000000   \n","9         0.921640        8.612685       -1.406337        8.929677   \n","10        1.000000        5.581058        0.100928        8.173586   \n","11        0.919816        9.000000       -2.500000        7.462543   \n","12        0.555150        6.058563       -1.783375        7.013663   \n","13       -2.500000        6.688769       -1.346836        6.546752   \n","14       -1.739023        5.000000       -2.500000        8.934220   \n","15       -0.193415        5.000000       -1.763051        8.911601   \n","16       -0.868076        6.899075       -2.357094        6.899075   \n","17       -2.337470        6.845520       -1.321150        7.812926   \n","18       -2.500000        9.000000       -1.013972        8.222575   \n","19       -1.438735        8.734394       -2.500000        8.854252   \n","20       -2.500000        6.979361        0.358294        5.867188   \n","21        1.000000        5.000000       -1.614255        7.443017   \n","22        1.000000        6.269356       -1.777364        8.872206   \n","23       -2.500000        6.945668       -2.500000        5.856597   \n","24        1.000000        9.000000       -2.500000        9.000000   \n","25        0.921640        9.000000       -2.500000        8.929677   \n","26        0.555150        6.058563       -1.783375        8.874329   \n","27       -2.500000        6.995078       -1.681115        8.474831   \n","28       -2.486268        5.494166       -1.058608        5.525169   \n","29       -1.585830        7.137483       -2.500000        9.000000   \n","\n","    Min_Threshold3  Max_Threshold3  Min_Threshold4  Max_Threshold4  \\\n","0         1.000000        6.276615        1.000000        5.620068   \n","1         0.864605        5.762497       -0.211366        5.931304   \n","2         0.454577        5.000000       -1.251862        5.000000   \n","3        -1.517770        6.145288        0.116261        8.241050   \n","4        -0.895456        5.655919        1.000000        9.000000   \n","5         1.000000        5.000000        0.012210        5.000000   \n","6        -1.552548        8.048886       -1.412026        9.000000   \n","7         0.388595        5.000000        0.568337        9.000000   \n","8         1.000000        5.000000        1.000000        5.620068   \n","9         1.000000        8.649072        0.731134        8.850317   \n","10        0.751805        6.467111        0.321589        7.635728   \n","11       -1.712006        9.000000       -1.809982        8.499060   \n","12       -0.889651        5.000000        0.082096        9.000000   \n","13       -0.473439        5.000000        1.000000        7.554996   \n","14        1.000000        8.995979       -1.757982        5.000000   \n","15        1.000000        7.707022       -0.475890        5.000000   \n","16       -0.287576        7.569593       -0.334092        6.986923   \n","17        0.817139        6.639562        0.282562        7.477598   \n","18        1.000000        7.940133       -1.281260        8.915542   \n","19        1.000000        7.128696        1.000000        5.000000   \n","20       -2.325128        5.000000       -1.553361        9.000000   \n","21        1.000000        7.216845        0.083578        5.000000   \n","22        0.249281        7.010816        0.013728        5.748915   \n","23        1.000000        6.253239       -1.276809        5.000000   \n","24        1.000000        5.000000        1.000000        5.000000   \n","25        0.740709        7.151827       -2.387043        7.647966   \n","26        1.000000        8.649072       -1.420631        7.864078   \n","27        0.315464        6.898417        1.000000        5.785397   \n","28        0.896542        6.370591        1.000000        5.645055   \n","29       -1.007128        8.184494        1.000000        7.327064   \n","\n","    Min_Threshold5  Max_Threshold5  Min_Threshold6  Max_Threshold6  \\\n","0         1.000000        8.681749        1.000000        8.748657   \n","1        -1.340304        5.000000       -0.209072        7.760002   \n","2         0.040180        6.080559       -0.229976        7.746080   \n","3        -1.815650        9.000000       -0.417373        5.157499   \n","4        -1.305070        7.508770        0.454413        7.108318   \n","5         1.000000        8.375118       -0.068469        9.000000   \n","6        -0.555866        8.464159       -1.252060        5.605640   \n","7         1.000000        7.277051       -2.500000        7.457110   \n","8         1.000000        8.681749        1.000000        8.748657   \n","9        -2.500000        5.000000       -1.154680        8.576633   \n","10        1.000000        6.302015        0.014918        9.000000   \n","11       -1.682064        5.000000       -1.281227        7.699881   \n","12       -2.108757        6.558063        1.000000        5.000000   \n","13       -1.188209        9.000000       -2.289905        6.641053   \n","14        1.000000        7.577163       -1.206961        5.236958   \n","15       -0.408703        7.566323        1.000000        9.000000   \n","16       -0.273360        7.673489        1.000000        7.524166   \n","17       -1.142909        8.961613       -2.215556        6.188587   \n","18        1.000000        8.502546       -1.326409        8.398827   \n","19        1.000000        9.000000        1.000000        7.969859   \n","20        0.687667        5.957286        0.156503        8.179476   \n","21       -0.136759        5.771946       -1.977589        6.975859   \n","22       -2.306030        5.063360        0.096497        5.275726   \n","23        1.000000        5.000000        0.156503        8.003465   \n","24        1.000000        9.000000        1.000000        7.953200   \n","25       -2.500000        5.000000        0.145513        9.000000   \n","26       -2.108757        5.000000       -0.461511        8.153280   \n","27       -2.040477        6.399901        0.191731        6.407787   \n","28        1.000000        8.198363        0.877842        8.394869   \n","29       -1.432210        8.384512       -2.500000        9.000000   \n","\n","    Min_Threshold7  Max_Threshold7  Min_Threshold8  Classification_Threshold  \n","0         1.000000        5.000000        1.000000                  1.792720  \n","1        -2.500000        5.000000        0.022641                  3.074763  \n","2        -1.302351        5.637896        0.485388                  3.112052  \n","3        -0.517976        5.222527        1.000000                  1.638921  \n","4        -2.500000        5.127444       -0.853306                  2.547014  \n","5        -1.533015        9.000000        1.000000                  4.014643  \n","6        -1.408749        8.784983        1.000000                  0.230816  \n","7        -2.173024        5.309422       -1.433755                  0.000000  \n","8         1.000000        5.000000        1.000000                  0.000000  \n","9         0.885723        5.355557        1.000000                  2.341229  \n","10       -1.816842        7.375160       -1.206248                  0.000000  \n","11       -2.111316        5.412747        0.588788                  4.774818  \n","12        1.000000        5.438562        0.984333                  1.596120  \n","13       -1.875815        7.239179       -1.498051                  1.343219  \n","14       -0.941809        9.000000        1.000000                  0.000000  \n","15       -0.634160        5.000000        0.582720                  2.109472  \n","16        1.000000        5.195959        0.804769                  0.000000  \n","17       -0.975260        5.272863        0.955138                  2.324567  \n","18       -1.375298        8.513964        0.873299                  0.310507  \n","19        1.000000        9.000000        1.000000                  4.070031  \n","20       -1.060701        5.000000        1.000000                  5.000000  \n","21       -0.738587        6.145567        1.000000                  2.393012  \n","22       -0.361210        6.258981        1.000000                  2.812965  \n","23        1.000000        5.000000        1.000000                  1.135637  \n","24        1.000000        9.000000        1.000000                  1.059567  \n","25       -0.515695        5.355557        1.000000                  2.341229  \n","26        1.000000        5.438562       -0.052988                  5.000000  \n","27       -2.500000        6.102496        0.711590                  2.711636  \n","28       -2.500000        5.166626        0.956411                  1.931600  \n","29       -0.795443        7.017673       -0.513190                  0.000000  "]},"execution_count":34,"metadata":{},"output_type":"execute_result"}],"source":["FinalSolFront1 = pd.read_csv('Output_FeatSelect/2025_06_02/FinalSolFront1.csv')\n","FinalSolFront1"]},{"cell_type":"code","execution_count":35,"metadata":{"tags":[],"trusted":true},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>0.11486744730180898</th>\n","      <th>0.5689328918023924</th>\n","      <th>0.5769835101709806</th>\n","      <th>0.6562675135438072</th>\n","      <th>0.7235441627194205</th>\n","      <th>0.7455677952387572</th>\n","      <th>0.7413399295453313</th>\n","      <th>0.0</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>0.055988</td>\n","      <td>0.137655</td>\n","      <td>0.167813</td>\n","      <td>0.163647</td>\n","      <td>0.972295</td>\n","      <td>0.762418</td>\n","      <td>0.757744</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>0.028122</td>\n","      <td>0.127703</td>\n","      <td>0.257181</td>\n","      <td>0.152625</td>\n","      <td>0.974849</td>\n","      <td>0.762851</td>\n","      <td>0.758164</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>0.010237</td>\n","      <td>0.097773</td>\n","      <td>0.422932</td>\n","      <td>0.741640</td>\n","      <td>0.653478</td>\n","      <td>0.763039</td>\n","      <td>0.758336</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>0.076793</td>\n","      <td>0.321552</td>\n","      <td>0.321552</td>\n","      <td>0.345694</td>\n","      <td>0.910690</td>\n","      <td>0.757485</td>\n","      <td>0.752787</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>0.086108</td>\n","      <td>0.176624</td>\n","      <td>0.424047</td>\n","      <td>0.032692</td>\n","      <td>0.998022</td>\n","      <td>0.759412</td>\n","      <td>0.754929</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>5</th>\n","      <td>0.000918</td>\n","      <td>0.010230</td>\n","      <td>0.384174</td>\n","      <td>0.970297</td>\n","      <td>0.203511</td>\n","      <td>0.763904</td>\n","      <td>0.759185</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>6</th>\n","      <td>0.079439</td>\n","      <td>0.142453</td>\n","      <td>0.142453</td>\n","      <td>0.995890</td>\n","      <td>0.028866</td>\n","      <td>0.762253</td>\n","      <td>0.757617</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>7</th>\n","      <td>0.110988</td>\n","      <td>0.568933</td>\n","      <td>0.573765</td>\n","      <td>0.989539</td>\n","      <td>0.051844</td>\n","      <td>0.745583</td>\n","      <td>0.741368</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>8</th>\n","      <td>0.066458</td>\n","      <td>0.519844</td>\n","      <td>0.535869</td>\n","      <td>0.440501</td>\n","      <td>0.869332</td>\n","      <td>0.748079</td>\n","      <td>0.743748</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>9</th>\n","      <td>0.081224</td>\n","      <td>0.190697</td>\n","      <td>0.190697</td>\n","      <td>0.987110</td>\n","      <td>0.061679</td>\n","      <td>0.761392</td>\n","      <td>0.756813</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>10</th>\n","      <td>0.001103</td>\n","      <td>0.008608</td>\n","      <td>0.285041</td>\n","      <td>0.011489</td>\n","      <td>0.999749</td>\n","      <td>0.763684</td>\n","      <td>0.758983</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>11</th>\n","      <td>0.041115</td>\n","      <td>0.568916</td>\n","      <td>0.576223</td>\n","      <td>0.706800</td>\n","      <td>0.674886</td>\n","      <td>0.755911</td>\n","      <td>0.751272</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>12</th>\n","      <td>0.058008</td>\n","      <td>0.099159</td>\n","      <td>0.099159</td>\n","      <td>0.828134</td>\n","      <td>0.541934</td>\n","      <td>0.761835</td>\n","      <td>0.757196</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>13</th>\n","      <td>0.083997</td>\n","      <td>0.146677</td>\n","      <td>0.386033</td>\n","      <td>0.994209</td>\n","      <td>0.038219</td>\n","      <td>0.761587</td>\n","      <td>0.757027</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>14</th>\n","      <td>0.096232</td>\n","      <td>0.533098</td>\n","      <td>0.533892</td>\n","      <td>0.540912</td>\n","      <td>0.810839</td>\n","      <td>0.749485</td>\n","      <td>0.745096</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>15</th>\n","      <td>0.038818</td>\n","      <td>0.568916</td>\n","      <td>0.570056</td>\n","      <td>0.969083</td>\n","      <td>0.194892</td>\n","      <td>0.756000</td>\n","      <td>0.751355</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>16</th>\n","      <td>0.047763</td>\n","      <td>0.082238</td>\n","      <td>0.406662</td>\n","      <td>0.448067</td>\n","      <td>0.865608</td>\n","      <td>0.761664</td>\n","      <td>0.757008</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>17</th>\n","      <td>0.083226</td>\n","      <td>0.146660</td>\n","      <td>0.344293</td>\n","      <td>0.982720</td>\n","      <td>0.108879</td>\n","      <td>0.761419</td>\n","      <td>0.756876</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>18</th>\n","      <td>0.111090</td>\n","      <td>0.568933</td>\n","      <td>0.576984</td>\n","      <td>0.029890</td>\n","      <td>0.998338</td>\n","      <td>0.745525</td>\n","      <td>0.741319</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>19</th>\n","      <td>0.052858</td>\n","      <td>0.222765</td>\n","      <td>0.424858</td>\n","      <td>0.011769</td>\n","      <td>0.999721</td>\n","      <td>0.760894</td>\n","      <td>0.756197</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>20</th>\n","      <td>0.066734</td>\n","      <td>0.114238</td>\n","      <td>0.423447</td>\n","      <td>0.415935</td>\n","      <td>0.880468</td>\n","      <td>0.760379</td>\n","      <td>0.755781</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>21</th>\n","      <td>0.024136</td>\n","      <td>0.206072</td>\n","      <td>0.424148</td>\n","      <td>0.246217</td>\n","      <td>0.949002</td>\n","      <td>0.762258</td>\n","      <td>0.757557</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>22</th>\n","      <td>0.089758</td>\n","      <td>0.565300</td>\n","      <td>0.573782</td>\n","      <td>0.755091</td>\n","      <td>0.610532</td>\n","      <td>0.743470</td>\n","      <td>0.739380</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>23</th>\n","      <td>0.111101</td>\n","      <td>0.568933</td>\n","      <td>0.577102</td>\n","      <td>0.752569</td>\n","      <td>0.611182</td>\n","      <td>0.745525</td>\n","      <td>0.741319</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>24</th>\n","      <td>0.052285</td>\n","      <td>0.219698</td>\n","      <td>0.424731</td>\n","      <td>0.440874</td>\n","      <td>0.869109</td>\n","      <td>0.760601</td>\n","      <td>0.755921</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>25</th>\n","      <td>0.066980</td>\n","      <td>0.565300</td>\n","      <td>0.565300</td>\n","      <td>0.010461</td>\n","      <td>0.999805</td>\n","      <td>0.746773</td>\n","      <td>0.742405</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>26</th>\n","      <td>0.067136</td>\n","      <td>0.234169</td>\n","      <td>0.335085</td>\n","      <td>0.284980</td>\n","      <td>0.936983</td>\n","      <td>0.759577</td>\n","      <td>0.754928</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>27</th>\n","      <td>0.105997</td>\n","      <td>0.483958</td>\n","      <td>0.505846</td>\n","      <td>0.618718</td>\n","      <td>0.756989</td>\n","      <td>0.748820</td>\n","      <td>0.744469</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>28</th>\n","      <td>0.057318</td>\n","      <td>0.100983</td>\n","      <td>0.108434</td>\n","      <td>0.992434</td>\n","      <td>0.048147</td>\n","      <td>0.761834</td>\n","      <td>0.757196</td>\n","      <td>0.0</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["    0.11486744730180898  0.5689328918023924  0.5769835101709806  \\\n","0              0.055988            0.137655            0.167813   \n","1              0.028122            0.127703            0.257181   \n","2              0.010237            0.097773            0.422932   \n","3              0.076793            0.321552            0.321552   \n","4              0.086108            0.176624            0.424047   \n","5              0.000918            0.010230            0.384174   \n","6              0.079439            0.142453            0.142453   \n","7              0.110988            0.568933            0.573765   \n","8              0.066458            0.519844            0.535869   \n","9              0.081224            0.190697            0.190697   \n","10             0.001103            0.008608            0.285041   \n","11             0.041115            0.568916            0.576223   \n","12             0.058008            0.099159            0.099159   \n","13             0.083997            0.146677            0.386033   \n","14             0.096232            0.533098            0.533892   \n","15             0.038818            0.568916            0.570056   \n","16             0.047763            0.082238            0.406662   \n","17             0.083226            0.146660            0.344293   \n","18             0.111090            0.568933            0.576984   \n","19             0.052858            0.222765            0.424858   \n","20             0.066734            0.114238            0.423447   \n","21             0.024136            0.206072            0.424148   \n","22             0.089758            0.565300            0.573782   \n","23             0.111101            0.568933            0.577102   \n","24             0.052285            0.219698            0.424731   \n","25             0.066980            0.565300            0.565300   \n","26             0.067136            0.234169            0.335085   \n","27             0.105997            0.483958            0.505846   \n","28             0.057318            0.100983            0.108434   \n","\n","    0.6562675135438072  0.7235441627194205  0.7455677952387572  \\\n","0             0.163647            0.972295            0.762418   \n","1             0.152625            0.974849            0.762851   \n","2             0.741640            0.653478            0.763039   \n","3             0.345694            0.910690            0.757485   \n","4             0.032692            0.998022            0.759412   \n","5             0.970297            0.203511            0.763904   \n","6             0.995890            0.028866            0.762253   \n","7             0.989539            0.051844            0.745583   \n","8             0.440501            0.869332            0.748079   \n","9             0.987110            0.061679            0.761392   \n","10            0.011489            0.999749            0.763684   \n","11            0.706800            0.674886            0.755911   \n","12            0.828134            0.541934            0.761835   \n","13            0.994209            0.038219            0.761587   \n","14            0.540912            0.810839            0.749485   \n","15            0.969083            0.194892            0.756000   \n","16            0.448067            0.865608            0.761664   \n","17            0.982720            0.108879            0.761419   \n","18            0.029890            0.998338            0.745525   \n","19            0.011769            0.999721            0.760894   \n","20            0.415935            0.880468            0.760379   \n","21            0.246217            0.949002            0.762258   \n","22            0.755091            0.610532            0.743470   \n","23            0.752569            0.611182            0.745525   \n","24            0.440874            0.869109            0.760601   \n","25            0.010461            0.999805            0.746773   \n","26            0.284980            0.936983            0.759577   \n","27            0.618718            0.756989            0.748820   \n","28            0.992434            0.048147            0.761834   \n","\n","    0.7413399295453313  0.0  \n","0             0.757744  0.0  \n","1             0.758164  0.0  \n","2             0.758336  0.0  \n","3             0.752787  0.0  \n","4             0.754929  0.0  \n","5             0.759185  0.0  \n","6             0.757617  0.0  \n","7             0.741368  0.0  \n","8             0.743748  0.0  \n","9             0.756813  0.0  \n","10            0.758983  0.0  \n","11            0.751272  0.0  \n","12            0.757196  0.0  \n","13            0.757027  0.0  \n","14            0.745096  0.0  \n","15            0.751355  0.0  \n","16            0.757008  0.0  \n","17            0.756876  0.0  \n","18            0.741319  0.0  \n","19            0.756197  0.0  \n","20            0.755781  0.0  \n","21            0.757557  0.0  \n","22            0.739380  0.0  \n","23            0.741319  0.0  \n","24            0.755921  0.0  \n","25            0.742405  0.0  \n","26            0.754928  0.0  \n","27            0.744469  0.0  \n","28            0.757196  0.0  "]},"execution_count":35,"metadata":{},"output_type":"execute_result"}],"source":["FinalSolFront1 = pd.read_csv('Output_FeatSelect/2025_06_02testSetGoalVals.csv')\n","FinalSolFront1"]},{"cell_type":"code","execution_count":36,"metadata":{"tags":[],"trusted":true},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>0.08703116284632584</th>\n","      <th>0.49049809961992397</th>\n","      <th>0.50250050010002</th>\n","      <th>0.9919028340080972</th>\n","      <th>0.9997780244173141</th>\n","      <th>0.7780581180604547</th>\n","      <th>0.7726068332403987</th>\n","      <th>0.5548286453274796</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>0.088608</td>\n","      <td>0.563113</td>\n","      <td>0.568114</td>\n","      <td>0.949393</td>\n","      <td>0.999778</td>\n","      <td>0.777972</td>\n","      <td>0.772556</td>\n","      <td>0.561279</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>0.072178</td>\n","      <td>0.563513</td>\n","      <td>0.572515</td>\n","      <td>0.997976</td>\n","      <td>0.999778</td>\n","      <td>0.778359</td>\n","      <td>0.772929</td>\n","      <td>0.569618</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>0.081535</td>\n","      <td>0.563513</td>\n","      <td>0.572515</td>\n","      <td>0.983806</td>\n","      <td>0.999556</td>\n","      <td>0.778731</td>\n","      <td>0.773517</td>\n","      <td>0.587727</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>0.099436</td>\n","      <td>0.563113</td>\n","      <td>0.563113</td>\n","      <td>0.985830</td>\n","      <td>0.999778</td>\n","      <td>0.778215</td>\n","      <td>0.772792</td>\n","      <td>0.587727</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>0.081407</td>\n","      <td>0.563113</td>\n","      <td>0.563113</td>\n","      <td>0.987854</td>\n","      <td>0.999778</td>\n","      <td>0.778042</td>\n","      <td>0.772660</td>\n","      <td>0.587727</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>94</th>\n","      <td>0.127607</td>\n","      <td>0.567313</td>\n","      <td>0.575915</td>\n","      <td>0.997976</td>\n","      <td>0.999778</td>\n","      <td>0.777981</td>\n","      <td>0.772576</td>\n","      <td>0.599030</td>\n","    </tr>\n","    <tr>\n","      <th>95</th>\n","      <td>0.127607</td>\n","      <td>0.567313</td>\n","      <td>0.575915</td>\n","      <td>0.995951</td>\n","      <td>0.999556</td>\n","      <td>0.777156</td>\n","      <td>0.771666</td>\n","      <td>0.599030</td>\n","    </tr>\n","    <tr>\n","      <th>96</th>\n","      <td>0.127607</td>\n","      <td>0.567313</td>\n","      <td>0.575915</td>\n","      <td>0.995951</td>\n","      <td>0.999778</td>\n","      <td>0.777458</td>\n","      <td>0.772052</td>\n","      <td>0.599030</td>\n","    </tr>\n","    <tr>\n","      <th>97</th>\n","      <td>0.127607</td>\n","      <td>0.567313</td>\n","      <td>0.576115</td>\n","      <td>0.991903</td>\n","      <td>0.999778</td>\n","      <td>0.777946</td>\n","      <td>0.772545</td>\n","      <td>0.599030</td>\n","    </tr>\n","    <tr>\n","      <th>98</th>\n","      <td>0.127607</td>\n","      <td>0.567313</td>\n","      <td>0.576115</td>\n","      <td>0.997976</td>\n","      <td>0.999778</td>\n","      <td>0.778024</td>\n","      <td>0.772588</td>\n","      <td>0.599030</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>99 rows × 8 columns</p>\n","</div>"],"text/plain":["    0.08703116284632584  0.49049809961992397  0.50250050010002  \\\n","0              0.088608             0.563113          0.568114   \n","1              0.072178             0.563513          0.572515   \n","2              0.081535             0.563513          0.572515   \n","3              0.099436             0.563113          0.563113   \n","4              0.081407             0.563113          0.563113   \n","..                  ...                  ...               ...   \n","94             0.127607             0.567313          0.575915   \n","95             0.127607             0.567313          0.575915   \n","96             0.127607             0.567313          0.575915   \n","97             0.127607             0.567313          0.576115   \n","98             0.127607             0.567313          0.576115   \n","\n","    0.9919028340080972  0.9997780244173141  0.7780581180604547  \\\n","0             0.949393            0.999778            0.777972   \n","1             0.997976            0.999778            0.778359   \n","2             0.983806            0.999556            0.778731   \n","3             0.985830            0.999778            0.778215   \n","4             0.987854            0.999778            0.778042   \n","..                 ...                 ...                 ...   \n","94            0.997976            0.999778            0.777981   \n","95            0.995951            0.999556            0.777156   \n","96            0.995951            0.999778            0.777458   \n","97            0.991903            0.999778            0.777946   \n","98            0.997976            0.999778            0.778024   \n","\n","    0.7726068332403987  0.5548286453274796  \n","0             0.772556            0.561279  \n","1             0.772929            0.569618  \n","2             0.773517            0.587727  \n","3             0.772792            0.587727  \n","4             0.772660            0.587727  \n","..                 ...                 ...  \n","94            0.772576            0.599030  \n","95            0.771666            0.599030  \n","96            0.772052            0.599030  \n","97            0.772545            0.599030  \n","98            0.772588            0.599030  \n","\n","[99 rows x 8 columns]"]},"execution_count":36,"metadata":{},"output_type":"execute_result"}],"source":["FinalSolFront1 = pd.read_csv(\"Output_FeatSelect/2025_06_02/Max Solution Performance Per Generation.csv\")\n","FinalSolFront1"]},{"cell_type":"code","execution_count":37,"metadata":{"tags":[],"trusted":true},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>0.4502567338852205</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>0.460779</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>0.457765</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>0.472083</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>0.467354</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>0.479878</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>94</th>\n","      <td>0.521935</td>\n","    </tr>\n","    <tr>\n","      <th>95</th>\n","      <td>0.518185</td>\n","    </tr>\n","    <tr>\n","      <th>96</th>\n","      <td>0.505430</td>\n","    </tr>\n","    <tr>\n","      <th>97</th>\n","      <td>0.502827</td>\n","    </tr>\n","    <tr>\n","      <th>98</th>\n","      <td>0.505292</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>99 rows × 1 columns</p>\n","</div>"],"text/plain":["    0.4502567338852205\n","0             0.460779\n","1             0.457765\n","2             0.472083\n","3             0.467354\n","4             0.479878\n","..                 ...\n","94            0.521935\n","95            0.518185\n","96            0.505430\n","97            0.502827\n","98            0.505292\n","\n","[99 rows x 1 columns]"]},"execution_count":37,"metadata":{},"output_type":"execute_result"}],"source":["FinalSolFront1 = pd.read_csv(\"Output_FeatSelect/2025_06_02/Average Performance Per Generation.csv\")\n","FinalSolFront1"]},{"cell_type":"code","execution_count":38,"metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["STDOUT:\n"," \n","STDERR:\n"," \n","Return Code: 0\n"]}],"source":["import subprocess\n","\n","upload_cmd = [\n","    \"dx\", \"upload\", \"Output_FeatSelect\",\n","    \"--path\", \"SADIQ/genetic algorithm/solutions/\",\n","    \"--recursive\"\n","]\n","\n","result = subprocess.run(upload_cmd, capture_output=True, text=True)\n","\n","print(\"STDOUT:\\n\", result.stdout)\n","print(\"STDERR:\\n\", result.stderr)\n","print(\"Return Code:\", result.returncode)\n"]},{"cell_type":"code","execution_count":39,"metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["STDOUT:\n"," \n","STDERR:\n"," \n","Return Code: 0\n"]}],"source":["import subprocess\n","\n","upload_cmd = [\n","    \"dx\", \"upload\", \"Output_PreProcess\",\n","    \"--path\", \"SADIQ/genetic algorithm/solutions/\",\n","    \"--recursive\"\n","]\n","\n","result = subprocess.run(upload_cmd, capture_output=True, text=True)\n","\n","print(\"STDOUT:\\n\", result.stdout)\n","print(\"STDERR:\\n\", result.stderr)\n","print(\"Return Code:\", result.returncode)\n","#Output_Model\n","#Output_PreProcessls"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["print(\"Hello\")"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["!ls Output_Model"]},{"cell_type":"code","execution_count":40,"metadata":{"tags":[],"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":["STDOUT:\n"," \n","STDERR:\n"," \n","Return Code: 0\n"]}],"source":["import subprocess\n","\n","upload_cmd = [\n","    \"dx\", \"upload\", \"Output_Model\",\n","    \"--path\", \"SADIQ/genetic algorithm/solutions/\",\n","    \"--recursive\"\n","]\n","\n","result = subprocess.run(upload_cmd, capture_output=True, text=True)\n","\n","print(\"STDOUT:\\n\", result.stdout)\n","print(\"STDERR:\\n\", result.stderr)\n","print(\"Return Code:\", result.returncode)\n","#Output_Model\n","#Output_PreProcessls"]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.5"}},"nbformat":4,"nbformat_minor":4}
